// devexpress-dashboard/model/items/metadata/_image-item.js
/**
* DevExpress Dashboard (_image-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { PropertyCategory } from '../../metadata/_base-metadata';
import { dashboardItemSerializationsInfo } from './_dashboard-item';
export let urlPath = { propertyName: 'urlPath', modelName: '@Url', defaultVal: '', simpleFormAdapterItem: 'textBoxEditor' };
export let image64 = { propertyName: 'image64', modelName: 'ImageData' };
export let imageType = {
    propertyName: 'imageType', displayName: 'DashboardWebStringId.Title.Image', defaultVal: 'linked', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'linked': 'DashboardWebStringId.Title.ImageLinked',
        'embedded': 'DashboardWebStringId.Title.ImageEmbedded'
    }
};
export let sizeMode = {
    propertyName: 'sizeMode', modelName: '@SizeMode', displayName: 'DashboardStringId.SizeMode', defaultVal: 'Clip', simpleFormAdapterItem: 'listEditor', values: {
        'Clip': 'DashboardWebStringId.Image.SizeMode.Clip',
        'Stretch': 'DashboardWebStringId.Image.SizeMode.Stretch',
        'Zoom': 'DashboardWebStringId.Image.SizeMode.Zoom',
        'Squeeze': 'DashboardWebStringId.Image.SizeMode.Squeeze'
    }, category: PropertyCategory.ViewModel
};
export let horizontalAlignment = {
    propertyName: 'horizontalAlignment', modelName: '@HorizontalAlignment', displayName: 'DashboardWebStringId.Image.HorizontalAlignment', defaultVal: 'Center', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'Left': 'DashboardWebStringId.Image.Alignment.Left',
        'Center': 'DashboardWebStringId.Image.Alignment.Center',
        'Right': 'DashboardWebStringId.Image.Alignment.Right'
    }, category: PropertyCategory.ViewModel
};
export let verticalAlignment = {
    propertyName: 'verticalAlignment', modelName: '@VerticalAlignment', displayName: 'DashboardWebStringId.Image.VerticalAlignment', defaultVal: 'Center', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'Top': 'DashboardWebStringId.Image.Alignment.Top',
        'Center': 'DashboardWebStringId.Image.Alignment.Center',
        'Bottom': 'DashboardWebStringId.Image.Alignment.Bottom'
    }, category: PropertyCategory.ViewModel
};
export let imageDashboardItemSerializationsInfo = dashboardItemSerializationsInfo.concat([urlPath, imageType, image64, sizeMode, horizontalAlignment, verticalAlignment]);
