// devexpress-dashboard/model/items/metadata/_text-box-item.js
/**
* DevExpress Dashboard (_text-box-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { valuesPropertyName } from '../../metadata/_base-metadata';
import { _baseInteractivityOptionsMeta } from '../options/interactivity-options';
import { dataDashboardItemSerializationsInfo } from './_data-dashboard-item';
export let textBoxText = {
    propertyName: 'text',
    modelName: 'Text',
    displayName: 'DashboardWebStringId.TextBox.Text',
    defaultVal: '',
    from: (value) => ko.observable(value),
    toJsonObject: (value) => {
        return value.replace(/\0*$/g, '').trim();
    },
};
export let textBoxValues = { propertyName: valuesPropertyName, modelName: 'Values', displayName: 'DashboardStringId.DescriptionValues', array: true };
export let textBoxDashboardItemSerializationsInfo = dataDashboardItemSerializationsInfo.concat([textBoxValues, textBoxText, _baseInteractivityOptionsMeta]);
