// devexpress-dashboard/viewer-parts/widgets/widgets-viewer/cards/_card-arrangement-info.js
/**
* DevExpress Dashboard (_card-arrangement-info.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { newCardMeasurements } from './_new-card-item';
export let margins = newCardMeasurements.margin * 2, paddings = newCardMeasurements.padding * 2, borders = newCardMeasurements.borderWidth * 2;
export class cardArrangementInfo {
    constructor() {
        this.etalonWidth = 11;
        this.etalonHeight = 10;
        this.arrangementDirection = 'column';
        this.ignorePadding = false;
    }
    getMarginOffset() {
        return this.ignorePadding ? margins : 0;
    }
    getScrollableHeight() {
        return Math.max(this.rowCount * this.cellHeight - this.getMarginOffset(), this.viewerHeight);
    }
    getScrollableWidth() {
        return Math.max(this.columnCount * this.cellWidth - this.getMarginOffset(), this.viewerWidth);
    }
    proportionFactor() {
        return Math.abs(((this.etalonWidth - this.etalonHeight) / (this.etalonWidth + this.etalonHeight)) - ((this.cellWidth - this.cellHeight) / (this.cellWidth + this.cellHeight)));
    }
    getCardContentHeight() {
        return this.cellHeight - margins - paddings - borders;
    }
    getCardContentWidth() {
        return this.cellWidth - margins - paddings - borders;
    }
    getCardWidth() {
        return this.cellWidth - margins;
    }
    getCardHeight() {
        return this.cellHeight - margins;
    }
    hasVerticalScroll() {
        return this.getScrollableHeight() > this.viewerHeight;
    }
    hasHorizontalScroll() {
        return this.getScrollableWidth() > this.viewerWidth;
    }
}
