// devexpress-richedit/lib/base/rich-utils/debug/chrome-dev-tools-custom-formatters/layout/anchor-objects-position-info.js
import { AnchorObjectsPositionInfo } from '../../../../../core/layout/document-layout';
import { Pair } from '@devexpress/utils/lib/class/pair';
import { NumberMapUtils } from '@devexpress/utils/lib/utils/map/number';
import { BaseFormatter, StdProps } from '../base-formatter';
import { SpanElement } from '../elements';
export class CCF_AnchorObjectsPositionInfo extends BaseFormatter {
    isHandleObject(obj) {
        return obj instanceof AnchorObjectsPositionInfo;
    }
    getShortDescription(_config) {
        if (NumberMapUtils.isEmpty(this.curr.cache))
            return new SpanElement().setText("Empty");
        return this.stdShow(new StdProps(this.getList()).showAsLine().setStdNumElements());
    }
    hasBody(_, _config) {
        return !NumberMapUtils.isEmpty(this.curr.cache);
    }
    getFullDescription(_config) {
        return this.stdShow(new StdProps(this.getList()).showAsColumn());
    }
    getList() {
        return this.mapNumberAsKeyToListOfPairs(this.curr.cache, (_, obj) => obj.modelPosition, (key, obj) => new Pair(key.toString(), obj.modelPosition));
    }
}
CCF_AnchorObjectsPositionInfo._foo = BaseFormatter.addToFormattersList(new CCF_AnchorObjectsPositionInfo());
