// devexpress-richedit/lib/base/rich-utils/debug/chrome-dev-tools-custom-formatters/layout/page-anchored-object-holder.js
import { PageAnchoredObjectHolder } from '../../../../../core/layout-formatter/floating/page-anchored-object-holder';
import { Pair } from '@devexpress/utils/lib/class/pair';
import { NumberMapUtils } from '@devexpress/utils/lib/utils/map/number';
import { BaseFormatter, StdProps } from '../base-formatter';
export class CCF_PageAnchoredObjectHolder extends BaseFormatter {
    isHandleObject(obj) {
        return obj instanceof PageAnchoredObjectHolder;
    }
    getShortDescription(_config) {
        return this.stdShow(new StdProps(this.getList()).showAsLine().setStdNumElements());
    }
    availableFullDescription(_config) {
        return !NumberMapUtils.isEmpty(this.curr.objects);
    }
    getFullDescription(_config) {
        return this.stdShow(new StdProps(this.getList()).showAsColumn());
    }
    getList() {
        return this.mapNumberAsKeyToListOfPairs(this.curr.objects, (key) => key, (key, obj) => new Pair(key.toString(), obj));
    }
}
CCF_PageAnchoredObjectHolder._foo = BaseFormatter.addToFormattersList(new CCF_PageAnchoredObjectHolder());
