// devexpress-dashboard/viewer-parts/viewer-items/filter-items/_combo-box-element.js
/**
* DevExpress Dashboard (_combo-box-element.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { format as stringFormat } from 'devextreme/core/utils/string';
import dxSelectBox from 'devextreme/ui/select_box';
import dxTagBox from 'devextreme/ui/tag_box';
import { KEY_EXPR } from '../../../data/data-controllers/_filter-element-data-controller';
import { getLocalizationById } from '../../../data/localization/_default';
import { $unwrap } from '../../../data/_jquery-helpers';
import { localizationId } from '../../../data/_localization-ids';
import { ALL_ELEMENT, localizer } from '../../../data/_localizer';
import { RenderHelper } from '../../widgets/_render-helper';
import { filterElementBaseItem } from './_base-element';
var MULTITAG_COUNT = 9;
export let cssComboBoxClassNames = {
    item: 'dx-dashboard-combobox-filter-item',
    multiText: 'dx-dashboard-filter-item-multitext',
    margins: 'dx-dashboard-combobox-margins'
};
export class comboBoxFilterElement extends filterElementBaseItem {
    constructor(container, options) {
        super(container, options);
        this._isFixedHeight = true;
    }
    get dataController() { return this._dataController; }
    set dataController(dataController) { this._dataController = dataController; }
    get filterDataController() { return this._dataController; }
    get _shouldApplySelectionOnInitialRender() {
        return false;
    }
    _setSelectionUnsafe(values) {
        super._setSelectionUnsafe(values);
        this._lock();
        try {
            this.widget.option('value', this._getSelectedKeys());
        }
        finally {
            this._unlock();
        }
    }
    _clearSelectionUnsafe() {
        if (!!this.options.useNeutralFilterMode) {
            this._lock();
            try {
                this.widget.option('value', null);
            }
            finally {
                this._unlock();
            }
        }
    }
    get _isBottomFloatingToolbarPosition() {
        return false;
    }
    get _allowPreview() {
        return true;
    }
    _getWidgetName() {
        return this.isMultiSelectable ? 'dxTagBox' : 'dxSelectBox';
    }
    _createWidgetDiv() {
        let div = super._createWidgetDiv();
        if (this.visualMode === 'content')
            div.classList.add(cssComboBoxClassNames.margins);
        return div;
    }
    _createWidget(div, opts) {
        return this.isMultiSelectable ? new dxTagBox(div, opts) : new dxSelectBox(div, opts);
    }
    _getMinContentHeight() {
        var element = document.createElement('div');
        this._createWidget(element, this._getOptions(false));
        return RenderHelper.getElementBox(element).height;
    }
    _generateInnerBorderClassesUnsafe(element) {
        var classes = super._generateInnerBorderClassesUnsafe(element);
        if (!this._isPaneEmpty()) {
            classes.push(cssComboBoxClassNames.item);
        }
        if (element) {
            if (this._isPaneEmpty()) {
                element.classList.remove(cssComboBoxClassNames.item);
            }
            else {
                element.classList.add(cssComboBoxClassNames.item);
            }
        }
        return classes;
    }
    _getSelectedKeys() {
        const selection = this._dataController.selection;
        if (this.isMultiSelectable) {
            return selection.map(this._dataController.getDataSourceItemKey);
        }
        else {
            return selection && selection.length ? this._dataController.getDataSourceItemKey(selection[0]) : null;
        }
    }
    _getOptions(includeActions) {
        let itemTemplate = (item, _, element) => {
            var node = $unwrap(element);
            var expr = this._getDisplayExpr();
            if (expr === 'html') {
                node.innerHTML = item[expr];
            }
            else {
                node.innerText = item[expr];
            }
        };
        let multiTagPreparingHandler = (args) => {
            if (this._dataController.dataSource.length === args.selectedItems.length)
                args.text = ALL_ELEMENT.text;
            else if (args.selectedItems.length < MULTITAG_COUNT)
                args.cancel = true;
            else
                args.text = stringFormat(getLocalizationById('DashboardWebStringId.FilterElementCheckedComboBoxSelected'), args.selectedItems.length);
        };
        let tagBoxSelectionChangedHandler = e => {
            this._raiseItemClick(e.removedItems.length > 0 ? e.removedItems : e.addedItems);
        };
        let selectBoxValueChangedHandler = e => {
            if (e.value != null) {
                const selectedItem = this._dataController.getDataSourceItemByKey(e.value);
                this._raiseItemClick([selectedItem]);
            }
        };
        let options = {
            dataSource: this.getDataSource(),
            displayExpr: this._getDisplayExpr(),
            valueExpr: KEY_EXPR,
            value: this._getSelectedKeys(),
            placeholder: localizer.getString(localizationId.FilterElementRadioComboBoxNoDataCaption),
            encodeNoDataText: true,
            noDataText: getLocalizationById('DashboardStringId.FilterElementNoDataToDisplay'),
            itemTemplate,
            onMultiTagPreparing: multiTagPreparingHandler,
            maxDisplayedTags: 1,
            selectAllMode: 'allPages',
            searchEnabled: this._enableSearch
        };
        let addtionalOptions = this.isMultiSelectable ?
            {
                onSelectionChanged: includeActions ? tagBoxSelectionChangedHandler : undefined,
                showSelectionControls: this.isMultiSelectable,
                showDropDownButton: true,
                multiline: false
            }
            : {
                onValueChanged: includeActions ? selectBoxValueChangedHandler : undefined
            };
        return Object.assign(Object.assign(Object.assign({}, options), addtionalOptions), { keyExpr: KEY_EXPR, multiSelectEnabled: this.isMultiSelectable, selectAllText: ALL_ELEMENT.text, pageLoadMode: 'scrollBottom', dropDownOptions: {
                container: this.controlContainer
            } });
    }
    _resizeUnsafe() {
        super._resizeUnsafe();
        this.widget.repaint();
    }
}
