// devexpress-dashboard/model/data-sources/metadata/_federation-data-source.js
/**
* DevExpress Dashboard (_federation-data-source.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { itemType, name } from '../../metadata/_base-metadata';
import { dataSourceSerializationsInfo } from './_data-source';
export let aliasProperty = { propertyName: 'alias', modelName: '@Alias' };
export let queryTypeProperty = { propertyName: '_queryType', modelName: '@QueryType' };
export let queryNodeSerializationsInfo = [itemType, aliasProperty, queryTypeProperty];
export let queryNodes = { propertyName: 'queries', modelName: 'Queries', array: true };
export let idProperty = { propertyName: 'id', modelName: '@ID' };
export let contextItemSerializationsInfo = [itemType, idProperty];
export let context = { propertyName: 'context', modelName: 'Context', array: true };
export let dataSourceProperty = { propertyName: 'dataSource', modelName: 'DataSource' };
export let dataMemberProperty = { propertyName: 'dataMember', modelName: 'DataMember' };
export let sourceSerializationsInfo = [itemType, dataSourceProperty, dataMemberProperty, name];
export let sources = { propertyName: 'sources', modelName: 'Sources', array: true };
export let federationDataSourceSerializationsInfo = dataSourceSerializationsInfo.concat([queryNodes, context, sources]);
