// devexpress-dashboard/common/extensions/data-inspector-extension/data-inspector-extension.js
/**
* DevExpress Dashboard (data-inspector-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { getLocalizationById } from '../../../data/localization/_default';
import { DataDashboardItem } from '../../../model';
import { DisposableObject } from '../../../model/disposable-object';
import { DashboardLayoutModeHelper } from '../../../viewer-parts/_dashboard-layout-mode-helper';
import { dashboardToolbarItemNames } from '../../../viewer-parts/widgets/caption-toolbar/caption-toolbar-options';
import { defaultExtensions, extensionNameMap } from '../../control-options';
import { UnderlyingDataProvider } from '../../data/_underlying-data-provider';
import { OptionsManager } from '../../internal/_options-manager';
import { findItemForApi } from '../../internal/_utils';
import { DataInspectorViewModel } from './_data-inspector-view-model';
const name = 'dataInspector';
const nameAlias = 'data-inspector';
export class DataInspectorExtension extends DisposableObject {
    constructor(dashboardControl, options = {}) {
        super();
        this._optionsManager = new OptionsManager();
        this._defaultOptions = {
            allowInspectAggregatedData: false,
            allowInspectRawData: false
        };
        this.name = name;
        this._addContextToolbarItem = (toolbarOptions, dashboardItem) => {
            if ((this._optionsManager.get('allowInspectAggregatedData') || this._optionsManager.get('allowInspectRawData')) && dashboardItem instanceof DataDashboardItem) {
                toolbarOptions.actionItems.push({
                    hint: getLocalizationById('DashboardStringId.ActionShowDataInspector'),
                    icon: 'dx-dashboard-data-inspector',
                    name: dashboardToolbarItemNames.dataInspector,
                    type: 'button',
                    click: () => {
                        this._viewModel.show(dashboardItem, this._optionsManager.get('allowInspectAggregatedData'), this._optionsManager.get('allowInspectRawData'));
                    }
                });
            }
        };
        this._optionsManager.initialize({
            extensionName: name,
            dashboardControl: dashboardControl,
            defaultOptions: this._defaultOptions,
            eventsHolder: this,
            initOptions: options,
            optionChanged: (args) => this._optionChanged(args)
        });
        this._dashboardControl = dashboardControl;
        this._viewModel = new DataInspectorViewModel({
            getContainer: () => DashboardLayoutModeHelper.isMobile ? window.document.body : dashboardControl.getWidgetContainer(),
            onGridContentReady: (e) => { this._optionsManager.raiseEvent('gridContentReady', e); },
            onGridInitialized: (e) => { this._optionsManager.raiseEvent('gridInitialized', e); },
            onDialogShowing: (e) => { this._optionsManager.raiseEvent('dialogShowing', e); },
            onDialogShown: (e) => { this._optionsManager.raiseEvent('dialogShown', e); },
            onDialogHidden: (e) => { this._optionsManager.raiseEvent('dialogHidden', e); },
        });
        this._customTemplate = {
            name: 'dx-dashboard-data-inspector-extension',
            data: this._viewModel
        };
    }
    _optionChanged(args) {
        switch (args.name) {
            case 'allowInspectAggregatedData':
            case 'allowInspectRawData':
                this.hideDataInspector();
                return 'updateItemToolbars';
            default:
                return null;
        }
    }
    start() {
        this._dashboardControl.customTemplates.push(this._customTemplate);
        if (this._dashboardControl._serviceClient()) {
            this._viewModel.setUnderlyingDataProvider(new UnderlyingDataProvider(this._dashboardControl._serviceClient()));
        }
        this.toDispose(this._dashboardControl._serviceClient.subscribe(serviceClient => {
            this._viewModel.setUnderlyingDataProvider(new UnderlyingDataProvider(serviceClient));
        }));
        this._dashboardControl._dashboardContext.addContextToolbarItems.add(this._addContextToolbarItem);
    }
    stop() {
        this._dashboardControl.customTemplates.remove(this._customTemplate);
        this._dashboardControl._dashboardContext.addContextToolbarItems.remove(this._addContextToolbarItem);
    }
    showDataInspector(dashboardItemName, inspectedType) {
        var dashboard = this._dashboardControl.dashboard();
        if (dashboard) {
            var dashboardItem = findItemForApi(dashboard, dashboardItemName, DataDashboardItem);
            this._viewModel.show(dashboardItem, this._optionsManager.get('allowInspectAggregatedData'), this._optionsManager.get('allowInspectRawData'), inspectedType);
        }
    }
    currentInspectedType() {
        return this._viewModel.visible() && this._viewModel.inspectedDataType() || null;
    }
    hideDataInspector() {
        this._viewModel.visible(false);
    }
}
defaultExtensions[nameAlias] = (dashboardControl, options) => new DataInspectorExtension(dashboardControl, options);
extensionNameMap[nameAlias] = name;
