// devexpress-dashboard/designer/items/surfaces/_delta-numeric-format-surface.js
/**
* DevExpress Dashboard (_delta-numeric-format-surface.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { AccordionTab, KnownTabs } from '../../properties-controller/_accordion-tab';
import { SharedComposers } from '../properties-composers/_shared-composers';
export class DeltaNumericFormatSurface {
    constructor(model, propertiesController) {
        this.model = model;
        this.propertiesController = propertiesController;
        this._disposables = [];
        this.propertiesTabs = ko.observableArray([]);
    }
    updatePropertiesTabs() {
        this.propertiesTabs([new AccordionTab(KnownTabs.NumericFormat, this.model.title, SharedComposers.getNumericFormatWrapper(this.model.numericFormat))
        ]);
        this.propertiesController.secondaryModel({
            displayText: this.model.title,
            data: this
        });
    }
    startEditing(args) {
        args.createImmediately = false;
        this.updatePropertiesTabs();
    }
    dispose() {
        this._disposables.forEach((d) => {
            d.dispose();
        });
    }
}
