// devexpress-dashboard/designer/items/surfaces/_map-custom-shapefile-surface.js
/**
* DevExpress Dashboard (_map-custom-shapefile-surface.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { MapCustomShapefilePropertiesComposer } from '../properties-composers/_map-custom-shapefile-properties-composer';
export class MapCustomShapeFileSurface {
    constructor(model, propertiesController) {
        this.model = model;
        this.propertiesController = propertiesController;
        this._disposables = [];
        this.propertiesTabs = ko.observableArray([]);
    }
    startEditing(args) {
        args.createImmediately = false;
        var composer = new MapCustomShapefilePropertiesComposer();
        var tabs = composer.composeTabs(this.model);
        this.propertiesTabs(tabs);
        this.propertiesController.secondaryModel({
            displayText: 'DashboardWebStringId.Map.Area.Custom',
            data: this
        });
    }
    dispose() {
        this._disposables.forEach((d) => {
            d.dispose();
        });
    }
}
