// devexpress-dashboard/designer/items/surfaces/_pie-map-item-surface.js
/**
* DevExpress Dashboard (_pie-map-item-surface.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { DataField } from '../../../model/data-sources/_data-field';
import { PieMapItemPropertiesComposer } from '../properties-composers/_pie-map-item-properties-composer';
import { DataItemCollectionSurface } from '../sections/_data-item-collection-surface';
import { SingleDataItemSurface } from '../sections/_single-data-item-surface';
import { SectionDescriptors, surfaceItemsFactory } from '../_section-descriptors';
import { DataDashboardItemSurface } from './_base-item-surface';
export class PieMapItemSurface extends DataDashboardItemSurface {
    constructor(dashboardItem, dashboardModel, dataSourceBrowser, notificationController) {
        super(dashboardItem, dashboardModel, dataSourceBrowser, notificationController);
    }
    fillSections() {
        this.dataSections.push(new SingleDataItemSurface({
            itemSurface: this,
            sectionInfo: SectionDescriptors.Latitude,
            fieldConstraint: field => DataField.isNumeric(field),
            warning: ko.computed(() => !this.dashboardItem.latitude() && !!(this.dashboardItem.longitude() || this.dashboardItem.values().length || this.dashboardItem.argument() || this.dashboardItem.tooltipDimensions().length || this.dashboardItem.tooltipMeasures().length))
        }));
        this.dataSections.push(new SingleDataItemSurface({
            itemSurface: this,
            sectionInfo: SectionDescriptors.Longitude,
            fieldConstraint: field => DataField.isNumeric(field),
            warning: ko.computed(() => !this.dashboardItem.longitude() && !!(this.dashboardItem.latitude() || this.dashboardItem.values().length || this.dashboardItem.argument() || this.dashboardItem.tooltipDimensions().length || this.dashboardItem.tooltipMeasures().length))
        }));
        this.dataSections.push(new DataItemCollectionSurface(this, SectionDescriptors.Values));
        this.dataSections.push(new SingleDataItemSurface({
            itemSurface: this,
            sectionInfo: SectionDescriptors.Argument,
            warning: ko.computed(() => !this.dashboardItem.argument() && !!(this.dashboardItem.latitude() || this.dashboardItem.longitude() || this.dashboardItem.values().length || this.dashboardItem.tooltipDimensions().length || this.dashboardItem.tooltipMeasures().length))
        }));
        this.dataSections.push(new DataItemCollectionSurface(this, SectionDescriptors.TooltipDimensions));
        this.dataSections.push(new DataItemCollectionSurface(this, SectionDescriptors.TooltipMeasures));
    }
    getPropertiesComposer() {
        return new PieMapItemPropertiesComposer(this._dashboardItemCustomization);
    }
}
surfaceItemsFactory.register('PieMap', PieMapItemSurface);
