// devexpress-dashboard/designer/data-source-browser/rename-data-source-popup/_rename-data-source-viewmodel.js
/**
* DevExpress Dashboard (_rename-data-source-viewmodel.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getLocalizationById } from '../../../data/localization/_default';
import { NameGenerator } from '../../../model/internal/_helper-classes';
export class RenameDataSourceViewModel {
    constructor(dataSourceBrowser) {
        this.title = getLocalizationById('DashboardWebStringId.DataSources.RenameDataSource');
        this.label = getLocalizationById('DashboardWebStringId.DataSources.NewDataSourceName');
        this.isDataSourceNameValid = ko.observable(true);
        this.popupVisible = ko.observable(false);
        this.toolbarItems = [];
        this.dsName = ko.observable('');
        this._dataSourceNameValid = (dsName) => {
            let otherDataSources = this._dataSourceBrowser._dataSources().filter(ds => ds.componentName() !== this._dataSource.componentName());
            return dsName && NameGenerator.isValidName(dsName, otherDataSources, 'name');
        };
        this._dataSourceBrowser = dataSourceBrowser;
        this.toolbarItems = [{
                toolbar: 'bottom',
                location: 'after',
                widget: 'dxButton',
                options: {
                    text: getLocalizationById('DashboardWebStringId.DataSources.Rename'),
                    type: 'default',
                    disabled: ko.computed(() => !this.isDataSourceNameValid()),
                    onClick: () => {
                        this._dataSource.name(this.dsName());
                        this.popupVisible(false);
                    }
                }
            },
            {
                toolbar: 'bottom',
                location: 'after',
                widget: 'dxButton',
                options: {
                    text: getLocalizationById('DashboardWebStringId.DataSources.Cancel'),
                    onClick: () => { this.popupVisible(false); }
                }
            }];
        this.nameValidationRules = [{
                type: 'required'
            }, {
                type: 'custom',
                validationCallback: (params) => {
                    return this._dataSourceNameValid(params.value);
                },
                message: getLocalizationById('DashboardWebStringId.DataSources.DataSourceNameExistsMessage')
            }];
        this.nameTextBoxOptions = {
            value: this.dsName
        };
        this.nameValidatorOptions = {
            isValid: this.isDataSourceNameValid,
            validationRules: this.nameValidationRules
        };
    }
    show(dataSource) {
        this._dataSource = dataSource;
        this.dsName(dataSource.name());
        this.popupVisible(true);
    }
}
