// devexpress-dashboard/common/internal/_interfaces.js
/**
* DevExpress Dashboard (_interfaces.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { getHeight, getWidth } from '../../data/_jquery-helpers';
export var KeyCodes = {
    Esc: 27,
    Delete: 46,
    Z: 90,
    Y: 89,
    S: 83
};
export class SingleItemSizeController {
    constructor(_element, requestRepaint, itemMargins = 0) {
        this._element = _element;
        this.requestRepaint = requestRepaint;
        this.itemMargins = itemMargins;
        this.renderImmediately = true;
    }
    getWidth() {
        return getWidth(this._element) - this.itemMargins;
    }
    getHeight() {
        return getHeight(this._element) - this.itemMargins;
    }
    setConstraints(constraints) {
    }
}
export class SingleTabItemSizeController {
    constructor(requestRepaint, width, height) {
        this.requestRepaint = requestRepaint;
        this.width = width;
        this.height = height;
        this.renderImmediately = true;
    }
    getWidth() {
        return this.width();
    }
    getHeight() {
        return this.height();
    }
    setConstraints(constraints) {
    }
}
