// devexpress-dashboard/common/internal/_update-controller.js
/**
* DevExpress Dashboard (_update-controller.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { distinct } from '../../data/_utils';
export class UpdateController {
    constructor(_actionMap, _mapActionByState) {
        this._actionMap = _actionMap;
        this._mapActionByState = _mapActionByState;
        this._locker = 0;
        this._actions = [];
    }
    beginUpdate() {
        this._locker++;
    }
    endUpdate() {
        if (this._locker > 0)
            this._locker--;
        this.commitUpdate();
    }
    commitUpdate() {
        if (this._locker === 0 && this._actions.length) {
            this.beginUpdate();
            let actions = this._actions;
            this._actions = [];
            while (actions.length) {
                actions = distinct(actions
                    .map(action => this._mapActionByState(this.getControlState.bind(this), action))
                    .map(action => {
                    var masterAction = (this._actionMap[action].masterActions || []).filter(masterAction => actions.indexOf(masterAction) !== -1).pop();
                    return masterAction || action;
                }));
                const actualAction = actions.shift();
                actualAction && this._actionMap[actualAction].action();
            }
            this.endUpdate();
        }
    }
    addAction(actionName) {
        const actualAction = this._mapActionByState(this.getControlState.bind(this), actionName);
        this._actions.push(actualAction);
    }
    switchControlState(controlState) {
        this._state = controlState;
    }
    getControlState() {
        return this._state;
    }
}
