// devexpress-dashboard/designer/ui-widgets/collection-editor/_editable-list-editor-viewmodel.js
/**
* DevExpress Dashboard (_editable-list-editor-viewmodel.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getLocalizationById } from '../../../data/localization/_default';
import { CollectionEditorViewModelBase } from './_collection-editor-viewmodel-base';
export class EditableListEditorViewModel extends CollectionEditorViewModelBase {
    constructor(params, selectedItem) {
        super(params);
        this.edit = () => this.editItemHandler && this.editItemHandler(this.selectedValue);
        this.dataField = params.propertyName;
        this.dataSource = params.dataSource;
        this.getDisplayText = params.getDisplayText;
        this.enableEditItem = params.enableEditItem;
        this.editItemHandler = params.editItemHandler;
        this.selectedItem = selectedItem;
        this.onSelectionChanged = params.onSelectionChanged;
        this.isToolbarVisible = true;
        this.allowEditItem = true;
    }
    getListOptions() {
        this._innerSelection = ko.unwrap(this.selectedItem);
        return Object.assign(Object.assign({}, super.getListOptions()), { dataSource: this.dataSource, selectedItems: [this._innerSelection] });
    }
    _getDisplayText(itemData) {
        return this.getDisplayText ? this.getDisplayText(itemData) : getLocalizationById(ko.unwrap(itemData[this.dataField]));
    }
    _listSelectionChanged() {
        this.selectedItem && this.selectedItem(this.selectedValue);
        this.onSelectionChanged && this.onSelectionChanged(this.selectedValue);
    }
    _updateActionsState() {
        const selection = this.selectedValue;
        this.editEnabled(!!this.enableEditItem ? this.enableEditItem(selection) : !!selection);
    }
}
