// devexpress-dashboard/common/mobile-layout-extension/_mobile-layout-item.js
/**
* DevExpress Dashboard (_mobile-layout-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getWidth } from '../../data/_jquery-helpers';
import { CardItem } from '../../model/items/card/card-item';
import { dataGridItem } from '../../viewer-parts/viewer-items/data-grid-item/_data-grid-item';
import { dateFilterElement } from '../../viewer-parts/viewer-items/filter-items/_date-filter-element';
import { cardsItem } from '../../viewer-parts/viewer-items/_cards-item';
import { cssClasses } from '../../viewer-parts/widgets/caption-toolbar/_caption-toolbar-css-classes';
import { HiddenCaptionToolbar } from '../../viewer-parts/widgets/caption-toolbar/_hidden-caption-toolbar';
import { StaticCaptionToolbar } from '../../viewer-parts/widgets/caption-toolbar/_static-toolbar';
import { DashboardItemContext } from '../viewer/_viewer-interfaces';
export class MobileLayoutItemViewModel {
    constructor(dashboardContext, repaintRequest, dashboardItem, _fullscreenItemModel) {
        this.dashboardContext = dashboardContext;
        this.repaintRequest = repaintRequest;
        this.dashboardItem = dashboardItem;
        this._fullscreenItemModel = _fullscreenItemModel;
        this._cachedItemWidth = 0;
        this._swipeToActionCoef = 100;
        this.itemOffsetInPixels = ko.observable(0);
        this.actionReadyCoef = ko.computed(() => {
            var offsetInPixels = this.itemOffsetInPixels();
            if (offsetInPixels > 0) {
                var coef = Math.abs(offsetInPixels / this._swipeToActionCoef);
                if (coef > 1)
                    coef = 1;
                return coef;
            }
            else {
                return 0;
            }
        });
        this.isReadyForAction = ko.computed(() => {
            return this.actionReadyCoef() >= 1;
        });
        this.maximizeIconOpacity = ko.computed(() => {
            return this.actionReadyCoef();
        });
        this.click = (data, args) => {
            this._performShowFullscreenItem();
            this.unselectItem(data, args);
        };
        this.selectItem = (data, args) => {
            if (this.canMaximizeItem) {
                args.target.classList.add('dx-dashboard-mobile-layout-item-selected');
            }
        };
        this.unselectItem = (data, args) => {
            args.target.classList.remove('dx-dashboard-mobile-layout-item-selected');
        };
        this.swipestart = (data, args) => {
            this._cachedItemWidth = getWidth(this._getStandaloneItemElement(args.target));
        };
        this.swipeupdate = (data, args) => {
            if (args.offset > 0) {
                this.itemOffsetInPixels(args.offset * this._cachedItemWidth);
            }
            else {
                this._reset();
            }
        };
        this.swipeend = (data, args) => {
            if (this.isReadyForAction()) {
                this._performShowFullscreenItem();
            }
            this._reset();
        };
        this.localContext = new DashboardItemContext({
            addContextToolbarItems: (options) => {
                options.stateItems = [];
                options.actionItems = [];
                options.navigationItems = [];
            },
            createCaptionToolbar: (viewerItem, container, controlContainer, popupContainer, viewOptions) => {
                return viewOptions.hasCaption ? new StaticCaptionToolbar(container, controlContainer, popupContainer, viewOptions.encodeHtml, cssClasses.caption, viewOptions.captionToolbarSeparatorRequired, true) :
                    new HiddenCaptionToolbar();
            },
            viewerItemCreated: (dashboardItem, viewerItem) => customizeMobileViewerItems(viewerItem),
            beforeApplyViewerItemOptions: (item, options, isCreation) => {
                options.ParentContainer = undefined;
                setCardAutoArrangementMode(item, options);
            },
            itemCreatingType: 'secondary'
        });
    }
    get canMaximizeItem() {
        return this.dashboardItem._uiState() === 'live';
    }
    _reset() {
        this.itemOffsetInPixels(0);
    }
    _performShowFullscreenItem() {
        if (this.canMaximizeItem) {
            this._fullscreenItemModel.maximizeItem(this.dashboardItem);
        }
    }
    _getStandaloneItemElement(element) {
        return element.parentElement.querySelector('dashboard-standalone-item');
    }
}
export function setCardAutoArrangementMode(item, options) {
    if (item instanceof CardItem) {
        options.ViewModel.ContentDescription.ArrangementMode = 'Auto';
    }
}
export function customizeMobileViewerItems(viewerItem) {
    if (viewerItem instanceof dataGridItem) {
        viewerItem._customizeViewOptions = (options) => {
            options.columnHidingEnabled = true;
            options.allowColumnResizing = false;
        };
        viewerItem._getColumnWidthProperty = () => {
            return 'width';
        };
        viewerItem._getColumnWidthMode = () => {
            return 'AutoFitToContents';
        };
        viewerItem._getDefaultBestCharacterCount = (index) => {
            return 10;
        };
    }
    if (viewerItem instanceof cardsItem) {
        viewerItem._getIgnorePadding = () => {
            return false;
        };
    }
    if (viewerItem instanceof dateFilterElement) {
        viewerItem._mobileLayout = () => {
            return true;
        };
    }
}
