// devexpress-dashboard/common/mobile-layout-extension/_mobile-layout.js
/**
* DevExpress Dashboard (_mobile-layout.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { $unwrap, createJQueryCallbacks, getHeight, getWidth } from '../../data/_jquery-helpers';
import { createSvgIconElement } from '../../data/_utils';
import { getLocalizationById } from '../../data/index.internal';
import { DataDashboardItem } from '../../model/items/data-dashboard-item';
import { DateFilterItem } from '../../model/items/filter-items/date-filter-item';
import { FilterElementItemBase } from '../../model/items/filter-items/filter-element-item-base';
import { DashboardTabPage } from '../../model/items/tab-container-item/dashboard-tab-page';
import { TabContainerItem } from '../../model/items/tab-container-item/tab-container-item';
import { DashboardLayoutGroup } from '../../model/layout/dashboard-layout-group';
import { titleHeight } from '../../viewer-parts/title/_dashboard-title-view-constants';
import { cssClasses } from '../../viewer-parts/widgets/caption-toolbar/_caption-toolbar-css-classes';
import { MobileLayoutCaptionToolbar } from '../../viewer-parts/widgets/caption-toolbar/_mobile-layout-caption-toolbar';
import { dashboardToolbarItemNames } from '../../viewer-parts/widgets/caption-toolbar/caption-toolbar-options';
import { FullscreenItemModel } from '../docking-layout/_docking-layout-fullscreen-item';
import { DashboardItemContext } from '../viewer/_viewer-interfaces';
import { DashboardTitleModel } from '../viewer/title/_dashboard-title-model';
import { DashboardTitleContext } from '../viewer/title/_title-component';
import { createFullscreenItemViewModel } from './_mobile-layout-fullscreen-item';
import { MobileLayoutItemViewModel, customizeMobileViewerItems, setCardAutoArrangementMode } from './_mobile-layout-item';
import { ItemMasterFiltersViewModel, MasterFiltersEditorModel } from './_mobile-layout-master-filters-editor';
export class DashboardMobileLayoutController {
    constructor(dashboard, dashboardContext, findExtension, _encodeHtml = false, viewerApi) {
        this.dashboard = dashboard;
        this.dashboardContext = dashboardContext;
        this._encodeHtml = _encodeHtml;
        this.selectedDashboardItem = ko.computed(() => null);
        this.emptyItemTemplates = ko.observableArray();
        this.selectedLayoutItem = ko.computed(() => null);
        this.visibleItemsProvider = null;
        var fullScreenItemLocalContext = new DashboardItemContext({
            addContextToolbarItems: (options) => {
                var dashboardItem = this.fullscreenItemModel.dashboardItem();
                if (dashboardItem instanceof DataDashboardItem)
                    this.masterFiltersEditorModel.addFilterButton(options.stateItems, dashboardItem);
                options.navigationItems.push({
                    name: dashboardToolbarItemNames.backButton,
                    type: 'button',
                    template: () => {
                        let div = document.createElement('div');
                        div.classList.add(cssClasses.flexParent, cssClasses.ellipsisText);
                        let icon = createSvgIconElement(cssClasses.iconBack);
                        let textDiv = document.createElement('div');
                        textDiv.classList.add(cssClasses.buttonBack, cssClasses.truncated);
                        textDiv.innerText = getLocalizationById('DashboardWebStringId.MobileLayout.Back');
                        div.append(icon, textDiv);
                        return div;
                    },
                    click: () => {
                        this.fullscreenItemModel.restoreDownItem();
                    }
                });
            },
            createCaptionToolbar: (viewerItem, container, controlContainer, popupContainer, viewOptions) => {
                return new MobileLayoutCaptionToolbar(container, controlContainer, popupContainer, viewOptions.encodeHtml, cssClasses.caption, viewOptions.captionToolbarSeparatorRequired);
            },
            viewerItemCreated: (dashboardItem, viewerItem) => customizeMobileViewerItems(viewerItem),
            beforeApplyViewerItemOptions: (item, options, isCreation) => {
                options.ParentContainer = undefined;
                setCardAutoArrangementMode(item, options);
            },
            itemCreatingType: 'secondary'
        });
        var exportExtension = findExtension('dashboardExport');
        if (exportExtension) {
            exportExtension._initializeSecondaryExportItem(fullScreenItemLocalContext);
        }
        this.fullscreenItemModel = new FullscreenItemModel(dashboardContext, fullScreenItemLocalContext);
        this.masterFiltersEditorModel = new MasterFiltersEditorModel();
        this.dashboardTitleContext = new DashboardTitleContext(this._encodeHtml, findExtension, false, viewerApi);
        let flatItems = this._getDashboardItemsInLayoutOrder(dashboard.layout());
        this.items = groupLayoutItems(flatItems).map(item => new DashboardMobileLayoutItem(item.itemComponentNames.map(itemName => dashboard.findItem(itemName))
            .filter(item => !(item instanceof FilterElementItemBase) && !(item instanceof DateFilterItem)), item.groupName))
            .filter(mobileLayoutItem => mobileLayoutItem.dashboardItems.length > 0);
    }
    get allowMaximizeItems() { return true; }
    set allowMaximizeItems(value) { }
    get fullscreenItemProvider() {
        return this.fullscreenItemModel;
    }
    _getDashboardItemsInLayoutOrder(layoutItem) {
        var result = [];
        if (layoutItem.dashboardItem() && !(layoutItem instanceof DashboardLayoutGroup)) {
            let parentContainer = layoutItem.item && layoutItem.item.parentContainer() ? this.dashboard.findItem(layoutItem.item.parentContainer()) : undefined;
            let groupName = this._getGroupName(parentContainer);
            result.push({
                groupName: this._getGroupName(parentContainer),
                groupComponentName: this._getGroupComponentName(parentContainer),
                itemComponentName: layoutItem.dashboardItem()
            });
        }
        if (layoutItem instanceof DashboardLayoutGroup) {
            result = result.concat(layoutItem.childNodes().reduce((acc, value) => acc.concat(this._getDashboardItemsInLayoutOrder(value)), []));
        }
        return result;
    }
    _getGroupName(parentContainer) {
        if (parentContainer) {
            let showCaption = parentContainer instanceof DashboardTabPage ? this._getParentTabContainer(parentContainer).showCaption() : parentContainer.showCaption();
            return showCaption ? parentContainer.name() : parentContainer.name() + groupWithoutCaptionItemKey;
        }
        else {
            return ungroupedItemKey;
        }
    }
    _getGroupComponentName(parentContainer) {
        if (parentContainer) {
            let showCaption = parentContainer instanceof DashboardTabPage ? this._getParentTabContainer(parentContainer).showCaption() : parentContainer.showCaption();
            return showCaption ? parentContainer.componentName() : parentContainer.componentName() + groupWithoutCaptionItemKey;
        }
        else {
            return ungroupedItemKey;
        }
    }
    _getParentTabContainer(tabPage) {
        return this.dashboard.items()
            .filter(item => item instanceof TabContainerItem)
            .filter(tabContainer => tabContainer.tabPages().indexOf(tabPage) !== -1)[0];
    }
}
export class DashboardMobileLayoutItem {
    constructor(dashboardItems, groupName) {
        this.dashboardItems = dashboardItems;
        this.groupName = groupName;
    }
}
export function groupLayoutItems(flatItems) {
    let groupedItems = [];
    let previousGroupedItem;
    flatItems.forEach(item => {
        if (previousGroupedItem && previousGroupedItem.groupComponentName === item.groupComponentName) {
            previousGroupedItem.itemComponentNames.push(item.itemComponentName);
        }
        else {
            previousGroupedItem = {
                groupName: item.groupName,
                groupComponentName: item.groupComponentName,
                itemComponentNames: [item.itemComponentName]
            };
            groupedItems.push(previousGroupedItem);
        }
    });
    return groupedItems;
}
ko.components.register('dashboard-mobile-layout-widget', {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            var disposables = [];
            var layoutModel = params.componentArgs;
            let dashboard = ko.unwrap(layoutModel.dashboard);
            var toolbarHeight = ko.observable(titleHeight);
            var contentToolbarHeight = ko.observable(contentToolbarHeight);
            let viewModel = {};
            let element = componentInfo.element;
            let currentWidth = getWidth(element);
            let currentHeight = getHeight(element);
            let titleWidth = ko.observable(currentWidth);
            let repaintRequest = createJQueryCallbacks();
            var layoutWidget = null;
            let getListSize = () => ({ width: getWidth(element), height: getHeight(element) - toolbarHeight() });
            var resizeControl = () => {
                if (getWidth(element) != currentWidth || getHeight(element) != currentHeight) {
                    if (layoutWidget) {
                        layoutWidget.option(getListSize());
                    }
                    repaintRequest.fire();
                    currentWidth = getWidth(element);
                    currentHeight = getHeight(element);
                    titleWidth(currentWidth);
                }
            };
            let resizeCallback = () => resizeControl();
            window.addEventListener('resize', resizeCallback);
            var interval = setInterval(() => resizeControl(), 300);
            let titleModel = new DashboardTitleModel(layoutModel.dashboardTitleContext, layoutModel.dashboard, customizeToolbarOptions);
            ko.utils.domNodeDisposal.addDisposeCallback(element, function () {
                clearInterval(interval);
                titleModel.dispose();
                window.removeEventListener('resize', resizeCallback);
            });
            var customizeToolbarOptions = (options) => {
                options.contentItems = options.contentItems.filter(item => [dashboardToolbarItemNames.titleFilterIcon, dashboardToolbarItemNames.titleFilterText].indexOf(item.name) === -1);
                layoutModel.masterFiltersEditorModel.addFilterButton(options.actionItems, layoutModel.dashboard);
            };
            var actionToolbarViewModel = createToolbarViewModel(ko.computed(() => {
                return {
                    toolbarOptions: customizeActionToolbar(titleModel.toolbarOptions().toolbarOptions),
                    allowHideEmptyToolbar: true
                };
            }), titleWidth, toolbarHeight, layoutModel.dashboardTitleContext, cssClasses.actionToolbar);
            var contentToolbarViewModel = createToolbarViewModel(ko.computed(() => {
                return {
                    toolbarOptions: customizeContentToolbar(titleModel.toolbarOptions().toolbarOptions),
                    allowHideEmptyToolbar: true
                };
            }), titleWidth, contentToolbarHeight, layoutModel.dashboardTitleContext, cssClasses.contentToolbar);
            return {
                itemMasterFiltersViewModel: new ItemMasterFiltersViewModel(layoutModel.masterFiltersEditorModel, layoutModel.dashboardContext, repaintRequest),
                fullscreenItemViewModel: createFullscreenItemViewModel(layoutModel.fullscreenItemModel, layoutModel.masterFiltersEditorModel, repaintRequest),
                titleViewModel: actionToolbarViewModel,
                layoutViewModel: createLayoutViewModel(element, getListSize(), layoutModel.dashboardContext, repaintRequest, layoutModel.items, layoutModel.fullscreenItemModel, (widget) => layoutWidget = widget, contentToolbarViewModel, layoutModel.dashboard.title.visible())
            };
        }
    },
    template: { element: 'dx-dashboard-mobile-layout-widget' }
});
export let ungroupedItemKey = '_dx_dashboard_ungrouped_mobile_layout_item_key';
export let groupWithoutCaptionItemKey = '_dx_dashboard_group_without_caption_mobile_layout_item_key';
export let dashboardTitleKey = '_dx_dashboard_mobile_layout_title_key';
let mobileLayoutCssClasses = {
    ungroupedItem: 'dx-dashboard-ungrouped-item',
    dashboardDisplayNone: 'dx-dashboard-display-none',
    groupWithoutCaption: 'dx-dashboard-group-without-caption'
};
var createToolbarViewModel = (options, width, height, context, className) => {
    return {
        options: options,
        width: width,
        height: height,
        encodeHtml: context.encodeHtml,
        className: className
    };
};
var customizeActionToolbar = (options) => {
    return {
        staticItems: options.navigationItems,
        actionItems: options.actionItems,
        stateItems: options.stateItems,
        navigationItems: []
    };
};
var customizeContentToolbar = (options) => {
    return {
        staticItems: options.staticItems,
        actionItems: [],
        stateItems: [],
        navigationItems: []
    };
};
export var createLayoutViewModel = (element, listSize, dashboardContext, repaintRequest, items, fullscreenItemModel, getWidgetCallback, titleViewModel, titleVisible) => {
    var title = {
        data: titleViewModel,
        name: 'dx-dashboard-mobile-title'
    };
    let layoutItems = items.map(layoutItem => {
        let items = layoutItem.dashboardItems.map(dashboardItem => {
            return {
                data: new MobileLayoutItemViewModel(dashboardContext, repaintRequest, dashboardItem, fullscreenItemModel),
                name: 'dx-dashboard-mobile-layout-item'
            };
        });
        return {
            key: layoutItem.groupName,
            items: items
        };
    });
    return Object.assign(Object.assign({ indicateLoading: false }, listSize), { activeStateEnabled: false, focusStateEnabled: false, hoverStateEnabled: false, grouped: true, dataSource: (titleVisible ? [{
                key: dashboardTitleKey,
                items: [title]
            }] : []).concat(layoutItems), onContentReady: (e) => {
            getWidgetCallback(e.component);
        }, groupTemplate: (data, index, element) => {
            if (data.key === ungroupedItemKey || data.key === dashboardTitleKey) {
                let className = data.key === dashboardTitleKey || (data.key === ungroupedItemKey && (titleVisible && index === 1 || !titleVisible && index === 0)) ? mobileLayoutCssClasses.dashboardDisplayNone : mobileLayoutCssClasses.ungroupedItem;
                $unwrap(element).classList.add(className);
                return document.createElement('div');
            }
            else if (data.key.search(groupWithoutCaptionItemKey) !== -1) {
                $unwrap(element).classList.add(mobileLayoutCssClasses.groupWithoutCaption);
                return document.createElement('div');
            }
            else {
                let div = document.createElement('div');
                div.innerText = data.key;
                return div;
            }
        } });
};
