// devexpress-dashboard/model/items/range-filter/_range-state-controller.js
/**
* DevExpress Dashboard (_range-state-controller.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { deepExtend } from '../../../data/_jquery-helpers';
import { arrayEquals } from '../../../data/_utils';
import { ItemState, RangeFilterSelection, RangeFilterState } from '../../dashboard-state';
import { DimensionFilterValues } from '../../data-item/_dimension-filter-values';
import { toStringArray } from '../../internal/_date-utils';
export class RangeStateController extends Disposable {
    constructor(item) {
        super();
        this.item = item;
    }
    get defaultDateTimePeriodName() { return this.item.defaultDateTimePeriodName; }
    get currentSelectedDateTimePeriodName() { return this.item.currentSelectedDateTimePeriodName; }
    get dateTimePeriods() { return this.item.dateTimePeriods; }
    get _selectionValues() { return this.item._selectionValues; }
    _getSelectionByPeriod(period) {
        return this.item._getSelectionByPeriod(period);
    }
    _setSelection(stateSelection) {
        this.item._setSelection(stateSelection);
    }
    initialize() {
        if (this.defaultDateTimePeriodName()) {
            this.item._processItemSetPredefinedPeriod(this.defaultDateTimePeriodName());
        }
        this.dateTimePeriods.subscribe(newPeriods => {
            if (!newPeriods.filter(p => p.name() === this.defaultDateTimePeriodName())[0]) {
                this.defaultDateTimePeriodName(undefined);
            }
        });
        var subscribeOnDefaultPeriodNameChange = () => {
            this.item._defaultDateTimePeriodIndexSubscription && this.item._defaultDateTimePeriodIndexSubscription.dispose();
            this.item._defaultDateTimePeriodIndexSubscription = null;
            var newPeriod = this.dateTimePeriods().filter(p => p.name() === this.defaultDateTimePeriodName())[0];
            if (newPeriod) {
                this.item._defaultDateTimePeriodIndexSubscription = newPeriod.name.subscribe((changedName) => {
                    this.defaultDateTimePeriodName(changedName);
                });
            }
        };
        this.addDisposable(this.defaultDateTimePeriodName.subscribe(newName => {
            subscribeOnDefaultPeriodNameChange();
            this.item._processItemSetPredefinedPeriod(newName);
        }));
        subscribeOnDefaultPeriodNameChange();
        this.addDisposable(this.item._state = ko.computed(() => {
            var state = new ItemState();
            if (this.defaultDateTimePeriodName() && this.defaultDateTimePeriodName() !== this.currentSelectedDateTimePeriodName()) {
                state.RangeFilterState = null;
            }
            if (this.currentSelectedDateTimePeriodName()) {
                if (this.currentSelectedDateTimePeriodName() !== this.defaultDateTimePeriodName()) {
                    state.RangeFilterState = {
                        PeriodName: this.currentSelectedDateTimePeriodName()
                    };
                }
            }
            else {
                var selection = toStringArray(this._selectionValues());
                if (selection && selection.length && (!!selection[0][0] || !!selection[0][1])) {
                    state.RangeFilterState = {
                        Selection: {
                            Minimum: selection[0][0],
                            Maximum: selection[0][1]
                        }
                    };
                }
            }
            return state;
        }));
    }
    setState(itemState) {
        var obsoleteItemState = itemState;
        let rangeFilterState = itemState.RangeFilterState;
        let selectedPeriod;
        let selection;
        if (rangeFilterState) {
            let rangeSelection = rangeFilterState.Selection;
            if (rangeFilterState.PeriodName) {
                let period = this.dateTimePeriods().filter(p => p.name() === rangeFilterState.PeriodName)[0];
                if (period) {
                    selectedPeriod = rangeFilterState.PeriodName;
                    selection = this._getSelectionByPeriod(period);
                }
            }
            else if (rangeSelection) {
                if (!!rangeSelection.Minimum || !!rangeSelection.Maximum) {
                    selection = [[rangeSelection.Minimum, rangeSelection.Maximum]];
                }
            }
        }
        else {
            selection = itemState.MasterFilterValues || obsoleteItemState.Selection;
        }
        this.currentSelectedDateTimePeriodName(selectedPeriod);
        this._setSelection(selection);
    }
    removeSelectionFromState(state) {
        var itemState = deepExtend({}, state);
        itemState.RangeFilterState = undefined;
        return itemState;
    }
    setPredefinedPeriodToState(state, periodName) {
        var itemState = deepExtend({}, state);
        itemState.RangeFilterState = { PeriodName: periodName };
        return itemState;
    }
    setSelectionToState(state, selection) {
        var itemState = deepExtend({}, state);
        if (!itemState.RangeFilterState) {
            itemState.RangeFilterState = new RangeFilterState();
        }
        let rangeSelection = new RangeFilterSelection();
        if (selection && selection.length > 0 && selection[0].length > 0) {
            rangeSelection.Minimum = selection[0][0];
            rangeSelection.Maximum = selection[0][1];
        }
        itemState.RangeFilterState.Selection = rangeSelection;
        return itemState;
    }
    extendContentState(content) {
        if (this.currentSelectedDateTimePeriodName()) {
            let selectedPeriod = this.dateTimePeriods().filter(p => p.name() === this.currentSelectedDateTimePeriodName())[0];
            if (selectedPeriod) {
                content.ViewModel.SelectedPeriodIndex = this.dateTimePeriods().indexOf(selectedPeriod);
            }
        }
    }
    _getDisplayFilterValues(limitCount) {
        var metaData = this.item._dataManager() ? this.item._dataManager().getMetaData() : undefined, selection = this.item._outputFilter(), outFilterValues = [], argumentDimension = selection && selection.dimensions && selection.dimensions[0];
        if (argumentDimension) {
            let dimensionId = argumentDimension['@DefaultId'];
            let format = metaData ? metaData.getDimensionFormat(dimensionId) : undefined;
            if (!arrayEquals(selection.range, this.item._fullRange())) {
                let filterValues = new DimensionFilterValues(this.item._getDimensionDisplayName(dimensionId));
                filterValues.Values.push({
                    Type: 'Range',
                    RangeLeft: selection.range[0],
                    RangeRight: selection.range[1],
                    Format: format
                });
                outFilterValues.push(filterValues);
            }
        }
        return outFilterValues;
    }
}
