// devexpress-dashboard/designer/extensions/_available-data-sources-view-model.js
/**
* DevExpress Dashboard (_available-data-sources-view-model.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export class AvailableDataSourcesViewModel {
    constructor(_dataSources, selectedDataSources, uiState, errorState, _showCreateDataSourceWizardDelegate) {
        this._dataSources = _dataSources;
        this.selectedDataSources = selectedDataSources;
        this.uiState = uiState;
        this.errorState = errorState;
        this._showCreateDataSourceWizardDelegate = _showCreateDataSourceWizardDelegate;
        this.showCreateDataSourceWizard = (federationSources) => {
            if (this.canCreateDataSourceWizard) {
                this._showCreateDataSourceWizardDelegate()(federationSources);
            }
        };
    }
    getDataSources(dataSourcesFilter = (() => true)) {
        return this._dataSources().filter(dataSourcesFilter);
    }
    get canCreateDataSourceWizard() {
        return !!this._showCreateDataSourceWizardDelegate();
    }
}
