// devexpress-dashboard/designer/extensions/available-data-sources-extension.js
/**
* DevExpress Dashboard (available-data-sources-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { NotificationController } from '../../common/notification-controller/notificator';
import { getLocalizationById } from '../../data/localization/_default';
import { Dashboard } from '../../model/dashboard';
import { DisposableObject } from '../../model/disposable-object';
import { NameGenerator } from '../../model/internal/_helper-classes';
import { subscribeArrayChange } from '../../model/internal/_knockout-utils';
import { defineObsoleteMethod } from '../../model/internal/_obsolete-helper';
import { AvailableDataSourcesViewModel } from './_available-data-sources-view-model';
const name = 'availableDataSources';
const nameAlias = 'available-data-sources';
export class AvailableDataSourcesExtension extends DisposableObject {
    constructor(dashboardControl) {
        super();
        this.dashboardControl = dashboardControl;
        this.name = name;
        this.templateName = 'dx-dashboard-datasource-available-datasources';
        this.selectedDataSources = ko.observableArray();
        this.dataSources = ko.observableArray();
        this._errorState = ko.observable(null);
        this._uiState = ko.observable('empty');
        let showCreateDataSourceWizardDelegate = ko.computed(() => {
            var dataSourceWizardExtension = (this.dashboardControl.findExtension('dataSourceWizard'));
            if (dataSourceWizardExtension) {
                return (federationSources) => {
                    let createdDataSourcePromise = federationSources ? dataSourceWizardExtension._showDataSourceCreatingDialog(federationSources) : dataSourceWizardExtension.showDataSourceCreatingDialog();
                    createdDataSourcePromise.done(dataSource => {
                        NameGenerator.validateName(dataSource, this.dataSources(), 'name', 1, true);
                        this.dataSources.push(dataSource);
                    });
                };
            }
            else {
                return null;
            }
        });
        this.toDispose(showCreateDataSourceWizardDelegate);
        let isInitialized = false;
        let uiStateComputed = ko.computed(() => {
            if (!isInitialized) {
                this.loadAvailableDataSources();
                isInitialized = true;
            }
            return this._uiState();
        }, this, {
            deferEvaluation: true
        });
        this.toDispose(uiStateComputed);
        this.viewModel = new AvailableDataSourcesViewModel(this.dataSources, this.selectedDataSources, uiStateComputed, this._errorState, showCreateDataSourceWizardDelegate);
        defineObsoleteMethod({
            target: this,
            memberName: 'loadAvaliableDataSources',
            oldMemberDisplayName: 'AvailableDataSourcesExtension.loadAvaliableDataSources',
            newMemberDisplayName: 'AvailableDataSourcesExtension.loadAvailableDataSources',
            action: () => this.loadAvailableDataSources()
        });
    }
    start() {
        if (this.dataSources().length > 0) {
            this.selectedDataSources([this.dataSources()[0]]);
        }
        subscribeArrayChange(this.dataSources, {
            added: (item) => this.selectedDataSources([item])
        });
    }
    stop() {
        this._errorState(null);
    }
    loadAvailableDataSources() {
        if (this.dashboardControl._endpointCollection.dataSourceUrls) {
            this._uiState('loading');
            this.dashboardControl.remoteService.getFromServer(this.dashboardControl._endpointCollection.dataSourceUrls.GetDataSourcesAction)
                .then((result) => {
                let dataSources = deserializeArray(result, (item) => Dashboard._createDataSource(item, new ModelSerializer()))();
                dataSources.forEach(dataSource => {
                    if (!dataSource.name()) {
                        dataSource.name(NameGenerator.generateName(getLocalizationById('DashboardStringId.DefaultDataSourceName') + ' ', dataSources, 'name', 1));
                    }
                });
                this.dataSources(dataSources);
                this._uiState('live');
            }, (errorInfo) => {
                let errorDetail = NotificationController._getDetailedErrorMessage(errorInfo);
                this._errorState({
                    title: getLocalizationById('DashboardWebStringId.DataSources.AvailableDataSourcesError'),
                    detail: errorDetail
                });
                this._uiState('error');
            });
        }
    }
}
designerExtensions[nameAlias] = (dashboardControl, options) => new AvailableDataSourcesExtension(dashboardControl);
extensionNameMap[nameAlias] = name;
