// devexpress-dashboard/designer/extensions/convert.js
/**
* DevExpress Dashboard (convert.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { asyncDebounce } from '../../data/_utils';
import { getLocalizationById } from '../../data/localization/_default';
import { Dashboard } from '../../model/dashboard';
import { DashboardItem } from '../../model/items/dashboard-item';
import { DataDashboardItem } from '../../model/items/data-dashboard-item';
import { itemTypesMap } from '../../model/serializable-model';
import { DashboardItemMenuSizes } from '../items/_dashboard-item-menu';
const name = 'itemConversionPanel';
const nameAlias = 'item-conversion-panel';
export class ConversionPanelExtension {
    constructor(dashboardControl) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._subscriptions = [];
    }
    _contextMenuSubscriber(itemContextMenu) {
        if (!!itemContextMenu) {
            var item = this.dashboardControl._actualLayoutController().selectedDashboardItem();
            if (item instanceof DataDashboardItem) {
                this._updateContextMenu(itemContextMenu, item, this.dashboardControl.dashboard(), this.dashboardControl._serviceClient());
            }
        }
    }
    start() {
        var contextMenuExtension = this.dashboardControl.findExtension('itemMenu');
        if (contextMenuExtension) {
            this._subscriptions.push(contextMenuExtension._itemContextMenu.subscribe(this._contextMenuSubscriber, this));
            this._contextMenuSubscriber(contextMenuExtension._itemContextMenu());
        }
    }
    stop() {
        this._subscriptions.forEach(s => s.dispose());
        this._subscriptions = [];
    }
    _updateContextMenu(itemContextMenu, dashboardItem, dashboard, serviceClient) {
        var toolboxExtension = this.dashboardControl.findExtension('toolbox');
        var _convertItem = asyncDebounce(serviceClient.convertItem.bind(serviceClient), (result) => {
            var tmpDashboard = new Dashboard(result);
            var newItem = tmpDashboard.items()[0];
            newItem.dataSource(dashboardItem.dataSource());
            newItem.componentName(undefined);
            dashboard._changeItem(dashboardItem, newItem);
        });
        const groups = toolboxExtension && toolboxExtension
            .toolboxGroups()
            .map(group => ({
            groupTitle: getLocalizationById(group.title),
            items: group.items()
                .filter(item => !!item.type && item.type !== 'Group' && item.type !== 'Image' && item.type !== 'TabContainer' && !itemTypesMap[item.type].customItemType)
                .map(item => ({
                title: getLocalizationById(item.title),
                icon: item.icon,
                convert: () => _convertItem(dashboardItem, DashboardItem._getCommonItemType(item.type)),
                disabled: dashboardItem.itemType() === item.type,
                type: item.type
            }))
        }))
            .filter(group => group.items.length) || [];
        const viewModel = {
            groups,
            duplicate: () => dashboard._duplicateItem(dashboardItem),
            duplicateActionCaption: getLocalizationById('DashboardWebStringId.Duplicate'),
        };
        itemContextMenu.contextMenuItems.push({
            menuItemId: nameAlias,
            icon: 'dx-dashboard-convert',
            title: 'DashboardWebStringId.ConvertTo',
            panelWidth: DashboardItemMenuSizes.OptionsPanelWidth(),
            templateName: 'dx-dashboard-convert-to',
            detailVisible: ko.observable(false),
            customData: viewModel,
            index: 400
        });
    }
}
designerExtensions[nameAlias] = (dashboardControl, options) => new ConversionPanelExtension(dashboardControl);
extensionNameMap[nameAlias] = name;
