// devexpress-dashboard/designer/extensions/open-dashboard.js
/**
* DevExpress Dashboard (open-dashboard.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import ArrayStore from 'devextreme/data/array_store';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { createJQueryDeferred } from '../../data/_jquery-helpers';
import { DashboardMenuItem } from '../toolbox-extension/toolbox-items';
const name = 'openDashboard';
const nameAlias = 'open-dashboard';
export class OpenDashboardExtension {
    constructor(dashboardControl) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._menuItem = new DashboardMenuItem(nameAlias, 'DashboardWebStringId.DashboardMenuOpen', 108, 79);
        this._menuItem.template = 'dx-dashboard-form-open';
        this._menuItem.data = {
            dashboardsListStore: (loadOptions) => {
                let def = createJQueryDeferred();
                this.dashboardControl.requestDashboardList().done(items => {
                    new ArrayStore(items).load({ filter: loadOptions.searchValue ? [loadOptions.searchExpr, loadOptions.searchOperation, loadOptions.searchValue] : null }).then(function (data) {
                        def.resolve(data);
                    });
                });
                return def;
            },
            openDashboard: (e) => {
                let toolboxExtension = this.dashboardControl.findExtension('toolbox');
                if (toolboxExtension) {
                    toolboxExtension.menuVisible(false);
                }
                let openDashboardCallback = () => this.loadDashboard(e.itemData.id);
                let saveExtension = this.dashboardControl.findExtension('saveDashboard');
                if (saveExtension) {
                    saveExtension.ensureDashboardSaved(openDashboardCallback);
                }
                else {
                    openDashboardCallback();
                }
            }
        };
    }
    start() {
        let toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.push(this._menuItem);
        }
    }
    stop() {
        let toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.remove(this._menuItem);
        }
    }
    loadDashboard(dashboardId) {
        return this.dashboardControl.loadDashboard(dashboardId);
    }
}
designerExtensions[nameAlias] = (dashboardControl, options) => new OpenDashboardExtension(dashboardControl);
extensionNameMap[nameAlias] = name;
