// devexpress-dashboard/designer/extensions/undo-engine-extension.js
/**
* DevExpress Dashboard (undo-engine-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { UndoEngine } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { KeyCodes } from '../../common/internal/_interfaces';
import { createSvgIconElement } from '../../data/_utils';
import { getLocalizationById } from '../../data/localization/_default';
import { DisposableObject } from '../../model/disposable-object';
import { subscribeAndPerform } from '../../model/internal/_knockout-utils';
import { defineObsoleteProperty } from '../../model/internal/_obsolete-helper';
import { UndoEngineContainer } from '../../model/internal/_undo-engine-helper';
import { LegacySettings } from '../../viewer-parts';
import { createToolbarSeparator } from '../toolbar-extension/_toolbar-extension';
import { DashboardToolbarGroup, DashboardToolbarItem } from '../toolbox-extension/toolbox-items';
const name = 'undoRedo';
const nameAlias = 'undo-redo';
export class UndoRedoExtension extends DisposableObject {
    constructor(dashboardControl) {
        super();
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._undoEngine = ko.observable();
        this._predefinedToolbarItems = [];
        this._defaultToolbarItems = [];
        this.isChanged = ko.computed({
            read: () => !!(this._undoEngine() && this._undoEngine().isDirty()),
            write: (value) => this._undoEngine() && this._undoEngine().isDirty(value)
        });
        defineObsoleteProperty({
            target: this,
            memberName: 'undoEngine',
            oldMemberDisplayName: 'UndoRedoExtension.undoEngine',
            warmMessage: 'The undoEngine property is obsolete.',
            action: () => {
                return this._undoEngine;
            }
        });
    }
    reset() {
        if (this._undoEngine()) {
            this._undoEngine().reset();
        }
    }
    processKeyEvent(keyEventType, eventArgs) {
        if (keyEventType === 'keyup' && eventArgs.ctrlKey) {
            if (eventArgs.keyCode === KeyCodes.Z && this._undoEngine().undoEnabled()) {
                this._undoEngine().undo();
                return true;
            }
            else if (eventArgs.keyCode === KeyCodes.Y && this._undoEngine().redoEnabled()) {
                this._undoEngine().redo();
                return true;
            }
        }
        else if (keyEventType === 'keydown' && eventArgs.ctrlKey && [KeyCodes.Z, KeyCodes.Y].indexOf(eventArgs.keyCode) !== -1) {
            eventArgs.preventDefault();
            return true;
        }
        return false;
    }
    start() {
        this.toDispose(this.dashboardControl.dashboard.subscribe(prevDashboard => {
            this.reset();
        }, null, 'beforeChange'));
        this.toDispose(this.dashboardControl.dashboard.subscribe(newDashboard => {
            var undoEngine = new DashboardUndoEngine(newDashboard, null, 'getInfo');
            UndoEngineContainer.undoEngine = undoEngine;
            this._undoEngine(undoEngine);
        }));
        this._controlOptionChangedHandler = this._onControlOptionChanged.bind(this);
        this.dashboardControl.on('optionChanged', this._controlOptionChangedHandler);
        this._setShowConfirmationDialog(this.dashboardControl.showConfirmationOnBrowserClosing);
        let undoButton = this._createToolbarItem('undoButton', () => this.undo(), 'dx-dashboard-undo', getLocalizationById('DashboardStringId.UndoText'), ko.pureComputed(() => !this.undoEnabled()));
        let redoButton = this._createToolbarItem('redoButton', () => this.redo(), 'dx-dashboard-redo', getLocalizationById('DashboardStringId.RedoText'), ko.pureComputed(() => !this.redoEnabled()));
        let separator = Object.assign({ name: 'undoRedoSeparator', location: 'before' }, createToolbarSeparator());
        this._predefinedToolbarItems = [undoButton, redoButton, separator];
        if (LegacySettings.showUndoRedoButtonsInToolbox) {
            let toolboxExtension = this.dashboardControl.findExtension('toolbox');
            if (!!toolboxExtension) {
                let undoItem = new DashboardToolbarItem('undo', () => this.undo(), 'dx-dashboard-undo', 'DashboardStringId.UndoText');
                undoItem.disabled = ko.pureComputed(() => !this.undoEnabled());
                let redoItem = new DashboardToolbarItem('redo', () => this.redo(), 'dx-dashboard-redo', 'DashboardStringId.RedoText');
                redoItem.disabled = ko.pureComputed(() => !this.redoEnabled());
                this._toolboxToolbarGroup = new DashboardToolbarGroup('undo-redo', 'Undo/Redo', 50, undoItem, redoItem);
                toolboxExtension.toolbarGroups.push(this._toolboxToolbarGroup);
            }
        }
        else {
            this._defaultToolbarItems = this._predefinedToolbarItems.map((item, index) => ({ name: item.name, index: UndoRedoExtension._toolbarItemsIndex + index }));
        }
        this._addToolbarItems(this.dashboardControl.findExtension('designerToolbar'));
    }
    _createToolbarItem(name, action, icon, hint, disabled) {
        let subscription;
        return {
            name: name,
            location: 'before',
            widget: 'dxButton',
            cssClass: `dx-dashboard-undo-redo-button ${icon}`,
            options: {
                hint: hint,
                stylingMode: 'text',
                onClick: action,
                focusStateEnabled: false,
                activeStateEnabled: false,
                disabled: disabled(),
                template: (args) => createSvgIconElement(icon),
                onInitialized: (args) => {
                    subscription = subscribeAndPerform(disabled, () => {
                        args.component && args.component.option('disabled', disabled());
                    });
                },
                onDisposed: () => {
                    subscription.dispose();
                }
            }
        };
    }
    _addToolbarItems(toolbar) {
        if (toolbar) {
            toolbar._unregisterDefaultItems(this._defaultToolbarItems);
            toolbar._unregisterPredefinedItems(this._predefinedToolbarItems);
            toolbar._registerDefaultItems(this._defaultToolbarItems);
            toolbar._registerPredefinedItems(this._predefinedToolbarItems);
            toolbar._update();
        }
    }
    _removeToolbarItems(toolbar) {
        if (toolbar) {
            toolbar._unregisterDefaultItems(this._defaultToolbarItems);
            toolbar._unregisterPredefinedItems(this._predefinedToolbarItems);
            toolbar._update();
        }
    }
    undo() {
        this._undoEngine() && this._undoEngine().undo();
    }
    redo() {
        this._undoEngine() && this._undoEngine().redo();
    }
    undoEnabled() {
        return !!(this._undoEngine() && this._undoEngine().undoEnabled());
    }
    redoEnabled() {
        return !!(this._undoEngine() && this._undoEngine().redoEnabled());
    }
    stop() {
        this._removeToolbarItems(this.dashboardControl.findExtension('designerToolbar'));
        let toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension && this._toolboxToolbarGroup) {
            toolboxExtension.toolbarGroups.remove(this._toolboxToolbarGroup);
        }
        this.reset();
        if (this._controlOptionChangedHandler)
            this.dashboardControl.off('optionChanged', this._controlOptionChangedHandler);
        this._setShowConfirmationDialog(false);
    }
    _onControlOptionChanged(args) {
        switch (args.name) {
            case 'showConfirmationOnBrowserClosing':
                this._setShowConfirmationDialog(args.value);
                break;
        }
    }
    _setShowConfirmationDialog(value) {
        if (value) {
            this._beforeWindowUnloadHandler = this._onBeforeWindowUnload.bind(this);
            window.onbeforeunload = this._beforeWindowUnloadHandler;
        }
        else {
            if (this._beforeWindowUnloadHandler)
                window.onbeforeunload = null;
            this._beforeWindowUnloadHandler = null;
        }
    }
    _onBeforeWindowUnload(event) {
        if (this.dashboardControl.isDesignMode() && this.isChanged()) {
            event.preventDefault();
            return event.returnValue = getLocalizationById('DashboardWebStringId.LayoutHasBeenChangedDialogMessage') + ' ' + getLocalizationById('DashboardWebStringId.SaveConfirmationDialogMessage');
        }
    }
}
UndoRedoExtension._toolbarItemsIndex = 0;
class DashboardUndoEngine extends UndoEngine {
    constructor(target, ignoredProperties, getInfoMethodName) {
        super(target, ignoredProperties, getInfoMethodName);
    }
    validatePropertyName(target, propertyName) {
        return propertyName;
    }
}
designerExtensions[nameAlias] = (dashboardControl, options) => new UndoRedoExtension(dashboardControl);
extensionNameMap[nameAlias] = name;
