// devexpress-dashboard/viewer-parts/widgets/widgets-viewer/_arrangement-info.js
/**
* DevExpress Dashboard (_arrangement-info.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export let positioningDirection = { Vertical: 'Vertical', Horizontal: 'Horizontal' };
export class ArrangementInfo {
    constructor(totalItemCount, itemsOnRowCount, itemWidth, itemHeight, itemMargin, direction, options) {
        this.totalItemCount = totalItemCount;
        this.itemsOnRowCount = itemsOnRowCount;
        this.itemWidth = itemWidth;
        this.itemHeight = itemHeight;
        this.direction = direction;
        this.itemMargin = itemMargin;
        this.options = options;
        this.itemsOnColumnCount = Math.ceil(this.totalItemCount / this.itemsOnRowCount);
    }
    getHeight(useMargin) {
        var that = this;
        var margin = useMargin ? 2 * that.itemMargin.Height : 0;
        switch (that.direction) {
            case positioningDirection.Horizontal:
                return that.itemHeight - margin;
            case positioningDirection.Vertical:
                return that.itemWidth - margin;
            default:
                return -1;
        }
    }
    getWidth(useMargin) {
        var that = this, margin = useMargin ? 2 * that.itemMargin.Width : 0;
        switch (that.direction) {
            case positioningDirection.Horizontal:
                return that.itemWidth - margin;
            case positioningDirection.Vertical:
                return that.itemHeight - margin;
            default:
                return 0;
        }
    }
}
