// devexpress-dashboard/viewer-parts/widgets/widgets-viewer/_base-widget-item.js
/**
* DevExpress Dashboard (_base-widget-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { off, on } from 'devextreme/events';
import { deepExtend } from '../../../data/_jquery-helpers';
import { cssClassNames } from './_css-class-names';
function patchStyles(options) {
    options.style.borderWidth = options.borderWidth || 0;
    options.selectionStyle.borderWidth = options.borderWidth || 0;
    options.hoverStyle.borderWidth = options.borderWidth || 0;
}
export class BaseWidgetItem {
    constructor(options) {
        this._options = deepExtend({}, this._getDefaultOptions(), options);
        this._type = this._options.type || 'unknown';
        this._isSelected = !!this._options.isSelected;
        this.tag = this._options.tag;
        patchStyles(this._options);
        this._hoverEnabled = !!this._options.hoverEnabled;
    }
    _getDefaultOptions() {
        return {
            cursor: 'default',
            isSelected: false,
            hoverEnabled: false,
            style: {
                backgroundColor: '#ffffff',
                borderStyle: 'solid',
                borderColor: '#C8C8CC'
            },
            selectionStyle: {
                backgroundColor: 'rgba(95,139,149,0.35)',
                borderStyle: 'solid',
                borderColor: '#5F8B95'
            },
            hoverStyle: {
                backgroundColor: 'rgba(95,139,149,0.25)',
                borderStyle: 'solid',
                borderColor: '#5F8B95'
            }
        };
    }
    dispose() {
        var that = this;
        that._type = null;
        that._isSelected = null;
        that.tag = null;
        that._hoverEnabled = null;
        this._disposeClickEvent();
        this._disposeHoverEvent();
        if (that._itemDiv && that._itemDiv.parentElement) {
            that._itemDiv.parentElement.removeChild(that._itemDiv);
        }
        that._itemDiv = null;
    }
    _getStyle(isSelected) {
        return isSelected ? this._options.selectionStyle : this._options.style;
    }
    select() {
        var that = this;
        that._isSelected = true;
        if (that._itemDiv) {
            that._itemDiv.classList.add(cssClassNames.selectedItem);
        }
    }
    _hover(isHover) {
        var that = this;
        if (that._hoverEnabled) {
            if (isHover) {
                that._itemDiv.classList.add(cssClassNames.hoveredItem);
            }
            else {
                that._itemDiv.classList.remove(cssClassNames.hoveredItem);
            }
        }
    }
    clearSelection() {
        var that = this;
        that._isSelected = false;
        if (that._itemDiv) {
            that._itemDiv.classList.remove(cssClassNames.selectedItem);
        }
    }
    _disposeClickEvent() {
        if (this._itemDiv) {
            off(this._itemDiv, 'click.cardItem');
        }
    }
    _disposeHoverEvent() {
        if (this._itemDiv) {
            off(this._itemDiv, 'mouseenter');
            off(this._itemDiv, 'mouseleave');
        }
    }
    setClickHandler(handler) {
        var that = this;
        if (typeof handler === 'function' && that._itemDiv) {
            this._disposeClickEvent();
            on(that._itemDiv, 'click.cardItem', function () {
                handler.call(null, { item: that });
            });
        }
    }
    setHoverHandler(handler) {
        var that = this;
        if (that._itemDiv) {
            this._disposeHoverEvent();
            on(that._itemDiv, 'mouseenter', function () {
                that._hover(true);
                if (typeof handler === 'function') {
                    handler.call(null, { item: that, state: true });
                }
            });
            on(that._itemDiv, 'mouseleave', function () {
                that._hover(false);
                if (typeof handler === 'function') {
                    handler.call(null, { item: that, state: false });
                }
            });
        }
    }
    draw(width, height, index, commonItemsOptions) {
        if (!this._itemDiv) {
            return this.initDraw(width, height, index);
        }
        return this._itemDiv;
    }
    initDraw(width, height, index) {
        var that = this, itemDiv;
        width = width || 0;
        height = height || 0;
        index = index || 0;
        that.index = index;
        this._disposeClickEvent();
        this._disposeHoverEvent();
        itemDiv = document.createElement('div');
        itemDiv.classList.add('dx-' + that._type);
        itemDiv.style.cursor = that._options.cursor;
        that._itemDiv = itemDiv;
        return itemDiv;
    }
    _applyExtraStyles() {
        if (this._isSelected) {
            this.select();
        }
        else {
            this.clearSelection();
        }
    }
    toggleSelection() {
        if (this._isSelected) {
            this.clearSelection();
        }
        else {
            this.select();
        }
    }
    getWidget() {
        return null;
    }
    finishRender(params) {
        params = params || {};
        var that = this, clickHandler = params.clickHandler, hoverHandler = params.hoverHandler, drawOptions = params.drawOptions;
        that.setClickHandler(clickHandler);
        that.setHoverHandler(hoverHandler);
        that.rerender(drawOptions);
        that._applyExtraStyles();
    }
    getItemContainer() {
        var itemDiv = this._itemDiv;
        if (itemDiv) {
            return itemDiv;
        }
        return;
    }
    _formStyle(selector, cssProperties) {
        var resultCss = '.dx-' + this._type + ' ' + selector + '{', cssProperty;
        for (cssProperty in cssProperties) {
            resultCss += cssProperty + ':' + cssProperties[cssProperty] + 'px;';
        }
        return resultCss + '}';
    }
    getCssStyle(width, height, _commonItemsOptions, prefix) {
        var styleOptions = { height: height };
        if (!this._options.ignoreProportions) {
            styleOptions['width'] = width;
        }
        return prefix + ' ' + this._formStyle('', styleOptions);
    }
    calcCommonItemSpecificOptions(width, height) {
    }
    resize(width, height, index, commonItemsOptions) {
    }
    rerender(drawOptions) {
    }
    detachItem() {
    }
    setHoverEnabledState(hoverEnabled) {
        this._hoverEnabled = hoverEnabled;
    }
}
