// devexpress-dashboard/viewer-parts/viewer/_events-helper.js
/**
* DevExpress Dashboard (_events-helper.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import 'devextreme/ui/widget/ui.widget';
import { itemDataTuple } from '../../data/item-data/_item-data-tuple';
import { ChartItem } from '../../model/items/chart/chart-item';
import { DataDashboardItem } from '../../model/items/data-dashboard-item';
import { GaugeItem } from '../../model/items/gauge/gauge-item';
import { GridItem } from '../../model/items/grid/grid-item';
import { ChartContext, GaugeContext, GridContext } from './item-widget-event-args';
export const createItemClickEventArgs = (dashboardItem, dataPoint, requestUnderlyingDataFunc) => {
    let itemData = dashboardItem._getItemData();
    const patchAxisName = (axisName) => {
        if (axisName === undefined)
            axisName = 'Default';
        return axisName;
    };
    const getAxis = (axisName) => {
        return itemData.getAxis(patchAxisName(axisName));
    };
    const getAxisPoint = (axisName) => {
        axisName = patchAxisName(axisName);
        return getAxis(axisName).getPointByUniqueValues(dataPoint.getValues(axisName));
    };
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getData: function () {
            return itemData;
        },
        getAxisPoint: getAxisPoint,
        getMeasures: () => {
            return itemData.getMeasuresByIds(dataPoint.getMeasureIds());
        },
        getDeltas: function () {
            var ids = dataPoint.getDeltaIds(), deltas = [];
            ids.forEach(id => {
                deltas.push(itemData.getDeltaById(id));
            });
            return deltas;
        },
        getDimensions: (axisName) => {
            return getAxis(axisName).getDimensions();
        },
        requestUnderlyingData: function (onCompleted, dataMembers) {
            var axisPoints = [];
            itemData.getAxisNames().forEach(axisName => {
                axisPoints.push(getAxisPoint(axisName));
            });
            requestUnderlyingDataFunc(dashboardItem.componentName(), {
                axisPoints: axisPoints,
                dataMembers: dataMembers
            }, onCompleted);
        }
    };
};
export const createItemSelectionChangedEventArgs = (dashboardItem, tuples) => {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getCurrentSelection: function () {
            let axisPointTuples = [];
            tuples.forEach(tuple => {
                axisPointTuples.push(dashboardItem._getItemData().createTuple(tuple));
            });
            return axisPointTuples;
        }
    };
};
export const createItemElementCustomColorEventArgs = (dashboardItem, eventArgs) => {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getTargetElement: function () {
            return new itemDataTuple(eventArgs.targetElement);
        },
        getMeasures: function () {
            return dashboardItem._getItemData().getMeasuresByIds(eventArgs.measureIds);
        },
        getColor: function () {
            return eventArgs.color;
        },
        setColor: function (color) {
            eventArgs.color = color;
        }
    };
};
export const createItemInteractivityEventArgs = (dashboardItem, interactivityOptions) => {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getSelectionMode: function () { return interactivityOptions.selectionMode; },
        setSelectionMode: function (value) { interactivityOptions.selectionMode = value; },
        isHighlightingEnabled: function () { return interactivityOptions.hoverEnabled; },
        enableHighlighting: function (value) { interactivityOptions.hoverEnabled = value; },
        getTargetAxes: function () { return interactivityOptions.targetAxes; },
        setTargetAxes: function (value) { interactivityOptions.targetAxes = value; },
        getDefaultSelection: function () {
            let realTuples = [];
            interactivityOptions.defaultSelectedValues.forEach(tuple => {
                realTuples.push(dashboardItem._getItemData().createTuple(tuple));
            });
            return realTuples;
        },
        setDefaultSelection: function (realTuples) {
            let tuples = [], tuple;
            realTuples.forEach(realTuple => {
                tuple = [];
                interactivityOptions.targetAxes.forEach(axisName => {
                    tuple.push({
                        AxisName: axisName,
                        Value: realTuple.getAxisPoint(axisName).getUniquePath()
                    });
                });
                tuples.push(tuple);
            });
            interactivityOptions.defaultSelectedValues = tuples;
        }
    };
};
const createBaseWidgetEventArgs = (dashboardItem) => {
    var itemData = dashboardItem instanceof DataDashboardItem ? dashboardItem._getItemData() : null;
    return {
        itemName: dashboardItem.componentName(),
        itemData: itemData,
        dashboardItem: dashboardItem,
        chartContext: dashboardItem instanceof ChartItem ? new ChartContext(dashboardItem) : null,
        gridContext: dashboardItem instanceof GridItem ? new GridContext(dashboardItem) : null,
        gaugeContext: dashboardItem instanceof GaugeItem ? new GaugeContext(dashboardItem) : null
    };
};
export const createWidgetEventArgs = (dashboardItem, widget) => {
    return Object.assign({ getWidget: function () { return widget; } }, createBaseWidgetEventArgs(dashboardItem));
};
export const createWidgetOptionsEventArgs = (dashboardItem, options) => {
    return Object.assign({ options: options }, createBaseWidgetEventArgs(dashboardItem));
};
