// devexpress-dashboard/viewer-parts/viewer/item-widget-event-args.js
/**
* DevExpress Dashboard (item-widget-event-args.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import 'devextreme/ui/widget/ui.widget';
import { DashboardDataIdField } from '../../data/_common';
export class ChartContext {
    constructor(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    getDashboardItemSeries(seriesOptions) {
        return this._dashboardItem
            .panes()
            .reduce((acc, pane) => acc.concat(pane.series()), [])
            .filter(series => seriesOptions[DashboardDataIdField] && series._getDataId() === seriesOptions[DashboardDataIdField])[0];
    }
}
export class GridContext {
    constructor(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    getDashboardItemColumn(columnOptions) {
        return this._dashboardItem
            .columns()
            .filter(column => columnOptions[DashboardDataIdField] && column._getDataId() === columnOptions[DashboardDataIdField])[0];
    }
}
export class GaugeContext {
    constructor(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    getDashboardItemGauge(gaugeOptions) {
        return this._dashboardItem
            .gauges()
            .filter(gauge => gaugeOptions[DashboardDataIdField] && gauge._getDataId() === gaugeOptions[DashboardDataIdField])[0];
    }
}
