// devexpress-dashboard/viewer-parts/title/_dashboard-title-toolbar-adapter.js
/**
* DevExpress Dashboard (_dashboard-title-toolbar-adapter.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as $ from 'jquery';
import { formatFilterValue } from '../../data/_formatter';
import { $wrap } from '../../data/_jquery-helpers';
import { localizationId } from '../../data/_localization-ids';
import { localizer } from '../../data/_localizer';
import { dashboardToolbarItemNames } from '../widgets/caption-toolbar/caption-toolbar-options';
import { cssClasses, Settings, _convertToExportFormat } from '../widgets/caption-toolbar/_caption-toolbar-css-classes';
import { titleHeight } from './_dashboard-title-view-constants';
import { FilterIconTooltip } from './_filter-icon-tooptip';
export class DashboardTitleToolbarAdapter {
    static getTitleOptions(titleViewModel, masterFilterValues, showExportDialog, showParametersDialog, allowExport) {
        let contentItems = [];
        let actionItems = [];
        if (titleViewModel) {
            let imageViewModel = titleViewModel.LayoutModel.ImageViewModel;
            let exportMenu = {
                title: localizer.getString(localizationId.buttonNames.ExportTo),
                items: [],
                itemClick: (itemData, itemElement, index) => { showExportDialog(_convertToExportFormat(itemData)); },
                type: 'icons'
            };
            exportMenu.items.push(cssClasses.iconExportToPDF);
            if (Settings.allowExportToImage)
                exportMenu.items.push(cssClasses.iconExportToImage);
            exportMenu.items.push(cssClasses.iconExportToExcel);
            if (imageViewModel) {
                contentItems.push({
                    name: dashboardToolbarItemNames.dashboardTitleImage,
                    template: () => {
                        let imageSrs = (imageViewModel.Url ? imageViewModel.Url : 'data:' + imageViewModel.MimeType + ';base64,' + imageViewModel.SourceBase64String);
                        let $image = $.fn.constructor('<img>').attr('src', imageSrs).height(titleHeight + 'px');
                        $image.on('load', () => {
                            let imageHeight = $image.height();
                            if (imageHeight > titleHeight) {
                                $image.width(Math.floor($image.width() * (titleHeight / imageHeight)));
                                $image.height(titleHeight);
                            }
                            else {
                                $image.css('margin-top', Math.ceil((titleHeight - imageHeight) / 2));
                            }
                            $image.css({ visibility: 'visible' });
                        });
                        return $image;
                    }
                });
            }
            if (titleViewModel.Text) {
                contentItems.push({
                    name: dashboardToolbarItemNames.dashboardTitle,
                    type: 'text',
                    text: titleViewModel.Text
                });
            }
            if (titleViewModel.IncludeMasterFilterValues && masterFilterValues && masterFilterValues.length == 1 && masterFilterValues[0].Values.length == 1)
                contentItems.push({
                    name: dashboardToolbarItemNames.titleFilterText,
                    template: () => {
                        return $.fn.constructor('<div/>').text(this._getMasterFilterText(masterFilterValues)).addClass([cssClasses.filterText, cssClasses.ellipsisText].join(' '));
                    }
                });
            if (titleViewModel.IncludeMasterFilterValues && masterFilterValues && (masterFilterValues.length > 1 || (masterFilterValues.length > 0 && masterFilterValues[0].Values.length > 1)))
                contentItems.push({
                    name: dashboardToolbarItemNames.titleFilterIcon,
                    icon: cssClasses.iconFilter,
                    type: 'button',
                    tooltip: {
                        template: (contentElement) => {
                            return FilterIconTooltip.getTooltipContent($wrap(contentElement), masterFilterValues);
                        },
                    }
                });
            if (allowExport)
                actionItems.push({
                    name: dashboardToolbarItemNames.exportMenu,
                    menu: exportMenu,
                    icon: cssClasses.iconItemExport,
                    type: 'menu',
                    hint: localizer.getString(localizationId.buttonNames.ExportTo)
                });
            if (titleViewModel.ShowParametersButton)
                actionItems.push({
                    name: dashboardToolbarItemNames.parameters,
                    click: (element) => { showParametersDialog(); },
                    icon: cssClasses.iconParameters,
                    type: 'button',
                    hint: localizer.getString(localizationId.labelName.ParametersFormCaption)
                });
        }
        return {
            contentItems: contentItems,
            actionItems: actionItems,
            navigationItems: []
        };
    }
    static _getMasterFilterText(masterFilterValues) {
        if (masterFilterValues) {
            masterFilterValues.forEach(dimFilterValue => {
                dimFilterValue.Values.forEach((val) => {
                    val.Format = val.Format || {};
                });
            });
            if (masterFilterValues.length == 1 && masterFilterValues[0].Values.length == 1) {
                return formatFilterValue(masterFilterValues[0].Values[0]);
            }
        }
        return undefined;
    }
}
