// devexpress-dashboard/designer/expression-editor/_flat-item-provider-wrapper.js
/**
* DevExpress Dashboard (_flat-item-provider-wrapper.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { createJQueryDeferred, jqueryWhenArray } from '../../data/_jquery-helpers';
export class FlatItemProviderWrapper {
    constructor(itemsProvider) {
        this.itemsProvider = itemsProvider;
    }
    getItems(path) {
        var def = createJQueryDeferred();
        this.itemsProvider.getItems(path).done(items => {
            if (!path.fullPath) {
                def.resolve(items);
            }
            else {
                var leafs = items.filter(i => !i.isList);
                var nodePromises = items
                    .filter(i => i.isList)
                    .map(node => {
                    var newPath = {
                        dataSource: path.dataSource,
                        fullPath: path.fullPath + '.' + node.name,
                        path: path.path + '.' + node.name,
                        id: path.id,
                        pathParts: path.pathParts && path.pathParts.concat(node.name) || undefined,
                        ref: path.ref
                    };
                    return this.itemsProvider.getItems(newPath);
                });
                if (nodePromises.length) {
                    jqueryWhenArray(nodePromises).done((...itemArray) => {
                        var itemFlatList = itemArray.reduce((acc, items) => acc.concat(items), []);
                        def.resolve(leafs.concat(itemFlatList));
                    });
                }
                else {
                    def.resolve(leafs);
                }
            }
        });
        return def.promise();
    }
}
