// devexpress-dashboard/viewer-parts/widgets/caption-toolbar/_caption-toolbar-arranger.js
/**
* DevExpress Dashboard (_caption-toolbar-arranger.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
function arrangeToolbarItems(staticItems = [], actionItems = [], stateItems = [], navigationItems = [], showStaticItemsOnCenter = false) {
    let items = [];
    navigationItems.forEach(button => {
        items.push(Object.assign(Object.assign({}, button), { location: staticItems.length !== 0 ? 'before' : 'after' }));
    });
    staticItems.forEach(button => {
        items.push(Object.assign(Object.assign({}, button), { location: showStaticItemsOnCenter ? 'center' : 'before' }));
    });
    actionItems.forEach(button => {
        items.push(Object.assign(Object.assign({}, button), { location: 'after' }));
    });
    if (actionItems.length !== 0 && stateItems.length !== 0) {
        items.push({
            location: 'after',
            isSeparator: true
        });
    }
    stateItems.forEach(button => {
        items.push(Object.assign(Object.assign({}, button), { location: 'after' }));
    });
    return items;
}
export function arrangeFloatingToolbarItems(itemOptions) {
    return arrangeToolbarItems(undefined, itemOptions.actionItems, itemOptions.stateItems, itemOptions.navigationItems);
}
export function arrangeHoveredToolbarItems(itemOptions, containerHovered, disabled) {
    return arrangeToolbarItems(itemOptions.staticItems, containerHovered && !disabled ? itemOptions.actionItems : undefined, itemOptions.stateItems, itemOptions.navigationItems);
}
export function arrangeTitleToolbarItems(itemOptions, showStaticItemsOnCenter) {
    return arrangeToolbarItems(itemOptions.staticItems, itemOptions.actionItems, itemOptions.stateItems, itemOptions.navigationItems, showStaticItemsOnCenter);
}
export function arrangeStaticToolbarItems(itemOptions, disabled) {
    return arrangeToolbarItems(itemOptions.staticItems, disabled ? undefined : itemOptions.actionItems, itemOptions.stateItems, itemOptions.navigationItems);
}
