// devexpress-dashboard/model/format-rules/style-settings/metadata/_appearance-settings.js
/**
* DevExpress Dashboard (_appearance-settings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Color } from '../../../color';
import { emptyStyleType, styleSettingsBaseSerializationsInfo } from './_style-settings-base';
export let appearanceType = {
    propertyName: 'appearanceType', modelName: '@AppearanceType', displayName: 'DashboardWebStringId.ConditionalFormatting.AppearanceType', defaultVal: emptyStyleType, simpleFormAdapterItem: 'selectBoxEditor', values: {
        'None': 'DashboardStringId.FormatConditionAppearanceNone',
        'Custom': 'DashboardStringId.FormatConditionAppearanceCustom',
        'PaleRed': 'DashboardStringId.FormatConditionAppearancePaleRed',
        'PaleYellow': 'DashboardStringId.FormatConditionAppearancePaleYellow',
        'PaleGreen': 'DashboardStringId.FormatConditionAppearancePaleGreen',
        'PaleBlue': 'DashboardStringId.FormatConditionAppearancePaleBlue',
        'PalePurple': 'DashboardStringId.FormatConditionAppearancePalePurple',
        'PaleCyan': 'DashboardStringId.FormatConditionAppearancePaleCyan',
        'PaleOrange': 'DashboardStringId.FormatConditionAppearancePaleOrange',
        'PaleGray': 'DashboardStringId.FormatConditionAppearancePaleGray',
        'Red': 'DashboardStringId.FormatConditionAppearanceRed',
        'Yellow': 'DashboardStringId.FormatConditionAppearanceYellow',
        'Green': 'DashboardStringId.FormatConditionAppearanceGreen',
        'Blue': 'DashboardStringId.FormatConditionAppearanceBlue',
        'Purple': 'DashboardStringId.FormatConditionAppearancePurple',
        'Cyan': 'DashboardStringId.FormatConditionAppearanceCyan',
        'Orange': 'DashboardStringId.FormatConditionAppearanceOrange',
        'Gray': 'DashboardStringId.FormatConditionAppearanceGray',
        'GradientRed': 'DashboardStringId.FormatConditionAppearanceGradientRed',
        'GradientYellow': 'DashboardStringId.FormatConditionAppearanceGradientYellow',
        'GradientGreen': 'DashboardStringId.FormatConditionAppearanceGradientGreen',
        'GradientBlue': 'DashboardStringId.FormatConditionAppearanceGradientBlue',
        'GradientPurple': 'DashboardStringId.FormatConditionAppearanceGradientPurple',
        'GradientCyan': 'DashboardStringId.FormatConditionAppearanceGradientCyan',
        'GradientOrange': 'DashboardStringId.FormatConditionAppearanceGradientOrange',
        'GradientTransparent': 'DashboardStringId.FormatConditionAppearanceGradientTransparent',
        'FontBold': 'DashboardStringId.FormatConditionAppearanceFontBold',
        'FontItalic': 'DashboardStringId.FormatConditionAppearanceFontItalic',
        'FontUnderline': 'DashboardStringId.FormatConditionAppearanceFontUnderline',
        'FontGrayed': 'DashboardStringId.FormatConditionAppearanceFontGrayed',
        'FontRed': 'DashboardStringId.FormatConditionAppearanceFontRed',
        'FontYellow': 'DashboardStringId.FormatConditionAppearanceFontYellow',
        'FontGreen': 'DashboardStringId.FormatConditionAppearanceFontGreen',
        'FontBlue': 'DashboardStringId.FormatConditionAppearanceFontBlue'
    }
};
export let fontStyle = {
    propertyName: 'fontStyle', modelName: '@FontStyle', displayName: 'DashboardWebStringId.ConditionalFormatting.FontStyle', defaultVal: null, simpleFormAdapterItem: 'selectBoxEditor', values: {
        'Regular': 'DashboardWebStringId.ConditionalFormatting.FontRegular',
        'Bold': 'DashboardStringId.FormatConditionAppearanceFontBold',
        'Italic': 'DashboardStringId.FormatConditionAppearanceFontItalic',
        'Underline': 'DashboardStringId.FormatConditionAppearanceFontUnderline',
        'Strikeout': 'DashboardStringId.FormatConditionAppearanceFontStrikeout'
    }
};
export let fontFamily = { propertyName: 'fontFamily', modelName: '@FontFamily', displayName: 'DashboardWebStringId.ConditionalFormatting.FontFamily', defaultVal: 'Segoe UI', simpleFormAdapterItem: 'textBoxEditor' };
export let backColor = { propertyName: 'backColor', modelName: '@BackColor', displayName: 'DashboardWebStringId.ConditionalFormatting.Appearance.BackColor', defaultVal: null, from: Color._colorFromModel, toJsonObject: Color._colorToModel, simpleFormAdapterItem: 'numberBoxEditor' };
export let foreColor = { propertyName: 'foreColor', modelName: '@ForeColor', displayName: 'DashboardWebStringId.ConditionalFormatting.Appearance.ForeColor', defaultVal: null, from: Color._colorFromModel, toJsonObject: Color._colorToModel, simpleFormAdapterItem: 'numberBoxEditor' };
export let appearanceSettingsSerializationsInfo = styleSettingsBaseSerializationsInfo.concat([appearanceType, backColor, foreColor, fontFamily, fontStyle]);
