// devexpress-dashboard/designer/toolbar-extension/toolbar-extension.js
/**
* DevExpress Dashboard (toolbar-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { CancellationToken } from '../../common/_helpers';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { OptionsManager, mergeOptions } from '../../common/internal/_options-manager';
import { createJQueryDeferred } from '../../data/_jquery-helpers';
import { type } from '../../data/_utils';
import { DisposableObject } from '../../model/disposable-object';
import { toolboxConstants } from '../toolbox-extension/_toolbox-extension';
import { createToolbarSeparator } from './_toolbar-extension';
import { ToolbarKoViewModel, toolbarAnimationTime } from './_toolbar-view-model';
const name = 'designerToolbar';
export class DesignerToolbarExtension extends DisposableObject {
    constructor(dashboardControl, options = {}) {
        super();
        this._defaultItems = [];
        this._predefinedItems = [];
        this.name = 'designerToolbar';
        this._optionsManager = new OptionsManager();
        this._showPanelAsync = (options, cancellationToken) => {
            var def = createJQueryDeferred();
            if (cancellationToken.canceled)
                def.reject();
            else {
                this._viewModel.showPanel();
                setTimeout(() => {
                    if (!cancellationToken.canceled) {
                        options.surfaceTop = this._viewModel.height;
                        def.resolve(options);
                    }
                    else {
                        def.reject();
                    }
                }, toolbarAnimationTime);
            }
            return def.promise();
        };
        this._hidePanelAsync = (options, cancellationToken) => {
            var def = createJQueryDeferred();
            if (cancellationToken.canceled)
                def.reject();
            else {
                this._viewModel.hidePanel();
                setTimeout(() => {
                    if (!cancellationToken.canceled) {
                        options.surfaceTop = 0;
                        def.resolve(options);
                    }
                    else {
                        def.reject();
                    }
                }, toolbarAnimationTime);
            }
            return def.promise();
        };
        this._dashboardControl = dashboardControl;
        this._optionsManager.initialize({
            extensionName: name,
            dashboardControl: dashboardControl,
            defaultOptions: {},
            eventsHolder: this,
            initOptions: options,
            optionChanged: (args) => {
                switch (args.name) {
                    case 'items':
                    case 'onPreparing':
                        this._update();
                        break;
                }
                return 'noop';
            }
        });
        this._predefinedItems = [
            Object.assign(Object.assign({}, createToolbarSeparator()), { name: 'separator' })
        ];
        let cancelableDesignerToViewerAction = {
            orderNo: 40,
            action: (options) => this._hidePanelAsync(options, CancellationToken.None),
            cancelableAction: this._hidePanelAsync
        };
        this.designerToViewerAction = cancelableDesignerToViewerAction;
        let cancelableViewerToDesignerAction = {
            orderNo: 40,
            action: (options) => this._showPanelAsync(options, CancellationToken.None),
            cancelableAction: this._showPanelAsync
        };
        this.viewerToDesignerAction = cancelableViewerToDesignerAction;
        this._viewModel = new ToolbarKoViewModel(() => this._dashboardControl.getWidgetContainer(), () => this._update());
        this.template = {
            data: this._viewModel,
            name: 'dx-dashboard-toolbar-extension'
        };
    }
    start() {
        let toolbox = this._dashboardControl.findExtension('toolbox');
        this._viewModel.left(toolbox ? toolboxConstants.leftPanelWidth : 0);
        this._extensionChangeSubscription && this._extensionChangeSubscription.dispose();
        this._extensionChangeSubscription = this._dashboardControl.subscribeExtensionsChanged({
            added: (extension) => {
                if (extension.name === 'toolbox')
                    this._viewModel.left(toolboxConstants.leftPanelWidth);
            },
            deleted: (extension) => {
                if (extension.name === 'toolbox')
                    this._viewModel.left(0);
            }
        });
        if (this._dashboardControl.isDesignMode())
            this._viewModel.showPanel();
        else
            this._viewModel.hidePanel();
        this._dashboardControl.surfaceTop(this._viewModel.height);
        this._update();
    }
    stop() {
        this._unsubscribe();
        this._viewModel.hidePanel();
        this._dashboardControl.surfaceTop(0);
    }
    _unsubscribe() {
        this._extensionChangeSubscription && this._extensionChangeSubscription.dispose();
    }
    _registerDefaultItems(defaultItems) {
        this._defaultItems.push(...defaultItems);
    }
    _unregisterDefaultItems(defaultItems) {
        this._defaultItems = this._defaultItems.filter(item => !defaultItems.some(n => n === item));
    }
    _registerPredefinedItems(predefinedItems) {
        this._predefinedItems.push(...predefinedItems);
    }
    _unregisterPredefinedItems(predefinedItems) {
        this._predefinedItems = this._predefinedItems.filter(item => !predefinedItems.some(n => n === item));
    }
    _update() {
        let args = {
            component: this._dashboardControl,
            dashboard: this._dashboardControl.dashboard(),
            items: []
        };
        let items = this._optionsManager.get('items');
        if (items) {
            args.items = items;
        }
        else {
            args.items = this._defaultItems
                .filter(defaultItem => type.isDefined(defaultItem.index))
                .sort((item1, item2) => item1.index - item2.index);
        }
        this._optionsManager.raiseEvent('preparing', args);
        let dxToolbarItems = args.items.reduce((acc, sourceItem) => {
            let item = typeof sourceItem === 'string' ? { name: sourceItem } : sourceItem;
            let defaultItem = this._predefinedItems.find(predefinedItem => predefinedItem.name === item.name);
            if (defaultItem) {
                let resultItem = {};
                mergeOptions(resultItem, defaultItem);
                mergeOptions(resultItem, item);
                resultItem.name = undefined;
                return acc.concat(resultItem);
            }
            return acc.concat([item]);
        }, []);
        this._viewModel.setToolbarItems(dxToolbarItems);
    }
    dispose() {
        this.stop();
        this._optionsManager.dispose();
        super.dispose();
    }
}
designerExtensions[name] = (dashboardControl, options) => new DesignerToolbarExtension(dashboardControl, options);
extensionNameMap[name] = name;
