// devexpress-dashboard/model/items/pie/metadata/_pie-item.js
/**
* DevExpress Dashboard (_pie-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { contentArrangementMode, contentLineCount, parseBool, valuesPropertyName } from '../../../metadata/_base-metadata';
import { chartItemBaseSerializationsInfo } from '../../metadata/_chart-item-base';
import { _chartItemInteractivityOptionsMeta } from '../../options/interactivity-options';
export let pieValues = { propertyName: valuesPropertyName, modelName: 'Values', displayName: 'DashboardWebStringId.Binding.Values', array: true };
export let labelContentType = {
    propertyName: 'labelContentType', modelName: '@LabelContentType', displayName: 'DashboardWebStringId.Chart.LabelContent', defaultVal: 'ArgumentAndPercent', simpleFormAdapterItem: 'listEditor', values: {
        'None': 'DashboardWebStringId.Pie.LabelContentType.None',
        'Argument': 'DashboardWebStringId.Pie.LabelContentType.Argument',
        'Percent': 'DashboardWebStringId.Pie.LabelContentType.Percent',
        'ArgumentAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ArgumentAndPercent',
        'Value': 'DashboardWebStringId.Pie.LabelContentType.Value',
        'ArgumentAndValue': 'DashboardWebStringId.Pie.LabelContentType.ArgumentAndValue',
        'ValueAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ValueAndPercent',
        'ArgumentValueAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ArgumentValueAndPercent'
    }
};
export let tooltipContentType = {
    propertyName: 'tooltipContentType', modelName: '@TooltipContentType', displayName: 'DashboardWebStringId.Chart.TooltipContent', defaultVal: 'ArgumentValueAndPercent', simpleFormAdapterItem: 'listEditor', values: {
        'None': 'DashboardWebStringId.Pie.LabelContentType.None',
        'Argument': 'DashboardWebStringId.Pie.LabelContentType.Argument',
        'Percent': 'DashboardWebStringId.Pie.LabelContentType.Percent',
        'ArgumentAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ArgumentAndPercent',
        'Value': 'DashboardWebStringId.Pie.LabelContentType.Value',
        'ArgumentAndValue': 'DashboardWebStringId.Pie.LabelContentType.ArgumentAndValue',
        'ValueAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ValueAndPercent',
        'ArgumentValueAndPercent': 'DashboardWebStringId.Pie.LabelContentType.ArgumentValueAndPercent'
    }
};
export let pieType = {
    propertyName: 'pieType', modelName: '@PieType', displayName: 'DashboardWebStringId.Pie.Type', defaultVal: 'Pie', valuesArray: [
        { value: 'Pie', displayValue: 'DashboardWebStringId.Pie' },
        { value: 'Donut', displayValue: 'DashboardWebStringId.Pie.Donut' }
    ]
};
export let showPieCaptions = { propertyName: 'showPieCaptions', modelName: '@ShowPieCaptions', displayName: 'DashboardWebStringId.Pie.ShowCaptions', defaultVal: true, simpleFormAdapterItem: 'onOffButtonGroupEditor', from: parseBool };
export let labelPosition = {
    propertyName: 'labelPosition', modelName: '@LabelPosition', displayName: 'DashboardWebStringId.Pie.LabelPosition', defaultVal: 'Outside', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'Outside': 'DashboardWebStringId.Pie.LabelPositionOutside',
        'Inside': 'DashboardWebStringId.Pie.LabelPositionInside'
    }
};
export let pieDashboardItemSerializationsInfo = chartItemBaseSerializationsInfo.concat([pieValues, labelContentType, tooltipContentType, pieType, contentArrangementMode, contentLineCount, showPieCaptions, _chartItemInteractivityOptionsMeta, labelPosition]);
