// devexpress-dashboard/model/data-item/dimension.js
/**
* DevExpress Dashboard (dimension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { DataField } from '../data-sources/_data-field';
import { safeComputed } from '../internal/_knockout-utils';
import { DataItem } from './data-item';
import { dimensionItemSerializationsInfo } from './metadata/_dimension';
export class Dimension extends DataItem {
    constructor(dataItemJSON = {}, serializer = new ModelSerializer()) {
        super(dataItemJSON, serializer);
        this.addDisposable(this.realSortMode = ko.computed({
            read: () => {
                if (this.sortMeasure())
                    return this.sortMeasure();
                var valueModes = ['DXDisplayText', 'DXValue', 'DXID', 'DXKey'];
                var actualModes = ['DisplayText', 'Value', 'ID', 'Key'];
                return valueModes[actualModes.indexOf(this.sortMode())];
            },
            write: (val) => {
                var valueModes = ['DXDisplayText', 'DXValue', 'DXID', 'DXKey'];
                var actualModes = ['DisplayText', 'Value', 'ID', 'Key'];
                if (valueModes.indexOf(val) !== -1) {
                    this.sortMode(actualModes[valueModes.indexOf(val)]);
                    this.sortMeasure(null);
                }
                else {
                    this.sortMode('Value');
                    this.sortMeasure(val);
                }
            }
        }));
        this._actualDateTimeGroupInterval = safeComputed({ dateTimeGroupInterval: this.dateTimeGroupInterval, dataMember: this.dataMember }, (args) => {
            return DataField.isOlap(args.dataMember) ? 'None' : args.dateTimeGroupInterval;
        });
    }
    getInfo() {
        return dimensionItemSerializationsInfo;
    }
    grabFrom(dataItem) {
        super.grabFrom(dataItem);
        if (!(dataItem instanceof Dimension))
            return;
        this.sortOrder(dataItem.sortOrder());
        this.sortMeasure(dataItem.sortMeasure());
        this.sortMode(dataItem.sortMode());
        this.dateTimeGroupInterval(dataItem.dateTimeGroupInterval());
        this.textGroupInterval(dataItem.textGroupInterval());
        this.isDiscreteNumericScale(dataItem.isDiscreteNumericScale());
        this.groupChildValues(dataItem.groupChildValues());
        this.coloringMode(dataItem.coloringMode());
        this.topNOptionsEnabled(dataItem.topNOptionsEnabled());
        this.topNOptionsMode(dataItem.topNOptionsMode());
        this.topNOptionsCount(dataItem.topNOptionsCount());
        this.topNOptionsMeasureName(dataItem.topNOptionsMeasureName());
        this.topNOptionsShowOthers(dataItem.topNOptionsShowOthers());
        this.groupIndex(dataItem.groupIndex());
    }
    isDefinitionEquals(dataItem) {
        return super.isDefinitionEquals(dataItem) && (dataItem instanceof Dimension) &&
            this.dateTimeGroupInterval() === dataItem.dateTimeGroupInterval() && this.textGroupInterval() == dataItem.textGroupInterval();
    }
    _getDefaultItemType() {
        return 'Dimension';
    }
}
