// devexpress-dashboard/common/viewer/_viewer-interfaces.js
/**
* DevExpress Dashboard (_viewer-interfaces.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { createJQueryCallbacks } from '../../data/_jquery-helpers';
export class DashboardItemContext {
    constructor(options = {}) {
        this.addContextToolbarItems = createJQueryCallbacks();
        this.viewerItemCreated = createJQueryCallbacks();
        this.viewerItemDispose = createJQueryCallbacks();
        this.beforeApplyViewerItemOptions = createJQueryCallbacks();
        this.captionToolbarCreated = createJQueryCallbacks();
        options.addContextToolbarItems && this.addContextToolbarItems.add(options.addContextToolbarItems);
        options.viewerItemCreated && this.viewerItemCreated.add(options.viewerItemCreated);
        options.viewerItemDispose && this.viewerItemDispose.add(options.viewerItemDispose);
        options.beforeApplyViewerItemOptions && this.beforeApplyViewerItemOptions.add(options.beforeApplyViewerItemOptions);
        this.createCaptionToolbar = options.createCaptionToolbar;
        this.itemFactory = options.itemFactory;
        this.ignoreDesignMode = options.ignoreDesignMode;
        this.disabled = options.disabled;
        this.visualMode = options.visualMode;
        this.boundaryContainer = options.boundaryContainer;
        this.itemCreatingType = options.itemCreatingType;
    }
}
