// devexpress-dashboard/model/dashboard.js
/**
* DevExpress Dashboard (dashboard.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { getLocalizationById } from '../data/localization/_default';
import { ColorSchemeEntry } from './colorization/color-scheme-entry';
import { getCustomPropertiesSerializationInfo } from './custom-properties/_custom-properties-utils';
import { DashboardState } from './dashboard-state';
import { _baseDataSourceTypesMap } from './data-sources/_data-source-factory-base';
import { EFDataSource } from './data-sources/ef-data-source';
import { FederationDataSource } from './data-sources/federation-data-source';
import { MongoDBDataSource } from './data-sources/mongodb-data-source';
import { SqlDataSource } from './data-sources/sql-data-source';
import { arrayEquals } from './internal/_array-utils';
import { DashboardUniqueNameGenerator } from './internal/_dashboard-component-name-generator';
import { getItemTitle } from './internal/_dashboard-item_helper';
import { serializeDate, toStringArray, tryConvertToDateTime } from './internal/_date-utils';
import { NameGenerator } from './internal/_helper-classes';
import { subscribeArrayChange } from './internal/_knockout-utils';
import { wrapWithUndoRedo } from './internal/_undo-engine-helper';
import { createDashboardItem } from './items/_dashboard-item-factory';
import { DataDashboardItem } from './items/data-dashboard-item';
import { GroupItem } from './items/group/group-item';
import { TabContainerItem } from './items/tab-container-item/tab-container-item';
import { DashboardLayoutCreator } from './layout/_dashboard-layout-creator';
import { deserializeDashboardLayoutNode } from './layout/_layout-utils';
import { DashboardLayoutGroup, DashboardLayoutRootGroup } from './layout/dashboard-layout-group';
import { DashboardLayoutItem } from './layout/dashboard-layout-item';
import { dashboardSerializationsInfo } from './metadata/_dashboard';
import { _getParametersQuery, Parameter } from './parameters/parameter';
import { SerializableModel } from './serializable-model';
export class Dashboard extends SerializableModel {
    constructor(dashboardJSON = {}, serializer = new ModelSerializer()) {
        super(dashboardJSON, serializer);
        this.dashboardJSON = dashboardJSON;
        this.dataSources = deserializeArray(dashboardJSON.DataSources, (item) => Dashboard._createDataSource(item, serializer));
        this.groups = deserializeArray(dashboardJSON.Groups, (group) => createDashboardItem(group, serializer));
        this.items = deserializeArray(dashboardJSON.Items, (item) => createDashboardItem(item, serializer));
        this._tabPages = ko.observableArray();
        this.addDisposable(this._componentNameGenerator = new DashboardUniqueNameGenerator('componentName', 1, this.dataSources, this.items, this.groups, this._tabPages));
        this.addDisposable(this._allItems = ko.pureComputed(() => [...this.items(), ...this.groups(), ...this._tabPages()]));
        this.addDisposable(subscribeArrayChange(this.dataSources, {
            added: (newDataSource) => {
                if (newDataSource instanceof FederationDataSource) {
                    newDataSource.context()
                        .forEach(contextItem => {
                        var dataSourceToAdd = contextItem.source();
                        if (dataSourceToAdd) {
                            if (!this.dataSources().includes(dataSourceToAdd)) {
                                if (!dataSourceToAdd.componentName() || this.dataSources().find(ds => ds.componentName() === dataSourceToAdd.componentName()) === undefined)
                                    this.dataSources.push(dataSourceToAdd);
                                else {
                                }
                            }
                            newDataSource.sources()
                                .filter(source => source.dataSource() === contextItem.id())
                                .forEach(source => source.dataSource(dataSourceToAdd.componentName()));
                        }
                    });
                    newDataSource.context.removeAll();
                }
                if (!newDataSource.name()) {
                    newDataSource.name(NameGenerator.generateName(getLocalizationById('DashboardStringId.DefaultDataSourceName') + ' ', ko.unwrap(this.dataSources), 'name', 1));
                }
            }
        }));
        this.addDisposable(ko.computed(() => {
            var newTabPagesList = this.items()
                .filter(item => item instanceof TabContainerItem)
                .reduce((acc, tabPage) => acc.concat(tabPage.tabPages()), []);
            this._tabPages()
                .filter(removedTabPage => newTabPagesList.indexOf(removedTabPage) === -1)
                .forEach(removedTabPage => this._tabPages.remove(removedTabPage));
            newTabPagesList
                .filter(newTabPage => this._tabPages().indexOf(newTabPage) === -1)
                .forEach(newTabPage => this._tabPages.push(newTabPage));
        }));
        this.addDisposable(this._componentNameGenerator);
        this.parameters = ko.observableArray();
        this.parameters(deserializeArray(dashboardJSON.Parameters, (item) => new Parameter(item, serializer))());
        this.colorScheme = deserializeArray(dashboardJSON.ColorScheme, (item) => new ColorSchemeEntry(item, serializer));
        this.addDisposable(subscribeArrayChange(this.dataSources, {
            deleted: this._processDeleteDataSource.bind(this)
        }));
        this.addDisposable(this._queryParameters = ko.computed(() => {
            return _getParametersQuery(this.parameters());
        }));
        this.addDisposable(this._dataDashboardItems = ko.computed(() => {
            return this.items().filter(item => item instanceof DataDashboardItem);
        }));
        this.addDisposable(this._masterFilterItems = ko.computed(() => {
            return this._dataDashboardItems()
                .filter(item => item._isMasterFilter() && this._interactivityGroupPathToRoot(item).every(containerItem => containerItem.interactivityOptions.isMasterFilter()));
        }));
        this.addDisposable(ko.computed(() => {
            this._dataDashboardItems().forEach(detailItem => {
                var masterItems = this._dataDashboardItems()
                    .filter(item => item !== detailItem && item._isMasterFilter())
                    .filter(item => {
                    if (item.isMasterFilterCrossDataSource())
                        return true;
                    return item.dataSource() == detailItem.dataSource() && item.dataMember() == detailItem.dataMember();
                })
                    .filter(item => {
                    let masterGroups = this._interactivityGroupPathToRoot(item).reverse();
                    let detailGroups = this._interactivityGroupPathToRoot(detailItem).reverse();
                    while (masterGroups.length && detailGroups.length && masterGroups[0] === detailGroups[0]) {
                        masterGroups.shift();
                        detailGroups.shift();
                    }
                    if (masterGroups.some(masterGroup => !masterGroup.interactivityOptions.isMasterFilter())) {
                        return false;
                    }
                    if (detailGroups.length === 0) {
                        return !detailItem._isIgnoreMasterFilter();
                    }
                    else {
                        return !detailGroups[0].interactivityOptions.ignoreMasterFilters();
                    }
                });
                if (!arrayEquals(detailItem._masterFilterItems.peek(), masterItems)) {
                    detailItem._masterFilterItems(masterItems);
                }
            });
        }));
        this.addDisposable(this._state = ko.computed({
            read: () => {
                var state = new DashboardState();
                if (this.parameters().length > 0) {
                    var parametersState = {};
                    this.parameters().forEach(param => {
                        parametersState[param.name()] = toStringArray(param._actualValue());
                    });
                    if (Object.keys(parametersState).length) {
                        state.Parameters = parametersState;
                    }
                }
                if (this.items().length > 0) {
                    var itemsState = {};
                    this.items().forEach(item => {
                        var itemState = item._state();
                        if (itemState && Object.keys(itemState).length) {
                            itemsState[item.componentName()] = itemState;
                        }
                    });
                    if (Object.keys(itemsState).length) {
                        state.Items = itemsState;
                    }
                }
                return state;
            },
            write: (dashboardState) => {
                var obsoleteDashboardState = dashboardState;
                var parametersState = dashboardState.Parameters || obsoleteDashboardState.parameters;
                if (parametersState) {
                    this.parameters().forEach(parameter => {
                        if (parametersState[parameter.name()] !== undefined) {
                            var parameterValue = parametersState[parameter.name()];
                            if (Array.isArray(parameterValue)) {
                                parameterValue = parameterValue.map(tryConvertToDateTime);
                            }
                            else {
                                parameterValue = tryConvertToDateTime(parameterValue);
                            }
                            parameter._value(parameterValue);
                        }
                    });
                }
                var itemsState = dashboardState.Items || obsoleteDashboardState.items;
                if (itemsState) {
                    Object.keys(itemsState).forEach((name) => {
                        var dashboardItemModel = this.findItem(name);
                        if (dashboardItemModel) {
                            dashboardItemModel._setState(itemsState[name]);
                        }
                    });
                }
            }
        }));
        var item = new DashboardLayoutRootGroup(this, dashboardJSON.LayoutTree || {}, serializer);
        this.layout(item);
        this.addDisposable(item);
        this.addDisposable(this._colorableItems = ko.pureComputed(() => this._dataDashboardItems().filter(dataDashboardItem => dataDashboardItem._isGloballyColored)));
    }
    static _createDataSource(dataSourceJSON, serializer) {
        var itemTypeName = dataSourceJSON['@ItemType'];
        var itemType = Dashboard._dataSourceTypesMap[itemTypeName];
        return new itemType(dataSourceJSON, serializer);
    }
    get stateString() {
        var state = this._state();
        return Object.keys(state).length ? JSON.stringify(state) : '';
    }
    set stateString(stateVal) {
        if (!stateVal)
            return;
        this._state(JSON.parse(stateVal));
    }
    dispose() {
        var _a, _b;
        super.dispose();
        (_a = this.title) === null || _a === void 0 ? void 0 : _a.dispose();
        (_b = this.customProperties) === null || _b === void 0 ? void 0 : _b.dispose();
        [this.items, this.groups, this._tabPages, this.parameters, this.colorScheme].forEach((array) => {
            this.disposeObservableArray(array);
        });
        this.removeProperties();
        this.disposed = true;
    }
    getInfo() {
        return dashboardSerializationsInfo.concat(getCustomPropertiesSerializationInfo(this));
    }
    getJSON() {
        return new ModelSerializer({ useRefs: false, serializeDate: serializeDate }).serialize(this, this.getInfo());
    }
    findItem(itemId) {
        var item = this.items().filter(filterItem => filterItem.componentName() === itemId)[0];
        if (!item) {
            item = this.groups().filter(filterItem => filterItem.componentName() === itemId)[0];
        }
        if (!item) {
            item = this._tabPages().filter(filterItem => filterItem.componentName() === itemId)[0];
        }
        return item;
    }
    rebuildLayout(clientWidth = 1, clientHeight = 1) {
        new DashboardLayoutCreator(clientWidth, clientHeight, this).rebuildLayout();
    }
    _getDisplayDashboardItem(tabPage) {
        if (!tabPage || !tabPage.showItemAsTabPage())
            return tabPage;
        let itemsOnTabPage = this.items().concat(this.groups()).filter(item => item.parentContainer() === tabPage.componentName());
        return itemsOnTabPage.length === 1 && !(itemsOnTabPage[0] instanceof GroupItem) ? itemsOnTabPage[0] : tabPage;
    }
    _changeItem(oldItem, newItem) {
        var dashboardLayoutItem = this.layout().findLayoutItem(oldItem);
        this.items.replace(oldItem, newItem);
        if (dashboardLayoutItem) {
            dashboardLayoutItem.item = newItem;
        }
    }
    _duplicateItem(item) {
        var dashboardLayoutItem = this.layout().findLayoutItem(item);
        var serializer = new ModelSerializer({ useRefs: false });
        var itemJSON = serializer.serialize(item);
        var itemCopy = createDashboardItem(itemJSON, serializer);
        itemCopy.componentName(undefined);
        this.items.push(itemCopy);
        var newDashboardLayoutItem = new DashboardLayoutItem();
        newDashboardLayoutItem.item = itemCopy;
        newDashboardLayoutItem.weight(dashboardLayoutItem.weight());
        dashboardLayoutItem.insert(newDashboardLayoutItem, 'left');
    }
    _createDashboardLayoutItem(modelItemJson) {
        if (!!modelItemJson) {
            var newItemModel = createDashboardItem(modelItemJson);
            newItemModel.name(NameGenerator.generateName(getItemTitle(newItemModel) + ' ', this.items().concat(this.groups()), 'name', 1));
            if (this.dataSources().length > 0 && newItemModel instanceof DataDashboardItem) {
                newItemModel.dataSource(this.dataSources()[0].componentName());
                if (this.dataSources()[0] instanceof SqlDataSource) {
                    var sqlDataSource = (this.dataSources()[0]);
                    sqlDataSource.queries().length > 0 && newItemModel.dataMember(sqlDataSource.queries()[0].name());
                }
                if (this.dataSources()[0] instanceof EFDataSource) {
                    var efDataSource = (this.dataSources()[0]);
                    if (!efDataSource._tables().length) {
                        var subscription = efDataSource._tables.subscribe((tables) => {
                            let dataDashboardItem = newItemModel;
                            if (dataDashboardItem.dataSource() === efDataSource.componentName() && !dataDashboardItem.dataMember.peek()) {
                                newItemModel.dataMember(tables[0].dataMember());
                            }
                            subscription.dispose();
                        });
                    }
                    else {
                        newItemModel.dataMember(efDataSource._tables()[0].dataMember());
                    }
                }
                if (this.dataSources()[0] instanceof FederationDataSource) {
                    var feredationDataSource = (this.dataSources()[0]);
                    feredationDataSource.queries().length > 0 && newItemModel.dataMember(feredationDataSource.queries()[0].alias());
                }
                if (this.dataSources()[0] instanceof MongoDBDataSource) {
                    let mongoDBDataSource = (this.dataSources()[0]);
                    mongoDBDataSource.queries().length > 0 && newItemModel.dataMember(mongoDBDataSource.queries()[0]._actualName());
                }
            }
            if (newItemModel instanceof GroupItem) {
                this.groups.push(newItemModel);
            }
            else {
                this.items.push(newItemModel);
            }
            return this._createDashboardLayoutNode(newItemModel);
        }
        return new DashboardLayoutGroup();
    }
    _createDashboardLayoutNode(dashboardItem) {
        var itemType = null;
        if (dashboardItem instanceof GroupItem) {
            itemType = 'LayoutGroup';
        }
        else if (dashboardItem instanceof TabContainerItem) {
            itemType = 'LayoutTabContainer';
        }
        else {
            itemType = 'LayoutItem';
        }
        var newLayoutItemModel = deserializeDashboardLayoutNode({ '@ItemType': itemType });
        newLayoutItemModel.item = dashboardItem;
        return newLayoutItemModel;
    }
    _findDataItem(itemId) {
        return this._dataDashboardItems().filter(item => item.componentName() == itemId)[0];
    }
    _interactivityGroupPathToRoot(dashboardItem) {
        let getParentContainerItem = (item) => {
            return (!!item.parentContainer() && this.findItem(item.parentContainer()) || undefined);
        };
        let parentContainers = [];
        let parentContainerItem = dashboardItem;
        do {
            parentContainerItem = getParentContainerItem(parentContainerItem);
            if (parentContainerItem) {
                parentContainers.push(parentContainerItem);
            }
        } while (parentContainerItem);
        return parentContainers;
    }
    _processDeleteDataSource(dataSource) {
        this._dataDashboardItems()
            .filter(item => item.dataSource() == dataSource.componentName())
            .forEach(item => item._clearBindings());
    }
}
Dashboard._dataSourceTypesMap = Object.assign({ 'FederationDataSource': FederationDataSource }, _baseDataSourceTypesMap);
__decorate([
    wrapWithUndoRedo
], Dashboard.prototype, "_changeItem", null);
__decorate([
    wrapWithUndoRedo
], Dashboard.prototype, "_duplicateItem", null);
__decorate([
    wrapWithUndoRedo
], Dashboard.prototype, "_processDeleteDataSource", null);
