// devexpress-dashboard/common/viewer/adapters/_pivot-item-viewer-adapter.js
/**
* DevExpress Dashboard (_pivot-item-viewer-adapter.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { DataItemViewerAdapter } from './_data-item-viewer-adapter';
export class PivotItemViewerAdapter extends DataItemViewerAdapter {
    constructor() {
        super(...arguments);
        this.expandValueHandler = (_, params) => {
            if (params.isRequestData) {
                this.dashboardItem._processItemExpandingChange({
                    values: params.values,
                    isColumn: params.isColumn
                });
            }
        };
        this.expandStateChangedHandler = (dxPivotState) => {
            var currentState = this.item.getExpandingState(!this.dashboardItem.autoExpandRowGroups(), !this.dashboardItem.autoExpandColumnGroups());
            this.dashboardItem._processExpandingStateChanged(currentState, dxPivotState);
        };
    }
    attachToModel(viewerItem, dataDashboardItem) {
        super.attachToModel(viewerItem, dataDashboardItem);
        viewerItem.expandValue.add(this.expandValueHandler);
        viewerItem.expandStateChanged.add(this.expandStateChangedHandler);
    }
    detachFromModel(viewerItem, dataDashboardItem) {
        viewerItem.expandValue.remove(this.expandValueHandler);
        viewerItem.expandStateChanged.remove(this.expandStateChangedHandler);
        super.detachFromModel(viewerItem, dataDashboardItem);
    }
}
