// devexpress-dashboard/common/notification-controller/notificator.js
/**
* DevExpress Dashboard (notificator.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { decodeHtml } from '../../data/_utils';
import { NotificationControllerViewModel } from './_notificator-view-model';
export class NotificationController {
    constructor() {
        this._viewModel = new NotificationControllerViewModel();
    }
    static _getErrorTextFromResponse(request) {
        return (request && request.responseJSON && request.responseJSON.Message) || '';
    }
    static _getDetailedErrorMessage(errorInfo) {
        var errorDetail = '';
        if (typeof errorInfo == 'string') {
            errorDetail = errorInfo;
        }
        else if (errorInfo && errorInfo['responseJSON']) {
            errorDetail = NotificationController._getErrorTextFromResponse(errorInfo);
        }
        return decodeHtml(errorDetail);
    }
    suspended(isSuspended) {
        this._viewModel.suspended(isSuspended);
    }
    showState(message) {
        this._viewModel.updateNotification('info', message);
    }
    showSuccess(message) {
        this._viewModel.updateNotification('success', message);
    }
    showError(title, errorInfo) {
        let errorDetail = NotificationController._getDetailedErrorMessage(errorInfo);
        this._viewModel.updateNotification('error', title, errorDetail);
        console.warn(title + (errorDetail ? ' - ' + errorDetail : ''));
    }
    reset() {
        this._viewModel.reset();
    }
}
