// devexpress-dashboard/model/items/chart/chart-axis.js
/**
* DevExpress Dashboard (chart-axis.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import { SerializableModel } from '../../serializable-model';
import { chartAxisSerializationsInfo, chartAxisXSerializationsInfo, chartAxisYSerializationsInfo, chartSecondaryAxisYSerializationsInfo, scatterChartAxisYSerializationsInfo } from './metadata/_chart-axis';
export class ChartAxis extends SerializableModel {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return chartAxisSerializationsInfo;
    }
}
export class ChartAxisX extends ChartAxis {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return chartAxisXSerializationsInfo;
    }
}
export class ChartAxisY extends ChartAxis {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return chartAxisYSerializationsInfo;
    }
}
export class ScatterChartAxisY extends ChartAxisY {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return scatterChartAxisYSerializationsInfo;
    }
}
export class ChartSecondaryAxisY extends ChartAxisY {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return chartSecondaryAxisYSerializationsInfo;
    }
}
