// devexpress-dashboard/model/colorization/color-scheme-definition.js
/**
* DevExpress Dashboard (color-scheme-definition.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getLocalizationById } from '../../data/localization/_default';
export class ColorSchemeDefinition {
    constructor(dataSource, dataMember, dimensionDefinitions, colorByMeasures = false, componentName = '', name = '') {
        this.dataSource = dataSource;
        this.dataMember = dataMember;
        this.dimensionDefinitions = dimensionDefinitions;
        this.colorByMeasures = colorByMeasures;
        this.componentName = componentName;
        this.name = name;
        this.dataItems = ko.computed(() => {
            return this.dimensionDefinitions.map(definition => {
                return false ? definition.dataMember() + ' (' + definition.dateTimeGroupInterval() + ')' : definition.dataMember();
            }).concat(this.colorByMeasures ? [getLocalizationById('DashboardWebStringId.Colorization.MeasureNames')] : []);
        });
        this.typeText = ko.computed(() => {
            return !this.componentName ? getLocalizationById('DashboardWebStringId.Coloring.Global') : (getLocalizationById('DashboardWebStringId.Coloring.Local') + '(' + this.name + ')');
        });
        this.dataSourceText = ko.computed(() => {
            return this.dataMember || this.dataSource;
        });
        this.key = ko.computed(() => {
            return [this.componentName, this.dataMember, this.dataSource].concat(this.dimensionDefinitions.map(value => value.dataMember() + '(' + value.dateTimeGroupInterval() + ')').sort()).concat(this.colorByMeasures ? [getLocalizationById('DashboardWebStringId.Colorization.MeasureNames')] : []).join('|');
        });
    }
    equals(definition) {
        return definition.key() === this.key();
    }
    _getDataSourceText(dataSourceName) {
        return this.dataMember || dataSourceName || this.dataSource;
    }
}
