// devexpress-dashboard/model/data-sources/_data-field.js
/**
* DevExpress Dashboard (_data-field.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as $ from 'jquery';
import * as ko from 'knockout';
import { SerializableModel } from '../serializable-model';
import { dataFieldSerializationsInfo } from './metadata/_data-field';
export let IsNumeric = (dataType) => (['Decimal', 'Float', 'Double', 'Long', 'Integer'].indexOf(dataType) > -1);
export let IsTextual = (dataType) => (['Enum', 'Text'].indexOf(dataType) > -1);
export let IsDateTime = (dataType) => (['DateTime'].indexOf(dataType) > -1);
export let IsOlapHierarchyField = dataField => !!dataField.groupIndex && dataField.groupIndex() !== undefined;
export class DataField extends SerializableModel {
    constructor(dataFieldJSON = {}, serializer = new ModelSerializer()) {
        super(dataFieldJSON, serializer);
        this.expanded = true;
        this.childNodes = deserializeArray(dataFieldJSON.ChildNodes, (item) => new DataField(item, serializer));
        if (!this.isDataFieldNode() || this.childNodes().length > 0) {
            this.fieldType(undefined);
        }
        this.groupDataItems = $.map(dataFieldJSON.GroupFieldTypes || {}, (type, name) => {
            return {
                dataMember: ko.observable(name),
                fieldType: ko.observable(type),
                displayName: ko.observable(dataFieldJSON.GroupCaptions[dataFieldJSON.GroupDataMembers.indexOf(name)]),
                isConvertible: ko.observable(true),
                isDataFieldNode: ko.observable(true)
            };
        });
        this.defaultNumericFormat = dataFieldJSON.DefaultNumericFormat;
    }
    static isNumeric(dataField) {
        return dataField && (IsNumeric(dataField.fieldType()) || (dataField.fieldType() == 'Custom' && DataField.isOlap(dataField.dataMember())));
    }
    static isDateTime(dataField) {
        return dataField && IsDateTime(dataField.fieldType());
    }
    static olapMarker() {
        return '].[';
    }
    static isOlap(dataMember) {
        return dataMember && ((dataMember.indexOf(DataField.olapMarker()) !== -1)
            || ((dataMember[0] === '[') && (dataMember.indexOf(']') === dataMember.length - 1)));
    }
    static isMeasure(dataField) {
        if (dataField.isAggregate && dataField.isAggregate())
            return true;
        var isOlap = DataField.isOlap(dataField.dataMember());
        return (!isOlap && DataField.isNumeric(dataField)) || (isOlap && DataField.isOrContainMeasures(dataField.dataMember()));
    }
    static isOrContainMeasures(dataMember) {
        return dataMember === '[Measures]' || dataMember.indexOf('[Measures]') !== -1;
    }
    static ifOlapThenOnlyMeasure(dataField) {
        return !DataField.isOlap(dataField.dataMember()) || DataField.isOrContainMeasures(dataField.dataMember());
    }
    static ifOlapThenOnlyDimension(dataField) {
        return !DataField.isOlap(dataField.dataMember()) || !DataField.isOrContainMeasures(dataField.dataMember());
    }
    static isContinous(dataField) {
        return dataField && (DataField.isNumeric(dataField) || dataField.fieldType() === 'DateTime');
    }
    static isOlapHierarchy(dataField) {
        return dataField && !!dataField.nodeType && dataField.nodeType() === 'OlapHierarchy';
    }
    getInfo() {
        return dataFieldSerializationsInfo;
    }
}
