// devexpress-dashboard/designer/data-source-browser/item-providers/_field-list-item-provider.js
/**
* DevExpress Dashboard (_field-list-item-provider.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { findDataMember, isNonCollectionDataField } from '../../../common/_data-source-browser';
import { createJQueryDeferred } from '../../../data/_jquery-helpers';
import { FederationDataSource } from '../../../model';
import { DataItem } from '../../../model/data-item/data-item';
import { OlapDataSource } from '../../../model/data-sources/olap-data-source';
import { SqlDataSource } from '../../../model/data-sources/sql-data-source';
export class DataFieldViewModel {
    constructor(name, displayName, hasItems, specifics, field, isList, isCorruptedCalcField) {
        this.name = name;
        this.displayName = displayName;
        this.hasItems = hasItems;
        this.specifics = specifics;
        this.field = field;
        this.isList = isList;
        this.isCorruptedCalcField = isCorruptedCalcField;
        this.innerActions = ko.observableArray();
    }
}
export class FieldListItemProvider {
    constructor(_dataSourceBrowserViewModel, _getDataFieldArrayCallback, isFieldValid) {
        this._dataSourceBrowserViewModel = _dataSourceBrowserViewModel;
        this._getDataFieldArrayCallback = _getDataFieldArrayCallback;
        this.isFieldValid = isFieldValid;
        this.loading = ko.observable(false);
        this._changeTrigger = ko.observable(false);
    }
    triggerItemsChanged() {
        this._changeTrigger.valueHasMutated();
    }
    getItems(pathRequest) {
        var deferred = createJQueryDeferred();
        var dataSource = this._dataSourceBrowserViewModel.selectedDataSource();
        this._changeTrigger();
        if (!!dataSource) {
            var { dataMember, fieldPath } = findDataMember(dataSource, pathRequest.path);
            this.loading(true);
            this._getDataFieldArrayCallback(dataSource.componentName(), dataMember, fieldPath).done(dataFields => {
                deferred.resolve(dataFields
                    .filter(field => !!field.dataMember())
                    .filter(field => !this.isFieldValid || this.isFieldValid(field))
                    .map(dataNode => {
                    var name = dataSource instanceof OlapDataSource ? dataNode.dataMember() : dataNode.name();
                    var dataFieldViewModel = new DataFieldViewModel(name, ko.unwrap(dataNode.displayName), !dataNode.isDataFieldNode(), DataItem.typesMap[dataNode.fieldType()] || 'string', dataNode, !isNonCollectionDataField(dataNode), dataNode.isCorruptedCalcField && dataNode.isCorruptedCalcField());
                    if (dataSource instanceof SqlDataSource) {
                        let isSqlQueryNode = pathRequest.path.length === 0;
                        let query = dataSource.queries().filter(query => query.name() === dataNode.dataMember())[0];
                        if (this._dataSourceBrowserViewModel.canEditDataSource && query != null && isSqlQueryNode) {
                            if (this._dataSourceBrowserViewModel.canEditCustomSqlQueries || query.type() !== 'CustomSqlQuery') {
                                dataFieldViewModel.innerActions.push({
                                    click: () => {
                                        this._dataSourceBrowserViewModel.editQuery(dataNode.dataMember());
                                    },
                                    icon: 'dx-dashboard-ds-edit',
                                    style: 'dx-dashboard-datasource-field-icon-edit'
                                });
                            }
                            dataFieldViewModel.innerActions.push({
                                click: () => {
                                    this._dataSourceBrowserViewModel.removeQuery(dataNode.dataMember());
                                },
                                icon: 'dx-dashboard-remove-small',
                                style: 'dx-dashboard-datasource-field-icon-remove'
                            });
                        }
                    }
                    if (dataSource instanceof FederationDataSource) {
                        let isQueryNode = pathRequest.path.length === 0;
                        let query = dataSource.queries().find(query => query.alias() === dataNode.dataMember());
                        if (this._dataSourceBrowserViewModel.canEditDataSource && query && isQueryNode) {
                            dataFieldViewModel.innerActions.push({
                                click: () => {
                                    this._dataSourceBrowserViewModel.editFederationQuery(dataNode.dataMember());
                                },
                                icon: 'dx-dashboard-ds-edit',
                                style: 'dx-dashboard-datasource-field-icon-edit'
                            });
                            dataFieldViewModel.innerActions.push({
                                click: () => {
                                    this._dataSourceBrowserViewModel.removeFederationQuery(dataNode.dataMember());
                                },
                                icon: 'dx-dashboard-remove-small',
                                style: 'dx-dashboard-datasource-field-icon-remove'
                            });
                        }
                    }
                    if (dataNode.nodeType() === 'CalculatedDataField') {
                        dataFieldViewModel.style = 'dx-dashboard-calculated-field';
                        var calcField = dataSource.calculatedFields().filter(calculatedField => calculatedField.name() === dataNode.dataMember())[0];
                        if (calcField) {
                            dataFieldViewModel.innerActions.push({
                                click: () => {
                                    this._dataSourceBrowserViewModel.editCalcField(calcField);
                                },
                                icon: 'dx-dashboard-ds-edit',
                                style: 'dx-dashboard-datasource-field-icon-edit'
                            });
                            dataFieldViewModel.innerActions.push({
                                click: () => {
                                    this._dataSourceBrowserViewModel.removeCalcField(calcField);
                                },
                                icon: 'dx-dashboard-remove-small',
                                style: 'dx-dashboard-datasource-field-icon-remove'
                            });
                        }
                    }
                    this.customizeDataFieldViewModel && this.customizeDataFieldViewModel(dataFieldViewModel);
                    return dataFieldViewModel;
                }));
            });
        }
        else {
            deferred.resolve([]);
        }
        deferred.always(() => {
            this.loading(false);
        });
        return deferred.promise();
    }
}
