// devexpress-dashboard/model/internal/_client-filter-manager.js
/**
* DevExpress Dashboard (_client-filter-manager.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable } from '@devexpress/analytics-core/analytics-utils';
export class ClientFilterManager extends Disposable {
    constructor(_updateTotals) {
        super();
        this._updateTotals = _updateTotals;
        this.addDisposable(_updateTotals.subscribe(value => {
            if (!value)
                this.clearFilterParams();
        }));
    }
    getFilterParams() {
        return this._filterParams;
    }
    setFilterParams(clientFilter) {
        if (this._updateTotals()) {
            this._filterParams = clientFilter;
        }
    }
    clearFilterParams() {
        this._filterParams = undefined;
    }
    setFilterState(filterValue) {
        this._filterState = filterValue;
    }
    getFilterState() {
        return this._filterState;
    }
}
