// devexpress-dashboard/model/internal/_expanding-manager.js
/**
* DevExpress Dashboard (_expanding-manager.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { toStringArray } from './_date-utils';
export class ExpandingManager {
    constructor() {
        this._expandingParams = undefined;
        this._expandingState = undefined;
    }
    getPivotExpandViewState() {
        return this._dxPivotState;
    }
    setExpandingParams(expandingParams) {
        this._expandingParams = expandingParams;
    }
    onViewStateChanged(expandingState, dxPivotState) {
        this._expandingState = expandingState;
        this._dxPivotState = dxPivotState;
    }
    canProvideExpandingState() {
        return this._expandingParams || this._expandingState;
    }
    calculateExpandingState() {
        var params = this._expandingParams;
        var state = this._expandingState;
        return {
            rows: state && toStringArray(state.rows) || [],
            columns: state && toStringArray(state.columns) || [],
            values: !!params ? toStringArray(params.values) : undefined,
            isColumn: !!params && params.isColumn
        };
    }
    resetExpandingParams() {
        this._expandingParams = undefined;
    }
    resetColumnViewState(autoExpand) {
        this._expandingState = { rows: this._expandingState.rows, columns: [] };
        if (this._dxPivotState) {
            this._dxPivotState.fields && this._dxPivotState.fields.filter(f => f.area === 'column').forEach(f => { f.expanded = autoExpand; });
            this._dxPivotState.columnExpandedPaths = [];
        }
    }
    resetRowViewState(autoExpand) {
        this._expandingState = { rows: [], columns: this._expandingState.columns };
        if (this._dxPivotState) {
            this._dxPivotState.fields && this._dxPivotState.fields.filter(f => f.area === 'row').forEach(f => { f.expanded = autoExpand; });
            this._dxPivotState.rowExpandedPaths = [];
        }
    }
}
