// devexpress-dashboard/model/internal/_knockout-utils.js
/**
* DevExpress Dashboard (_knockout-utils.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import CustomStore from 'devextreme/data/custom_store';
import DataSource from 'devextreme/data/data_source';
import * as ko from 'knockout';
export function subscribeArrayChange(array, handlers) {
    return array.subscribe(changes => {
        changes.forEach(arrayChange => {
            var change = arrayChange;
            if (change.status === 'added' && handlers.added) {
                handlers.added(change.value, change.index);
            }
            if (change.status === 'deleted' && handlers.deleted) {
                handlers.deleted(change.value);
            }
        });
    }, null, 'arrayChange');
}
export function subscribeWithPrev(target, callback) {
    var prevValue;
    var disposables = [];
    disposables.push(target.subscribe(value => {
        prevValue = value;
    }, null, 'beforeChange'));
    disposables.push(target.subscribe(value => {
        callback(prevValue, value);
        prevValue = undefined;
    }));
    return {
        dispose: () => {
            disposables.forEach(disposable => disposable.dispose());
            disposables = [];
        }
    };
}
export function subscribeArrayChangeWithCancel(target, callback) {
    let prevValue;
    let disposables = [];
    disposables.push(target.subscribe(value => {
        prevValue = value && value.slice();
    }, null, 'beforeChange'));
    disposables.push(target.subscribe(changes => {
        let ok = true;
        try {
            ok = callback(changes);
        }
        catch (error) {
            target(prevValue);
            prevValue = undefined;
            throw error;
        }
        if (!ok)
            target(prevValue);
        prevValue = undefined;
    }, null, 'arrayChange'));
    return {
        dispose: () => { disposables.forEach(element => element.dispose()); }
    };
}
export function syncArrayHelper(sourceArray, destArray, addHandler) {
    destArray(sourceArray.peek().map(item => addHandler(item)));
    return sourceArray.subscribe((changes) => {
        changes.forEach((arrayChange) => {
            var change = arrayChange;
            if (change.status === 'added') {
                destArray.splice(change.index, 0, addHandler(change.value));
            }
            else if (change.status === 'deleted') {
                destArray.splice(change.index, 1);
            }
        });
    }, null, 'arrayChange');
}
export function subscribeToArrayItemProperties(array, handler) {
    let subscriptionInfoList = [];
    let subscribe = (item) => {
        subscriptionInfoList.push({ object: item, subscriptions: [].concat(handler(item) || []) });
    };
    let unsubscribe = (item) => {
        var info = subscriptionInfoList.filter(info => info.object === item)[0];
        if (info) {
            info.subscriptions.forEach(subscription => subscription.dispose());
            subscriptionInfoList.splice(subscriptionInfoList.indexOf(info), 1);
        }
    };
    array().forEach(subscribe);
    let arrayChangeSubscribtion = subscribeArrayChange(array, {
        added: subscribe,
        deleted: unsubscribe
    });
    return {
        dispose: () => {
            arrayChangeSubscribtion.dispose();
            arrayChangeSubscribtion = null;
            subscriptionInfoList.forEach(info => info.subscriptions.forEach(subscription => subscription.dispose()));
            subscriptionInfoList.splice(0, subscriptionInfoList.length);
        }
    };
}
export function subscribeAndPerform(subscribable, action) {
    action(subscribable());
    return subscribable.subscribe(action);
}
export function safeSubscribe(observables, handler) {
    let computed = ko.pureComputed(() => Object.keys(observables).reduce((acc, key) => { acc[key] = observables[key](); return acc; }, {}));
    let prevDisposable = handler(computed());
    let subscription = computed.subscribe(args => {
        prevDisposable && prevDisposable.dispose();
        prevDisposable = handler(args);
    });
    return {
        dispose: () => {
            subscription.dispose();
            computed.dispose();
            prevDisposable && prevDisposable.dispose();
        }
    };
}
export function safeComputed(observables, handler) {
    var observable = ko.observable();
    var subscription = safeSubscribe(observables, (args) => {
        observable(handler(args));
    });
    var computed = ko.pureComputed(() => observable());
    var oldDispose = computed.dispose;
    computed.dispose = () => {
        subscription.dispose();
        oldDispose.call(computed);
    };
    return computed;
}
export function createObservableDataSource(observables, load) {
    let computed = ko.pureComputed(() => Object.keys(observables).reduce((acc, key) => { acc[key] = observables[key](); return acc; }, {}));
    let dataSource = new DataSource({
        store: new CustomStore({
            loadMode: 'raw',
            load: () => load(computed())
        })
    });
    var subscription = computed.subscribe(_ => {
        dataSource.reload();
    });
    return {
        dataSource: dataSource,
        dispose: () => {
            subscription.dispose();
            if (dataSource['_disposed'] !== true)
                dataSource.dispose();
            computed.dispose();
        }
    };
}
