// devexpress-dashboard/model/custom-properties/_custom-properties-utils.js
/**
* DevExpress Dashboard (_custom-properties-utils.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { floatFromModel, parseBool, PropertyCategory } from '../metadata/_base-metadata';
import { CustomProperties } from './custom-properties';
import { _customPropertiesMeta } from './custom-properties-metadata';
function _convertCustomPropertiesMetadata(customPropertiesMeta) {
    return (customPropertiesMeta || [])
        .map(d => {
        var serializationInfo = {
            propertyName: d.propertyName,
            modelName: d.propertyName,
            category: PropertyCategory.ViewModel,
            defaultVal: d.defaultValue,
        };
        if (d.valueType === 'boolean') {
            serializationInfo.from = parseBool;
        }
        if (d.valueType === 'number') {
            serializationInfo.from = floatFromModel;
        }
        return serializationInfo;
    });
}
export function _getCustomPropertiesSerializationInfoCore(customPropertiesMeta) {
    return {
        propertyName: 'customProperties',
        modelName: 'CustomProperties',
        category: PropertyCategory.ViewModel,
        type: CustomProperties,
        info: _convertCustomPropertiesMetadata(customPropertiesMeta)
    };
}
export function getCustomPropertiesSerializationInfo(owner) {
    const filteredCustomProperties = _customPropertiesMeta
        .filter(metadata => owner instanceof metadata.ownerType);
    return _getCustomPropertiesSerializationInfoCore(filteredCustomProperties);
}
