// devexpress-dashboard/viewer-parts/widgets/dialogs/_custom-time-period-dialog.js
/**
* DevExpress Dashboard (_custom-time-period-dialog.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import 'devextreme/ui/date_box';
import dxDateBox from 'devextreme/ui/date_box';
import { DashboardFormatHelper } from '../../../data/_format-helper';
import { localizationId } from '../../../data/_localization-ids';
import { localizer } from '../../../data/_localizer';
import { DashboardLayoutModeHelper } from '../../_dashboard-layout-mode-helper';
import { CalendarHelper } from '../../_utils';
import { dialogClasses, dialogForm } from './_dialog-form';
var customTimePeriodDialogClassNames = {
    emptyButton: 'dx-dashboard-custom-time-period-empty-button',
    leftPeriod: 'dx-dashboard-custom-time-period-dialog-left-period',
    rightPeriod: 'dx-dashboard-custom-time-period-dialog-right-period'
};
export class customTimePeriodDialog {
    constructor(options) {
        this.options = options;
        this.setRange = options.setRange;
        this._initialize();
    }
    _initialize() {
        this.dialogForm = new dialogForm({
            dialogContainer: this.options.container,
            width: 'auto',
            height: 'auto',
            allowScrolling: true,
            deferredRendering: true,
            title: localizer.getString(localizationId.labelName.SelectRange),
            buttons: [{
                    className: customTimePeriodDialogClassNames.emptyButton
                }, {
                    name: localizer.getString(localizationId.buttonNames.ButtonOK),
                    hide: true,
                    func: () => {
                        this.setRange({ startValue: this.leftCalendar.option('value'), endValue: this.rightCalendar.option('value') });
                    },
                    isDefault: true
                }, {
                    name: localizer.getString(localizationId.buttonNames.ButtonCancel),
                    hide: true,
                    func: () => { }
                }
            ],
            renderContent: () => {
                let form = document.createElement('div');
                form.classList.add(dialogClasses.form);
                let fromText = localizer.getString(localizationId.labelName.RangeFilterSelectRangeFromCaption);
                let toText = localizer.getString(localizationId.labelName.RangeFilterSelectRangeToCaption);
                let leftDate = document.createElement('div');
                leftDate.style.width = '100%';
                let rightDate = document.createElement('div');
                rightDate.style.width = '100%';
                let left = document.createElement('div');
                left.classList.add(customTimePeriodDialogClassNames.leftPeriod);
                let right = document.createElement('div');
                right.classList.add(customTimePeriodDialogClassNames.rightPeriod);
                let options = {
                    pickerType: DashboardLayoutModeHelper.isTouch ? 'rollers' : 'calendar',
                    type: this.format,
                    calendarOptions: {
                        maxZoomLevel: this.maxZoomLevel,
                    },
                    displayFormat: this.displayFormatFunc,
                    acceptCustomValue: false,
                    applyButtonText: localizer.getString(localizationId.buttonNames.ButtonOK),
                    cancelButtonText: localizer.getString(localizationId.buttonNames.ButtonCancel),
                    placeholder: localizer.getString(localizationId.buttonNames.SelectDate),
                    dropDownOptions: {
                        container: this.options.container
                    },
                    disabledDates: this.disabledDates,
                };
                let fromTextSpan = document.createElement('span');
                fromTextSpan.innerText = fromText;
                left.appendChild(fromTextSpan);
                left.appendChild(leftDate);
                this.leftCalendar = new dxDateBox(leftDate, Object.assign(Object.assign({}, options), { value: this.range.minimum || new Date() }));
                let toTextSpan = document.createElement('span');
                toTextSpan.innerText = toText;
                right.appendChild(toTextSpan);
                right.appendChild(rightDate);
                this.rightCalendar = new dxDateBox(rightDate, Object.assign(Object.assign({}, options), { value: this.range.maximum || new Date() }));
                form.appendChild(left);
                form.appendChild(document.createElement('br'));
                form.appendChild(right);
                return form;
            },
            disposeContent: () => {
                this.leftCalendar && this.leftCalendar.dispose();
                this.rightCalendar && this.rightCalendar.dispose();
            },
            setActualState: (width) => {
            }
        });
    }
    show(options) {
        this.range = options.range;
        if (options.isIntYearGroupInterval && this.range.minimum && this.range.maximum) {
            this.range.minimum = new Date(this.range.minimum, 0, 1);
            this.range.maximum = new Date(this.range.maximum, 0, 1);
        }
        this.format = CalendarHelper.getCalendarType(options.groupInterval);
        this.maxZoomLevel = CalendarHelper.getCalendarMaxZoomLevel(options.groupInterval);
        this.displayFormatFunc = value => {
            if (options.isIntYearGroupInterval)
                value = value.getFullYear();
            return DashboardFormatHelper.format(value, options.displayFormat);
        };
        this.disabledDates = options.groupInterval === 'WeekYear' ? (args) => args.date.getDay() !== options.firstDayOfWeek : undefined;
        this.dialogForm.showDialog();
    }
    dispose() {
        this.dialogForm && this.dialogForm.dispose();
    }
}
