// devexpress-dashboard/designer/items/properties-composers/_custom-range-properties-composer.js
/**
* DevExpress Dashboard (_custom-range-properties-composer.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { localize } from '@devexpress/analytics-core/analytics-internal';
import * as ko from 'knockout';
import { getLocalizationById } from '../../../data/localization/_default';
import { endLimit, mode, startLimit } from '../../../model/items/range-filter/metadata/_date-time-period';
import { fixedDateTimePeriodLimitProperty, flowDateTimePeriodLimitProperty } from '../../../model/items/range-filter/metadata/_limit-container';
import { date } from '../../../model/items/range-filter/metadata/_period-limit';
import { name } from '../../../model/metadata/_base-metadata';
import { actionButtons, flowModeSettingsEditor, textPreviewEditor } from '../../form-adapter/_form-adapter-editors';
import { ObjectPropertiesWrapper } from '../../form-adapter/_object-properties-wrapper';
import { AccordionTab, KnownTabs } from '../../properties-controller/_accordion-tab';
export class CustomRangePropertiesComposer {
    composeTabs(model, args) {
        var commonTab = new AccordionTab(KnownTabs.Common, 'DashboardWebStringId.AccordionTab.Common', this.getCommonWrapper(model, args.argument, args.rangeFilterItem));
        return [commonTab];
    }
    getCommonWrapper(model, argument, rangeFilterItem) {
        var p = [], visibilityRules = {}, dateOnly = ['Year', 'MonthYear', 'QuarterYear', 'DayMonthYear'].indexOf(argument.dateTimeGroupInterval()) !== -1;
        p.push(Object.assign(Object.assign({}, name), { validateBeforeSet: true, validationRules: [
                {
                    type: 'custom',
                    validationCallback: (options) => { return !rangeFilterItem.dateTimePeriods().filter(period => period.name() === options.value)[0]; },
                    message: getLocalizationById('DashboardStringId.UniqueNameValidationMessage')
                }
            ] }));
        p.push({
            container: startLimit,
            properties: [Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.StartMode', replacementPropertyName: 'start_mode' }, mode), Object.assign(Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.StartIntervalAndOffset', replacementPropertyName: 'start_flow_options' }, flowDateTimePeriodLimitProperty), { formAdapterItem: flowModeSettingsEditor() }), {
                    container: fixedDateTimePeriodLimitProperty,
                    properties: [Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.StartDate', replacementPropertyName: 'start_date', editorOptions: {
                                type: dateOnly ? 'date' : 'datetime'
                            } }, date)]
                }]
        });
        visibilityRules['start_date'] = ['start_mode', '=', 'Fixed'];
        visibilityRules['start_flow_options'] = ['start_mode', '=', 'Flow'];
        p.push({
            container: endLimit,
            properties: [Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.EndMode', replacementPropertyName: 'end_mode' }, mode), Object.assign(Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.EndIntervalAndOffset', replacementPropertyName: 'end_flow_options' }, flowDateTimePeriodLimitProperty), { formAdapterItem: flowModeSettingsEditor() }), {
                    container: fixedDateTimePeriodLimitProperty,
                    properties: [Object.assign({ displayName: 'DashboardWebStringId.RangeFilter.EndDate', replacementPropertyName: 'end_date', editorOptions: {
                                type: dateOnly ? 'date' : 'datetime'
                            } }, date)]
                }]
        });
        visibilityRules['end_date'] = ['end_mode', '=', 'Fixed'];
        visibilityRules['end_flow_options'] = ['end_mode', '=', 'Flow'];
        p.push({
            propertyName: 'dateSampleFakeProperty',
            formAdapterItem: textPreviewEditor({ target: model._getPeriodTextValue }),
        });
        p.push({
            propertyName: 'buttonsFakeProperty',
            formAdapterItem: actionButtons([
                {
                    text: localize('DashboardWebStringId.RangeFilter.SetDefaultCustomPeriod'),
                    onClick: () => rangeFilterItem.defaultDateTimePeriodName(model.name()),
                    visible: ko.computed(() => rangeFilterItem.defaultDateTimePeriodName() !== model.name())
                },
                {
                    text: localize('DashboardWebStringId.RangeFilter.ClearDefaultCustomPeriod'),
                    onClick: () => rangeFilterItem.defaultDateTimePeriodName(null),
                    visible: ko.computed(() => rangeFilterItem.defaultDateTimePeriodName() === model.name())
                }
            ]),
        });
        var wrapper = new ObjectPropertiesWrapper({
            model: model,
            properties: p,
            visibilityFilterRules: visibilityRules,
            modelExtention: { buttonsFakeProperty: {}, dateSampleFakeProperty: {} }
        });
        return wrapper;
    }
}
