// devexpress-dashboard/designer/items/properties-composers/_tab-container-item-properties-composer.js
/**
* DevExpress Dashboard (_tab-container-item-properties-composer.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { safeSubscribe, subscribeToArrayItemProperties } from '../../../model/internal/_knockout-utils';
import { layoutItemsSerializable } from '../../../model/layout/metadata/_dashboard-layout-node';
import { collectionEditor } from '../../form-adapter/_form-adapter-editors';
import { ObjectPropertiesWrapper } from '../../form-adapter/_object-properties-wrapper';
import { AccordionTab, KnownTabs } from '../../properties-controller/_accordion-tab';
import { CollectionEditorRefreshCallback } from '../../ui-widgets/collection-editor/_collection-editor-viewmodel-base';
import { DashboardItemPropertiesComposer } from './_base-properties-composer';
export class TabContainerItemPropertiesComposer extends DashboardItemPropertiesComposer {
    constructor(editTabPageHandler) {
        super(() => { });
        this.editTabPageHandler = editTabPageHandler;
    }
    _composeTabsCore() { }
    composeTabs(tabContainer, args) {
        let tabContainerLayoutItem = args.dashboard.layout().findLayoutItem(tabContainer);
        if (!tabContainerLayoutItem) {
            return [];
        }
        const customItemTemplate = (itemData, itemIndex, itemElement) => {
            const item = ko.unwrap(itemData.item);
            const div = document.createElement('div');
            div.innerText = item && args.dashboard._getDisplayDashboardItem(item).name() || '';
            div.style.fontWeight = tabContainer._activeTabPage() === item ? '800' : 'normal';
            itemElement.appendChild(div);
            return div;
        };
        const refreshCallback = new CollectionEditorRefreshCallback();
        const subscriptions = [];
        subscriptions.push(safeSubscribe({ activeTabPage: tabContainer._activeTabPage }, _ => refreshCallback.refresh()));
        subscriptions.push(subscribeToArrayItemProperties(tabContainerLayoutItem.childNodes, (node) => {
            const tabPage = node.item;
            return tabPage && [tabPage.showItemAsTabPage, tabPage.name].map(s => s.subscribe(() => refreshCallback.refresh()));
        }));
        const collectionEditorOptions = {
            createNewItemHandler: () => tabContainerLayoutItem._createTabPage(),
            editItemHandler: (layoutItem, args) => this.editTabPageHandler(layoutItem.item, args),
            removeItemHandler: (layoutItem) => layoutItem._createViewModel().delete(),
            enableRemoveItem: () => tabContainerLayoutItem.childNodes().length > 1,
            reorderItemsHandler: (layoutItem, direction) => {
                let prevActiveTab = tabContainer._activeTabPage();
                var index = tabContainerLayoutItem.childNodes().indexOf(layoutItem);
                tabContainerLayoutItem.childNodes.splice(index, 1);
                tabContainerLayoutItem.childNodes.splice(index + (direction === 'up' ? -1 : 1), 0, layoutItem);
                if (prevActiveTab !== tabContainer._activeTabPage()) {
                    tabContainer._activeTabPage(prevActiveTab);
                }
            },
            customTemplate: customItemTemplate,
            forceRefreshCallback: refreshCallback,
        };
        return [
            new AccordionTab(KnownTabs.TabContainer, 'DashboardStringId.DefaultNameTabContainerItem', new ObjectPropertiesWrapper({
                model: tabContainerLayoutItem,
                properties: [
                    Object.assign(Object.assign({}, layoutItemsSerializable), { displayName: 'DashboardWebStringId.TabPages', formAdapterItem: collectionEditor(collectionEditorOptions) })
                ],
                disposableModelSubscriptions: subscriptions,
            }))
        ];
    }
}
