// devexpress-dashboard/model/items/card/card-row-element.js
/**
* DevExpress Dashboard (card-row-element.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { TypedSerializableModel } from '../../serializable-model';
import { cardRowDataElementSerializationInfo, cardRowElementSerializationsInfo, cardRowIndicatorElementSerializationInfo, cardRowTextElementBaseSerializationInfo, cardRowTextElementSerializationInfo } from './metadata/_card-row-element';
export class CardRowElement extends TypedSerializableModel {
    constructor(modelJson, serializer) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardRowElementSerializationsInfo;
    }
}
export class CardRowTextElementBase extends CardRowElement {
    constructor(modelJson, serializer) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardRowTextElementBaseSerializationInfo;
    }
}
export class CardRowDataElement extends CardRowTextElementBase {
    constructor(modelJson, serializer) {
        super(modelJson, serializer);
        this.title = ko.computed(() => this.valueType());
    }
    getInfo() {
        return cardRowDataElementSerializationInfo;
    }
    _getDefaultItemType() {
        return 'CardRowDataElement';
    }
}
export class CardRowTextElement extends CardRowTextElementBase {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardRowTextElementSerializationInfo;
    }
    _getDefaultItemType() {
        return 'CardRowTextElement';
    }
}
export class CardRowIndicatorElement extends CardRowElement {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardRowIndicatorElementSerializationInfo;
    }
    _getDefaultItemType() {
        return 'CardRowIndicatorElement';
    }
}
