// devexpress-dashboard/model/items/card/card-row.js
/**
* DevExpress Dashboard (card-row.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer, deserializeArray } from '@devexpress/analytics-core/analytics-utils';
import { TypedSerializableModel } from '../../serializable-model';
import { CardRowDataElement, CardRowIndicatorElement, CardRowTextElement } from './card-row-element';
import { cardRowBaseSerializationInfo, cardRowSerializationInfo, cardSparklineRowSerializationInfo } from './metadata/_card-row';
export class CardRowBase extends TypedSerializableModel {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardRowBaseSerializationInfo;
    }
}
export class CardRow extends CardRowBase {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
        this.elements = deserializeArray(modelJson.CardRowElements, (item) => this.createRowElement(item, serializer));
    }
    createRowElement(elementJSON, serializer) {
        var itemType = elementJSON['@ItemType'];
        return new CardRow.elementItemTypes[itemType].constructor(elementJSON, serializer);
    }
    getInfo() {
        return cardRowSerializationInfo;
    }
    _getDefaultItemType() {
        return 'CardRow';
    }
}
CardRow.elementItemTypes = {
    'CardRowDataElement': {
        constructor: CardRowDataElement
    },
    'CardRowIndicatorElement': {
        constructor: CardRowIndicatorElement
    },
    'CardRowTextElement': {
        constructor: CardRowTextElement
    }
};
export class CardSparklineRow extends CardRowBase {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return cardSparklineRowSerializationInfo;
    }
    _getDefaultItemType() {
        return 'CardSparklineRow';
    }
}
