// devexpress-dashboard/model/items/card/card-layout.js
/**
* DevExpress Dashboard (card-layout.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer, deserializeArray } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { SerializableModel } from '../../serializable-model';
import { CardRow, CardSparklineRow } from './card-row';
import { cardLayoutSerializationInfo } from './metadata/_card-layout';
export class CardLayout extends SerializableModel {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
        this.templateID = ko.observable(-1);
        this.rows = deserializeArray(modelJson.CardRows, (item) => { return this.createRow(item, serializer); });
    }
    createRow(elementJSON, serializer) {
        var itemType = elementJSON['@ItemType'];
        return new CardLayout.rowTypes[itemType].constructor(elementJSON, serializer);
    }
    getInfo() {
        return cardLayoutSerializationInfo;
    }
    grabFrom(newLayout) {
        this.minWidth = newLayout.minWidth;
        this.maxWidth = newLayout.maxWidth;
        this.rows(newLayout.rows());
    }
}
CardLayout.rowTypes = {
    'CardRow': {
        constructor: CardRow
    },
    'CardSparklineRow': {
        constructor: CardSparklineRow
    }
};
