// devexpress-dashboard/model/metadata/_title.js
/**
* DevExpress Dashboard (_title.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { parseBool } from './_base-metadata';
export let titleText = { propertyName: 'text', modelName: '@Text', displayName: 'DashboardWebStringId.Title.Text', simpleFormAdapterItem: 'textBoxEditor' };
export let titleVisible = { propertyName: 'visible', displayName: 'DashboardWebStringId.Title.Visible', modelName: '@Visible', defaultVal: true, simpleFormAdapterItem: 'checkBoxEditor', from: parseBool };
export let includeMasterFilter = { propertyName: 'includeMasterFilter', displayName: 'DashboardWebStringId.Title.IncludeMasterFilter', modelName: '@IncludeMasterFilterState', defaultVal: true, simpleFormAdapterItem: 'checkBoxEditor', from: parseBool };
export let titleAlignment = {
    propertyName: 'alignment', modelName: '@Alignment', defaultVal: 'Center', displayName: 'DashboardWebStringId.Title.Aligment', simpleFormAdapterItem: 'listEditor', values: {
        'Left': 'DashboardWebStringId.Title.Aligment.Left',
        'Center': 'DashboardWebStringId.Title.Aligment.Center',
    }
};
export let titleImage64 = { propertyName: 'image64', modelName: 'ImageData' };
export let titleImageUrl = { propertyName: 'url', modelName: '@Url', simpleFormAdapterItem: 'textBoxEditor' };
export let titleImageType = {
    propertyName: 'imageType', displayName: 'DashboardWebStringId.Title.Image', defaultVal: 'none', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'embedded': 'DashboardWebStringId.Title.ImageEmbedded',
        'linked': 'DashboardWebStringId.Title.ImageLinked',
        'none': 'DashboardWebStringId.Title.ImageNone'
    }
};
export let dashboardTitleSerializationsInfo = [titleText, titleVisible, titleAlignment, includeMasterFilter, titleImageType, titleImage64, titleImageUrl];
