// devexpress-dashboard/designer/toolbox-extension/toolbox-extension.js
/**
* DevExpress Dashboard (toolbox-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { CancellationToken } from '../../common/_helpers';
import { designerExtensions, extensionNameMap } from '../../common/control-options';
import { DockingLayoutController } from '../../common/docking-layout/_docking-layout-controller';
import { createJQueryDeferred } from '../../data/_jquery-helpers';
import { DashboardLayoutNode } from '../../model';
import { DisposableObject } from '../../model/disposable-object';
import { safeComputed, subscribeAndPerform } from '../../model/index.internal';
import { getIconName, getItemJson } from '../../model/internal/_dashboard-item_helper';
import { defineObsoleteMethod, defineObsoleteProperty } from '../../model/internal/_obsolete-helper';
import { itemTypesMap } from '../../model/serializable-model';
import { toolboxConstants } from './_toolbox-extension';
import { ToolboxViewModel } from './_toolbox-view-model';
import { DashboardToolbarGroup, DashboardToolbarItem, DashboardToolboxGroup, DashboardToolboxItem } from './toolbox-items';
const name = 'toolbox';
const nameAlias = 'toolbox';
export class ToolboxExtension extends DisposableObject {
    constructor(dashboardControl) {
        super();
        this.dashboardControl = dashboardControl;
        this.name = name;
        this.menuItems = ko.observableArray();
        this.addMenuItem = (menuItem) => {
            if (!!this._findMenuItem(menuItem.id)) {
                throw Error("The '" + menuItem.id + "' menu item already exists.");
            }
            this.menuItems.push(menuItem);
        };
        this.removeMenuItem = (menuItemId) => {
            var menuItem = this._findMenuItem(menuItemId);
            this.menuItems.remove(menuItem);
        };
        this.selectMenuItem = (menuItem) => {
            if (!!menuItem) {
                this._viewModel.menuVisible(true);
                this._viewModel.menuItemClick(menuItem);
            }
        };
        this.toolboxGroups = ko.observableArray();
        this.addToolboxItem = (groupName, toolboxItem) => {
            var group = this._findToolboxGroup(groupName);
            if (!!group) {
                group.items.push(toolboxItem);
            }
            else {
                throw Error("The '" + groupName + "' group does not exist.");
            }
        };
        this.removeToolboxItem = (groupName, toolboxItemName) => {
            var group = this._findToolboxGroup(groupName);
            if (group) {
                this._unregisterToolboxItem(group, toolboxItemName);
            }
            else {
                throw Error("The '" + groupName + "' group does not exist.");
            }
        };
        this.toolbarGroups = ko.observableArray();
        this.addToolbarItem = (groupName, toolbarItem) => {
            var group = this.toolbarGroups().filter(group => group.name === groupName)[0];
            if (!!group) {
                group.items.push(toolbarItem);
            }
            else {
                throw Error("The '" + groupName + "' group does not exist.");
            }
        };
        this.removeToolbarItem = (groupName, toolbarItemName) => {
            var group = this.toolbarGroups().filter(group => group.name === groupName)[0];
            if (!!group) {
                var toolbarItem = group.items().filter(item => item.name === toolbarItemName)[0];
                group.items.remove(toolbarItem);
            }
            else {
                throw Error("The '" + groupName + "' group does not exist.");
            }
        };
        this.showPanelAsync = (options) => {
            return this._showPanelAsync(options, CancellationToken.None);
        };
        this._showPanelAsync = (options, cancellationToken) => {
            var def = createJQueryDeferred();
            if (cancellationToken.canceled)
                def.reject();
            else {
                this._viewModel.showDesignerPanel();
                setTimeout(() => {
                    if (!cancellationToken.canceled) {
                        options.surfaceLeft = toolboxConstants.leftPanelWidth;
                        def.resolve(options);
                    }
                    else {
                        def.reject();
                    }
                }, 500);
            }
            return def.promise();
        };
        this.hidePanelAsync = (options) => {
            return this._hidePanelAsync(options, CancellationToken.None);
        };
        this._hidePanelAsync = (options, cancellationToken) => {
            var def = createJQueryDeferred();
            if (cancellationToken.canceled)
                def.reject();
            else {
                this._viewModel.hideDesignerPanel();
                setTimeout(() => {
                    if (!cancellationToken.canceled) {
                        options.surfaceLeft = 0;
                        def.resolve(options);
                    }
                    else {
                        def.reject();
                    }
                }, 500);
            }
            return def.promise();
        };
        this._layoutItemPlaceholderService = (layoutItem) => {
            var currentDashboardItem = layoutItem._parent().viewModel.model;
            return {
                data: {
                    dashboardItems: ko.computed(() => {
                        var placeholderItems = [];
                        this._viewModel.toolboxGroupsSorted().forEach(group => {
                            group
                                .items()
                                .filter(toolboxItem => DashboardLayoutNode._canAttach(currentDashboardItem, { '@ItemType': toolboxItem.type }))
                                .forEach(toolboxItem => {
                                placeholderItems.push({
                                    type: toolboxItem.type,
                                    name: toolboxItem.name,
                                    iconName: toolboxItem.icon,
                                    title: toolboxItem.title
                                });
                            });
                        });
                        return placeholderItems;
                    }),
                    addDashboardItem: (data) => {
                        layoutItem.create(getItemJson(data.type), 'left');
                    }
                },
                name: 'dx-toolbox-extension-layout-item-placeholder'
            };
        };
        let cancelableDesignerToViewerAction = {
            orderNo: 40,
            action: (options) => {
                this.closeMenu();
                return this.hidePanelAsync(options);
            },
            cancelableAction: (options, cancellationToken) => {
                this.closeMenu();
                return this._hidePanelAsync(options, cancellationToken);
            }
        };
        this.designerToViewerAction = cancelableDesignerToViewerAction;
        let cancelableViewerToDesignerAction = {
            orderNo: 40,
            action: this.showPanelAsync,
            cancelableAction: this._showPanelAsync
        };
        this.viewerToDesignerAction = cancelableViewerToDesignerAction;
        this._createDefaultGroups();
        let draggableController = safeComputed({ dockingLayoutController: this.dashboardControl._dockingLayoutAdapter._dockingLayoutController }, (args) => {
            if (args.dockingLayoutController) {
                return args.dockingLayoutController.dragController;
            }
        });
        this.toDispose(draggableController);
        this._viewModel = new ToolboxViewModel(dashboardControl.isDesignMode(), this.dashboardControl, this.menuItems, this.toolboxGroups, this.toolbarGroups, draggableController);
        this.template = {
            data: this._viewModel,
            name: 'dx-dashboard-toolbox-extension'
        };
        if (dashboardControl.isDesignMode()) {
            dashboardControl.surfaceLeft(toolboxConstants.leftPanelWidth);
        }
        dashboardControl.subscribeExtensionsChanged({
            added: (extension) => {
                if (extension.name === 'dashboardPanel') {
                    this._switchToViewerToolbar = new DashboardToolbarGroup('viewer-button', '', 100);
                    var toViewerItem = new DashboardToolbarItem('toviewer', () => dashboardControl.switchToViewer());
                    toViewerItem.template = 'dx-dashboard-working-mode-extension-viewer-button';
                    toViewerItem.disabled = ko.pureComputed(() => !!this.dashboardControl.dashboard());
                    this._switchToViewerToolbar.items.push(toViewerItem);
                    this.toolbarGroups.push(this._switchToViewerToolbar);
                }
            },
            deleted: (extension) => {
                if (extension.name === 'dashboardPanel') {
                    this.toolbarGroups.remove(this._switchToViewerToolbar);
                }
            }
        });
        defineObsoleteProperty({
            target: this,
            memberName: 'settingsForm',
            oldMemberDisplayName: 'DevExpress.Dashboard.Designer.ToolboxExtension.settingsForm',
            newMemberDisplayName: 'DevExpress.Dasbhoard.Designer.DashboardMenuItem.template',
            action: () => { return this._viewModel.settingsForm; }
        });
        defineObsoleteProperty({
            target: this,
            memberName: 'settingsFormVisible',
            oldMemberDisplayName: 'DevExpress.Dashboard.Designer.ToolboxExtension.settingsFormVisible',
            newMemberDisplayName: 'DevExpress.Dasbhoard.Designer.DashboardMenuItem.template',
            action: () => { return this._viewModel.settingsFormVisible; }
        });
        defineObsoleteProperty({
            target: this,
            memberName: 'toggleMenu',
            oldMemberDisplayName: 'DevExpress.Dashboard.Designer.ToolboxExtension.toggleMenu',
            newMemberDisplayName: 'DevExpress.Dasbhoard.Designer.ToolboxExtension.openMenu/closeMenu',
            action: () => { return this._viewModel.toggleMenu; }
        });
        defineObsoleteMethod({
            target: this,
            memberName: 'menuItemClick',
            oldMemberDisplayName: 'DevExpress.Dashboard.Designer.ToolboxExtension.menuItemClick',
            newMemberDisplayName: 'DevExpress.Dasbhoard.Designer.ToolboxExtension.selectMenuItem',
            action: (menuItem) => { return this._viewModel.menuItemClick(menuItem); }
        });
    }
    get menuVisible() { return this._viewModel.menuVisible; }
    openMenu() {
        this._viewModel.showMenu();
    }
    closeMenu() {
        this._viewModel.closeMenu();
    }
    processKeyEvent(keyEventType, eventArgs) {
        return this._viewModel.processKeyEvent(keyEventType, eventArgs);
    }
    start() {
        var standardItems = Object.keys(itemTypesMap).filter(key => !itemTypesMap[key].customItemType);
        var itemTypeNames = standardItems.sort((t1, t2) => itemTypesMap[t1].index - itemTypesMap[t2].index);
        itemTypeNames.forEach(itemTypeName => this._registerToolboxItem(itemTypeName, itemTypesMap[itemTypeName]));
        this.dashboardControl.extensions.forEach(extension => this._registerCustomItemToolbox(extension));
        this.toDispose(this.dashboardControl.subscribeExtensionsChanged({
            added: (extension) => {
                this._registerCustomItemToolbox(extension);
            },
            deleted: (extension) => {
                this.toolboxGroups().forEach(group => this._unregisterToolboxItem(group, extension.name));
            }
        }));
        subscribeAndPerform(this.dashboardControl._actualLayoutController, (layout) => {
            if (layout && layout instanceof DockingLayoutController) {
                layout.layoutItemPlaceholderService = this._layoutItemPlaceholderService;
            }
        });
    }
    stop() {
        this.dashboardControl.surfaceLeft(0);
    }
    _registerCustomItemToolbox(extension) {
        var customItemExtension = extension;
        if (customItemExtension.metaData) {
            this._registerToolboxItem(customItemExtension.name, customItemExtension.metaData);
        }
    }
    _createDefaultGroups() {
        this.toolboxGroups.push(new DashboardToolboxGroup('common', 'DashboardWebStringId.AccordionTab.Common', 100));
        this.toolboxGroups.push(new DashboardToolboxGroup('maps', 'DashboardStringId.DescriptionMaps', 110));
        this.toolboxGroups.push(new DashboardToolboxGroup('filter', 'DashboardWebStringId.DataSources.Filter', 120));
        this.toolboxGroups.push(new DashboardToolboxGroup('layout', 'DashboardWebStringId.AccordionTab.Layout', 130));
    }
    _registerToolboxItem(itemTypeName, itemDescription) {
        var group = this._findToolboxGroup(itemDescription.groupName) || this._findToolboxGroup('custom');
        if (!group) {
            group = new DashboardToolboxGroup('custom', 'DashboardStringId.CustomItems', 130);
            this.toolboxGroups.push(group);
        }
        var itemClickHandlerCreator = (type) => () => {
            var layout = this.dashboardControl._actualLayoutController();
            layout && layout instanceof DockingLayoutController && layout.addDashboardItem({ type: type });
        };
        var item = new DashboardToolboxItem(itemTypeName, itemClickHandlerCreator(itemTypeName), getIconName(itemTypeName, itemDescription.icon), itemDescription.title, itemTypeName);
        item.disabled = ko.computed(() => !this.dashboardControl.dashboard());
        group.items.push(item);
        return itemClickHandlerCreator;
    }
    _unregisterToolboxItem(group, toolboxItemName) {
        var toolboxItem = group.items().filter(item => item.name === toolboxItemName)[0];
        if (!!toolboxItem) {
            group.items.remove(toolboxItem);
        }
    }
    _findToolboxGroup(groupName) {
        return this.toolboxGroups().filter(gr => gr.name === groupName)[0];
    }
    _findMenuItem(menuItemId) {
        return this.menuItems().filter(mi => mi.id === menuItemId)[0];
    }
}
designerExtensions[nameAlias] = (dashboardControl, options) => new ToolboxExtension(dashboardControl);
extensionNameMap[nameAlias] = name;
