// devexpress-dashboard/designer/toolbox-extension/toolbox-items.js
/**
* DevExpress Dashboard (toolbox-items.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
var leftPanelWidth = 81;
export class DashboardMenuItem {
    constructor(id, title, index, hotKey, click) {
        this.id = id;
        this.title = title;
        this.click = click;
        this.hotKey = hotKey;
        this.index = index;
        this.selected = ko.observable(false);
        this.disabled = ko.observable(false);
    }
}
export class DashboardToolboxItem {
    constructor(name, click, icon, title, type) {
        this.name = name;
        this.click = click;
        this.icon = icon;
        this.title = title;
        this.type = type;
        this.disabled = ko.observable(false);
    }
}
export class DashboardToolbarItem {
    constructor(name, click, icon, title) {
        this.name = name;
        this.click = click;
        this.icon = icon;
        this.title = title;
        this.disabled = ko.observable(false);
    }
}
export class DashboardToolboxGroup {
    constructor(name, title, index, ...items) {
        this.name = name;
        this.title = title;
        this.index = index;
        this.items = ko.observableArray(items);
    }
}
export class DashboardToolbarGroup {
    constructor(name, title, index, ...items) {
        this.name = name;
        this.title = title;
        this.index = index;
        this.items = ko.observableArray(items);
    }
}
