// devexpress-dashboard/designer/data-source-wizard/pages/_parameters-item-provider.js
/**
* DevExpress Dashboard (_parameters-item-provider.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { createJQueryDeferred } from '../../../data/_jquery-helpers';
import { getLocalizationById } from '../../../data/index.internal';
import { DataItem } from '../../../model/data-item/data-item';
export class ParametersItemProvider {
    constructor(dashboardParameters = []) {
        this.dashboardParameters = dashboardParameters;
    }
    getItems(pathRequest) {
        var items = [], fullPath = pathRequest.fullPath && pathRequest.fullPath.toLowerCase() || '';
        if (fullPath === '') {
            items = [{
                    displayName: getLocalizationById('DashboardWebStringId.DashboardParameters'),
                    name: 'Parameters',
                    isList: true,
                    specifics: 'parameters'
                }];
        }
        else if (fullPath === 'parameters') {
            items = this.dashboardParameters.map(parameter => {
                return {
                    displayName: parameter.name(),
                    name: parameter.name(),
                    isList: false,
                    specifics: DataItem.typesMap[parameter.type().replace('System.', '')]
                };
            });
        }
        return createJQueryDeferred().resolve(items).promise();
    }
}
