// devexpress-dashboard/designer/data-source-wizard/pages/choose-olap-connection-string-page.js
/**
* DevExpress Dashboard (choose-olap-connection-string-page.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { SqlDataSourceWizardPageId, WizardPageBase, _registerChooseSqlConnectionPage as _registerAnalyticChooseSqlConnectionPage } from '@devexpress/analytics-core/analytics-wizard';
import * as ko from 'knockout';
import { createJQueryDeferred } from '../../../data/_jquery-helpers';
import { OlapDataSourceWizardPageId } from './page-id';
export class DashboardChooseOlapConnectionStringPage extends WizardPageBase {
    constructor(connectionStrings) {
        super();
        this._selectedConnectionString = ko.observableArray([]);
        this._connectionStrings = connectionStrings;
        this._disposables.push(this._selectedConnectionString.subscribe(() => this._onChange()));
    }
    canNext() {
        return false;
    }
    canFinish() {
        return this._selectedConnectionString().length !== 0;
    }
    commit() {
        var deferred = createJQueryDeferred();
        if (this._selectedConnectionString()[0]) {
            deferred.resolve({
                connectionName: this._selectedConnectionString()[0].name
            });
        }
        else {
            deferred.resolve();
        }
        return deferred.promise();
    }
    initialize(state) {
        if (this._connectionStrings && this._connectionStrings.length === 1) {
            this._selectedConnectionString([this._connectionStrings[0]]);
        }
        else if (this._connectionStrings && this._connectionStrings.length > 0) {
            let matchingStrings = this._connectionStrings.filter(c => c.name == state.connectionName);
            var selectedString = matchingStrings.length > 0 ? matchingStrings[0] : this._connectionStrings[0];
            this._selectedConnectionString(selectedString ? [selectedString] : []);
        }
        else {
            this._selectedConnectionString([]);
        }
        return createJQueryDeferred().resolve().promise();
    }
}
export function _registerOlapConnectionStringsPage(factory, connectionStrings) {
    let sqlConnectionStringsMeta = factory.getMetadata(SqlDataSourceWizardPageId.ChooseConnectionPage);
    if (!sqlConnectionStringsMeta) {
        _registerAnalyticChooseSqlConnectionPage(factory, undefined);
        sqlConnectionStringsMeta = factory.getMetadata(SqlDataSourceWizardPageId.ChooseConnectionPage);
        factory.unregisterMetadata(SqlDataSourceWizardPageId.ChooseConnectionPage);
    }
    factory.registerMetadata(OlapDataSourceWizardPageId.ChooseConnectionPage, {
        create: () => {
            return new DashboardChooseOlapConnectionStringPage(connectionStrings);
        },
        setState: (result, state) => {
            state.connectionName = result.connectionName;
        },
        getState: (state) => {
            return state.olapDataSourceWizard;
        },
        resetState: (state, defaultState) => {
            state.connectionName = defaultState.connectionName;
        },
        template: sqlConnectionStringsMeta.template,
        description: sqlConnectionStringsMeta.description
    });
}
