// devexpress-dashboard/viewer-parts/conditional-formatting/_chart-item-style-settings-provider.js
/**
* DevExpress Dashboard (_chart-item-style-settings-provider.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { appearanceSettingsProvider } from './_appearance-settings-provider';
import { styleSettingsProvider } from './_style-settings-provider';
export class ChartItemStyleSettingsProvider {
    initialize(cfModel) {
        if (cfModel)
            this.cfModel = cfModel;
    }
    getDefaultBackColor() {
        return appearanceSettingsProvider.getDefaultColorizationColor();
    }
    getBackColor(styleSettingsInfo) {
        let styleIndexes = styleSettingsInfo ? styleSettingsInfo.styleIndexes : undefined, styleSettingsModel;
        if (styleIndexes && styleIndexes.length > 0) {
            let resultIndex = styleIndexes[styleIndexes.length - 1];
            styleSettingsModel = this.cfModel.FormatConditionStyleSettings[resultIndex];
            return this.getBackColorFromStyleSettingsModel(styleSettingsModel);
        }
    }
    getBackColorFromStyleSettingsModel(styleSettingsModel) {
        if (styleSettingsModel.RangeIndex) {
            let condition = this.cfModel.RuleModels[styleSettingsModel.RuleIndex].ConditionModel;
            styleSettingsModel = styleSettingsProvider._getRangeBackColorStyleSettings(styleSettingsModel, condition);
        }
        return appearanceSettingsProvider.getColor(styleSettingsModel);
    }
}
