// devexpress-dashboard/model/items/treemap/treemap-item.js
/**
* DevExpress Dashboard (treemap-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { ModelSerializer, deserializeArray } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { DataItemLink } from '../../data-item/data-item';
import { collectionItemType } from '../../internal/_utils';
import { BindingSectionTitles } from '../../metadata/_base-metadata';
import { itemTypesMap } from '../../serializable-model';
import { DataDashboardItem } from '../data-dashboard-item';
import { treeMapArgumentsMeta, treeMapvalues, treemapDashboardItemSerializationsInfo } from './metadata/_treemap-item';
export class TreemapItem extends DataDashboardItem {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
        this.values = ko.observableArray([]);
        this.__arguments = ko.observableArray([]);
        this.arguments = ko.observableArray([]);
        this.addDisposable(this.__arguments.subscribe((newArguments) => {
            DataDashboardItem._addColoringMeta(newArguments);
        }));
        this.__arguments(deserializeArray(modelJson.Arguments, (item) => new DataItemLink(this, item, serializer))());
        this._subscribeDataItemLinkArrays(treeMapArgumentsMeta);
        this.__values = deserializeArray(modelJson.Values, (item) => new DataItemLink(this, item, serializer));
        this._subscribeDataItemLinkArrays(treeMapvalues);
    }
    _getInfoCore() {
        return treemapDashboardItemSerializationsInfo;
    }
    _clearBindings() {
        super._clearBindings();
        this.disposeObservableArray(this.__arguments);
        this.disposeObservableArray(this.__values);
        this.__arguments.removeAll();
        this.__values.removeAll();
    }
    _isCalculationSupported() {
        return false;
    }
    _getDefaultItemType() {
        return 'Treemap';
    }
    _getCanColorByMeasures() { return true; }
    _getCanColorByDimensions() { return true; }
    _getAreMeasuresColoredByDefault() { return this.__values().length > 1 && this.__arguments().length === 0; }
    _getIsDimensionColoredByDefault(dimension) {
        return this.__arguments().length > 0 && this.__arguments()[0].dataItem() === dimension;
    }
    _getMasterFilterMode() { return this.interactivityOptions.masterFilterMode(); }
    _getDrillDownEnabled() { return this.interactivityOptions.isDrillDownEnabled(); }
    _getIgnoreMasterFilter() { return this.interactivityOptions.ignoreMasterFilters(); }
    _getInteractivityDimensionLinks() { return this.__arguments(); }
    _getLayersCount() {
        return !!this.__arguments && this.__arguments().length > 0 ? this.__values().length : 0;
    }
    _getLayerName() {
        return this._getDataItemDisplayName(this.__values()[this._selectedElementIndex() || 0].dataItem());
    }
    _getColorizableDataItemsInfo() {
        return [{
                items: this.__arguments(),
                prefixId: BindingSectionTitles.Arguments
            }];
    }
    _isSortingEnabled() {
        return false;
    }
}
__decorate([
    collectionItemType('Value')
], TreemapItem.prototype, "__values", void 0);
__decorate([
    collectionItemType('Argument')
], TreemapItem.prototype, "__arguments", void 0);
itemTypesMap['Treemap'] = { type: TreemapItem, groupName: 'common', title: 'DashboardStringId.DefaultNameTreemapItem', index: 40 };
