// devexpress-dashboard/common/extensions/_parameter-dialog-binder.js
/**
* DevExpress Dashboard (_parameter-dialog-binder.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { ParametersCollection } from '../../data/_parameters';
import { DisposableObject } from '../../model/disposable-object';
import { Parameter } from '../../model/parameters/parameter';
import { parameterTypes } from '../../viewer-parts/widgets/dialogs/_parameters-dialog';
export class ParameterDialogViewModel extends DisposableObject {
    constructor(_parameters, getParameterValues) {
        super();
        this._parameters = _parameters;
        this._getDashboardParameterType = (type) => {
            var result = parameterTypes.string;
            if (type) {
                if (type.indexOf('DateTime') > -1) {
                    result = parameterTypes.dateTime;
                }
                else if ((type.indexOf('Decimal') + (type.indexOf('Double'))) > -2) {
                    result = parameterTypes.float;
                }
                else if (type.indexOf('Boolean') > -1) {
                    result = parameterTypes.bool;
                }
                else if (type.indexOf('.Int') > -1) {
                    result = parameterTypes.int;
                }
            }
            return result;
        };
        this.setParameters = (newParameters) => {
            newParameters.forEach(newParameter => {
                var parameterModel = this._parameters.peek().filter(p => p.name.peek() === newParameter.getName())[0];
                if (parameterModel) {
                    parameterModel._value(newParameter.getValue());
                }
            });
        };
        this.parameterCollection = ko.computed(() => {
            var parameters = this._parameters();
            var parameterCollection = new ParametersCollection(parameters.map(param => {
                var values = this._getParameterValues(param, getParameterValues);
                return {
                    Name: param.name(),
                    DefaultValue: this._getParameterDefaultValue(param, values),
                    Description: param.description(),
                    Values: values,
                    ContainsDisplayMember: param.containsDisplayMember(),
                    Visible: param.parameterVisible(),
                    AllowMultiselect: param.allowMultiselect(),
                    AllowNull: param.allowNull(),
                    Type: this._getDashboardParameterType(param.type()),
                    OriginalType: param.type()
                };
            }));
            parameters.forEach(param => {
                var actualValue = param._actualValue();
                if (actualValue === Parameter.SelectAllValue) {
                    actualValue = this._getParameterValues(param, getParameterValues).map(valueViewModel => valueViewModel.Value);
                }
                parameterCollection.setParameterValue(param.name(), actualValue);
            });
            parameterCollection.collectionChanged.add(() => {
                this.setParameters(parameterCollection.getParameterList());
            });
            return parameterCollection;
        });
        this.toDispose(this.parameterCollection);
    }
    _getParameterValues(parameter, getParameterValues) {
        var parameterValues = null;
        if (!!parameter.staticListLookUpSettings()) {
            parameterValues = parameter.staticListLookUpSettings().values().map(item => {
                return {
                    Value: item.value(),
                };
            });
        }
        else if (!!parameter.dynamicListLookUpSettings()) {
            parameterValues = getParameterValues(parameter.type(), parameter.dynamicListLookUpSettings())();
        }
        return parameterValues;
    }
    _getParameterDefaultValue(param, paramValues) {
        if (param.lookUpSourceType() !== 'None' && param.allowMultiselect()) {
            if (param.selectAllValues()) {
                return paramValues && paramValues.map(value => value.Value) || [];
            }
            else {
                return param._valuesOfDefaultValues() || [];
            }
        }
        else {
            return param.defaultValue();
        }
    }
}
