// devexpress-dashboard/common/extensions/parameter-dialog-extension.js
/**
* DevExpress Dashboard (parameter-dialog-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { $unwrap } from '../../data/_jquery-helpers';
import { DisposableObject } from '../../model/disposable-object';
import { defineObsoleteMethod } from '../../model/internal/_obsolete-helper';
import { DashboardLayoutModeHelper } from '../../viewer-parts/_dashboard-layout-mode-helper';
import { parametersDialog } from '../../viewer-parts/widgets/dialogs/_parameters-dialog';
import { defaultExtensions, extensionNameMap } from '../control-options';
import { OptionsManager } from '../internal/_options-manager';
import { ParameterDialogViewModel } from './_parameter-dialog-binder';
const name = 'dashboardParameterDialog';
const nameAlias = 'dashboard-parameter-dialog';
export class DashboardParameterDialogExtension extends DisposableObject {
    constructor(dashboardControl, options) {
        super();
        this._customDialogContent = [];
        this._optionsManager = new OptionsManager();
        this.name = name;
        this.showDialogButton = ko.observable(true);
        this._updateViewModel = (dashboard) => {
            this._clear();
            if (!!dashboard) {
                this._viewModel = new ParameterDialogViewModel(dashboard.parameters, (parameterType, settings) => this._dashboardControl._dataSourceBrowser.getParameterValues(parameterType, settings));
                this._viewModel.parameterCollection.subscribe((newParamCollection) => {
                    var _a, _b;
                    const parameters = (_a = (this._viewModel.parameterCollection && this._viewModel.parameterCollection()._parameters)) !== null && _a !== void 0 ? _a : [];
                    const newParameters = (_b = newParamCollection === null || newParamCollection === void 0 ? void 0 : newParamCollection._parameters) !== null && _b !== void 0 ? _b : [];
                    if (this._shouldResetCustomDialogContent(parameters, newParameters)) {
                        this._clearContent();
                        this._customDialogContent.concat(this._parameterDialog)
                            .filter(dialog => dialog && dialog.setActualState)
                            .forEach(dialog => dialog.setActualState());
                    }
                });
            }
        };
        this._optionsManager.initialize({
            extensionName: name,
            dashboardControl: dashboardControl,
            defaultOptions: {},
            eventsHolder: this,
            initOptions: options,
            optionChanged: (args) => null
        });
        this._dashboardControl = dashboardControl;
        this._onShowing = this._optionsManager.getInitialOptions().onShowing || (() => { });
        this._onShown = this._optionsManager.getInitialOptions().onShown || (() => { });
        this._onHidden = this._optionsManager.getInitialOptions().onHidden || (() => { });
    }
    get onShowing() {
        return this._onShowing;
    }
    set onShowing(value) {
        if (this._onShowing) {
            this.off('showing', this._onShowing);
        }
        this._onShowing = value;
        this.on('showing', value);
    }
    get onShown() {
        return this._onShown;
    }
    set onShown(value) {
        if (this._onShown) {
            this.off('shown', this._onShown);
        }
        this._onShown = value;
        this.on('shown', value);
    }
    get onHidden() {
        return this._onHidden;
    }
    set onHidden(value) {
        if (this._onHidden) {
            this.off('hidden', this._onHidden);
        }
        this._onHidden = value;
        this.on('hidden', value);
    }
    start() {
        if (this._dashboardControl.dashboard()) {
            this._updateViewModel(this._dashboardControl.dashboard());
            this._subscribeDynamicLookUpValuesLoaded();
        }
        this.toDispose(this._dashboardControl.dashboard.subscribe(newDashboard => this._updateViewModel(newDashboard)));
        this.toDispose(this._dashboardControl.dashboard.subscribe(() => this._subscribeDynamicLookUpValuesLoaded()));
        defineObsoleteMethod({
            target: this,
            memberName: 'showDialog',
            oldMemberDisplayName: 'DashboardParameterDialogExtension.showDialog',
            newMemberDisplayName: 'DashboardParameterDialogExtension.show',
            action: () => this.show()
        });
        defineObsoleteMethod({
            target: this,
            memberName: 'hideDialog',
            oldMemberDisplayName: 'DashboardParameterDialogExtension.hideDialog',
            newMemberDisplayName: 'DashboardParameterDialogExtension.hide',
            action: () => this.hide()
        });
    }
    stop() {
        this._clear();
        this.dispose();
    }
    show() {
        if (this._parameterDialog) {
            this._parameterDialog.dispose();
        }
        this._parameterDialog = this._createParameterDialog();
        this._parameterDialog.show();
    }
    hide() {
        if (!!this._parameterDialog) {
            this._parameterDialog.hide();
        }
    }
    subscribeToContentChanges(callback) {
        return this._viewModel.parameterCollection.subscribe(callback);
    }
    getParameters() {
        return this._getParameters();
    }
    _getParameters() {
        if (!this._viewModel)
            throw 'Dashboard is not loaded';
        return this._viewModel.parameterCollection();
    }
    renderContent(element) {
        var customParameterDialog = this._createParameterDialog();
        this._customDialogContent.push(customParameterDialog);
        return customParameterDialog.generateContent($unwrap(element), () => {
            this._customDialogContent.splice(this._customDialogContent.indexOf(customParameterDialog), 1);
        });
    }
    _createParameterDialog() {
        return new parametersDialog({
            parametersDialogContainer: DashboardLayoutModeHelper.isMobile ? window.document.body : this._dashboardControl.getWidgetContainer(),
            getParametersCollection: () => this._getParameters(),
            submitParameters: (newParameters) => {
                var _a;
                (_a = this._dashboardControl._dataSourceBrowser) === null || _a === void 0 ? void 0 : _a.clearDimensionValuesCache();
                this._viewModel.parameterCollection.peek().setParameters(newParameters);
            },
            onShowing: (args) => this._optionsManager.raiseEvent('showing', args),
            onShown: (args) => this._optionsManager.raiseEvent('shown', args),
            onHidden: (args) => this._optionsManager.raiseEvent('hidden', args)
        });
    }
    _clearContent() {
        this._customDialogContent.forEach(dialog => dialog.dispose());
        this._customDialogContent = [];
    }
    _clear() {
        this._clearContent();
        if (this._parameterDialog) {
            this._parameterDialog.dispose();
            this._parameterDialog = undefined;
        }
        if (this._viewModel) {
            this._viewModel.dispose();
            this._viewModel = undefined;
        }
    }
    _subscribeDynamicLookUpValuesLoaded() {
        var dataSourceBrowser = this._dashboardControl._dataSourceBrowser;
        if (dataSourceBrowser) {
            dataSourceBrowser.dynamicLookUpValuesLoaded = (dynamicListLookUpSettings) => {
                this._dashboardControl
                    .dashboard()
                    .parameters()
                    .filter(param => {
                    return param.dynamicListLookUpSettings() && param.dynamicListLookUpSettings().equals(dynamicListLookUpSettings);
                })
                    .map(param => param.name())
                    .forEach(paramName => {
                    this._optionsManager.raiseEvent('dynamicLookUpValuesLoaded', { parameterName: paramName });
                });
            };
        }
    }
    _shouldResetCustomDialogContent(parameters, newParameters) {
        if (parameters.length !== newParameters.length)
            return true;
        for (const p of parameters) {
            const newParam = newParameters.find(np => np._name === p._name);
            if (!newParam || newParam._value !== p._value)
                return true;
        }
        return false;
    }
}
defaultExtensions[nameAlias] = (dashboardControl, options) => new DashboardParameterDialogExtension(dashboardControl, options);
extensionNameMap[nameAlias] = name;
