// devexpress-dashboard/common/extensions/url-state-extension.js
/**
* DevExpress Dashboard (url-state-extension.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { DisposableObject } from '../../model/disposable-object';
import { ViewerDataServiceClient } from '../_service-client';
import { defaultExtensions, extensionNameMap } from '../control-options';
const name = 'urlState';
const nameAlias = 'url-state';
export class UrlStateExtension extends DisposableObject {
    constructor(dashboardControl, options = {}) {
        super();
        this._dashboardContaierSubscriptions = [];
        this._defaultOptions = {
            includeDashboardIdToUrl: false,
            includeDashboardStateToUrl: false
        };
        this.name = name;
        this._options = Object.assign(Object.assign({}, this._defaultOptions), options);
        this._dashboardControl = dashboardControl;
    }
    start() {
        this.toDispose(this._dashboardControl.dashboardContainer.subscribe(dashboardContainer => {
            this._dashboardContaierSubscriptions.forEach(disposable => disposable.dispose());
            this._dashboardContaierSubscriptions = [];
            this._processDashboardChanged(dashboardContainer);
        }));
        this.toDispose(this._dashboardControl.isDesignMode.subscribe(isDesignMode => {
            this._updateDashboardState();
        }));
        this._processDashboardChanged(this._dashboardControl.dashboardContainer());
    }
    stop() {
        this.dispose();
    }
    _processDashboardChanged(dashboardInfo) {
        if (dashboardInfo) {
            if (this._options.includeDashboardIdToUrl) {
                this._updateUrl('dashboardId', dashboardInfo.id);
            }
            this._updateDashboardState();
            if (!!dashboardInfo.dashboard) {
                this._dashboardContaierSubscriptions.push(dashboardInfo.dashboard._state.subscribe(_ => this._updateDashboardState()));
            }
        }
    }
    _updateDashboardState() {
        var dashboard = this._dashboardControl.dashboardContainer() && this._dashboardControl.dashboardContainer().dashboard || null;
        var state = dashboard && !this._dashboardControl.isDesignMode() ? dashboard.stateString : null;
        if (this._options.includeDashboardStateToUrl) {
            this._updateUrl('dashboardState', state);
        }
    }
    _updateUrl(key, value) {
        var newUrl = this._replaceValue(key, value);
        if (newUrl) {
            if (newUrl.length > ViewerDataServiceClient.maxQueryStringLength) {
                newUrl = this._replaceValue(key, null);
            }
            this._setUrl(newUrl);
        }
    }
    _replaceValue(key, value) {
        var uri = this._getUrl();
        var re = new RegExp('([?&])' + key + '=.*?(&|$)', 'i');
        var separator = uri.indexOf('?') !== -1 ? '&' : '?';
        var newParameterValue = value ? key + '=' + encodeURIComponent(value) : '';
        if (uri.match(re)) {
            var separator = !!newParameterValue ? '$1' : '';
            return uri.replace(re, separator + newParameterValue + '$2');
        }
        else if (!!newParameterValue) {
            return uri + separator + newParameterValue;
        }
        else {
            return uri;
        }
    }
    _getUrl() {
        return location.href;
    }
    _setUrl(url) {
        if (!url)
            url = location.pathname;
        history.replaceState({}, '', url);
    }
}
defaultExtensions[nameAlias] = (dashboardControl, options) => new UrlStateExtension(dashboardControl, options);
extensionNameMap[nameAlias] = name;
