// devexpress-dashboard/designer/items/binding-details/_data-item-container-seed.js
/**
* DevExpress Dashboard (_data-item-container-seed.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { DataItemLink } from '../../../model';
import { IsOlapHierarchyField } from '../../../model/data-sources/_data-field';
import { DataItemContainer } from '../../../model/items/data-item-container';
import { itemType, name } from '../../../model/metadata/_base-metadata';
export class DataItemContainerSeed extends DataItemContainer {
    constructor(dataItemProvider, dataItemType, _fieldConstraint) {
        super();
        this.dataItemType = dataItemType;
        this._fieldConstraint = _fieldConstraint;
        this.name = ko.observable('NewContainer');
        this.dataLink = new DataItemLink(dataItemProvider);
        this.dataLink.itemType('Seed');
    }
    _getDataId() {
        return null;
    }
    _getContainerType() {
        return 'Stub';
    }
    _getBindingModel() {
        return [{
                propertyName: 'dataLink',
                dataItemType: this.dataItemType,
                emptyPlaceholder: 'Data',
                selectedPlaceholder: 'Configure Data',
                fieldConstraint: this._fieldConstraint
            }];
    }
    grow(dataItemProvider, bindingProperty, dataField) {
        var containerModels = [];
        if (!IsOlapHierarchyField(dataField)) {
            var cm = bindingProperty.creator(undefined, dataField);
            var dataItem = dataItemProvider._createDataItem(dataField, cm._getBindingModel()[0]);
            let newDataItemBinding = cm._getBindingModel()[0];
            cm[newDataItemBinding.propertyName].uniqueName(dataItem.uniqueName());
            containerModels.push(cm);
        }
        else {
            dataField.groupDataItems.forEach(groupItem => {
                var cm = bindingProperty.creator(undefined, groupItem);
                var dataItem = dataItemProvider._createDataItem(groupItem, cm._getBindingModel()[0]);
                dataItem.groupIndex(dataField.groupIndex());
                let newDataItemBinding = cm._getBindingModel()[0];
                cm[newDataItemBinding.propertyName].uniqueName(dataItem.uniqueName());
                containerModels.push(cm);
            });
        }
        return containerModels;
    }
    _getInfoCore() {
        return [itemType, name];
    }
    _getDefaultItemType() {
        return 'Stub';
    }
}
