// devexpress-dashboard/designer/filtering/_item-filter-items-provider.js
/**
* DevExpress Dashboard (_item-filter-items-provider.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import dateLocalization from 'devextreme/localization/date';
import { getDataFields } from '../../common/_data-source-browser';
import { createJQueryDeferred } from '../../data/_jquery-helpers';
import { DataItem } from '../../model/data-item/data-item';
import { Dimension } from '../../model/data-item/dimension';
import { IsNumeric, IsTextual } from '../../model/data-sources/_data-field';
import { getDataItemDisplayName } from '../_display-name-provider';
import { getValuesList } from './_filter-utils';
export var getRealDimensionType = (dimension, dataField) => {
    return isCategoricalDateTime(dimension, dataField) ? 'Integer' : dataField.fieldType();
};
export var isCategoricalDateTime = (dimension, dataField) => {
    return dataField.fieldType() === 'DateTime' && ['Year', 'Quarter', 'Month', 'Day', 'Hour', 'Minute', 'Second',
        'DayOfYear', 'DayOfWeek', 'WeekOfYear', 'WeekOfMonth'].indexOf(dimension.dateTimeGroupInterval() || 'Year') !== -1;
};
export class ItemFilterItemsProvider {
    constructor(dataItemValuesProvider, dataFieldProvider, parameters, dataDashboardItem, filterPredicate = () => true) {
        this.dataItemValuesProvider = dataItemValuesProvider;
        this.dataFieldProvider = dataFieldProvider;
        this.parameters = parameters;
        this.dataDashboardItem = dataDashboardItem;
        this.filterPredicate = filterPredicate;
    }
    getItems(pathRequest) {
        var deferred = createJQueryDeferred();
        if (pathRequest.fullPath === 'Parameters') {
            deferred.resolve(this.parameters()
                .map(parameter => {
                return {
                    displayName: parameter.name(),
                    name: parameter.name()
                };
            }));
        }
        else {
            this._getDashboardItemDataFields().done(dataFields => {
                deferred.resolve(this.dataDashboardItem
                    ._uniqueDataItems
                    .filter(this.filterPredicate)
                    .map(dataItem => {
                    let dataField = dataFields.filter(dataField => dataField.dataMember() === dataItem.dataMember())[0];
                    let itemType = 'string';
                    if (dataField) {
                        itemType = dataItem instanceof Dimension ? getRealDimensionType(dataItem, dataField) : dataField.fieldType();
                    }
                    return {
                        displayName: getDataItemDisplayName(this.dataFieldProvider, this.dataDashboardItem, dataItem),
                        name: dataItem.uniqueName(),
                        specifics: DataItem.typesMap[itemType] || 'string'
                    };
                }));
            });
        }
        return deferred.promise();
    }
    getValues(pathRequest) {
        var dataItem = this.dataDashboardItem._dimensions.filter(di => di.uniqueName() === pathRequest.path)[0];
        var def = createJQueryDeferred();
        if (!dataItem) {
            return def.resolve([]).promise();
        }
        this._getDashboardItemDataFields()
            .done(dataFields => {
            var dataField = dataFields.filter(dataField => dataField.dataMember() === dataItem.dataMember())[0];
            if (!dataField) {
                return def.resolve([]).promise();
            }
            if (dataItem.dateTimeGroupInterval() === 'DayOfWeek' && dataField.fieldType() === 'DateTime') {
                def.resolve([0, 1, 2, 3, 4, 5, 6].map(index => ({
                    value: index,
                    display: dateLocalization.getDayNames(undefined)[index]
                })));
            }
            else if (IsNumeric(dataField.fieldType()) || IsTextual(dataField.fieldType()) || isCategoricalDateTime(dataItem, dataField) ||
                dataItem.dateTimeGroupInterval() === 'WeekYear') {
                getValuesList({
                    dataItemValuesProvider: this.dataItemValuesProvider,
                    dataField: dataField,
                    dataItem: dataItem,
                    dataMember: this.dataDashboardItem.dataMember(),
                    dataSource: this.dataDashboardItem.dataSource()
                }).done((res) => def.resolve(res));
            }
            else {
                return def.resolve([]);
            }
        });
        return def.promise();
    }
    _getDashboardItemDataFields() {
        let fieldsNames = this.dataDashboardItem
            ._uniqueDataItems
            .filter(this.filterPredicate)
            .map(dataItem => dataItem.dataMember());
        return getDataFields(fieldsNames, this.dataDashboardItem.dataSource(), this.dataDashboardItem.dataMember(), this.dataFieldProvider);
    }
}
