// devexpress-dashboard/model/items/card/metadata/_card-layout-template.js
/**
* DevExpress Dashboard (_card-layout-template.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { CardLayout } from '../card-layout';
import { CardLayoutTemplateDataElement, CardLayoutTemplateDeltaElement, CardLayoutTemplateSparklineElement } from '../card-layout-template-element';
import { maxWidth, minWidth } from './_card-layout';
export let type = { propertyName: 'type', modelName: '@Type' };
export let templateSerializationInfo = [minWidth, maxWidth, type];
export let cardLayout = { propertyName: 'layout', modelName: 'Layout', type: CardLayout };
export let customTemplateSerializationInfo = templateSerializationInfo.concat([cardLayout]);
export let mainValue = { propertyName: 'mainValue', modelName: 'MainValue', type: CardLayoutTemplateDataElement };
export let subValue = { propertyName: 'subValue', modelName: 'SubValue', type: CardLayoutTemplateDataElement };
export let deltaIndicator = { propertyName: 'deltaIndicator', modelName: 'DeltaIndicator', type: CardLayoutTemplateDeltaElement };
export let cardLayoutSparkline = { propertyName: 'sparkline', modelName: 'Sparkline', type: CardLayoutTemplateSparklineElement };
export let bottomValue = { propertyName: 'bottomValue', modelName: 'BottomValue', type: CardLayoutTemplateDataElement };
export let bottomSubValue1 = { propertyName: 'bottomSubValue1', modelName: 'BottomSubValue1', type: CardLayoutTemplateDataElement };
export let bottomSubValue2 = { propertyName: 'bottomSubValue2', modelName: 'BottomSubValue2', type: CardLayoutTemplateDataElement };
export let deltaCompactTemplateSerializationInfo = templateSerializationInfo.concat([mainValue, subValue, bottomValue, bottomSubValue1, bottomSubValue2, deltaIndicator, cardLayoutSparkline]);
export let topValue = { propertyName: 'topValue', modelName: 'TopValue', type: CardLayoutTemplateDataElement };
export let bottomValue1 = { propertyName: 'bottomValue1', modelName: 'BottomValue1', type: CardLayoutTemplateDataElement };
export let bottomValue2 = { propertyName: 'bottomValue2', modelName: 'BottomValue2', type: CardLayoutTemplateDataElement };
export let autofitTemplateSerializationInfo = templateSerializationInfo.concat([topValue, mainValue, subValue, bottomValue1, bottomValue2, deltaIndicator, cardLayoutSparkline]);
export let nameValueTemplateSerializationInfo = templateSerializationInfo.concat([mainValue, subValue, bottomValue, deltaIndicator, cardLayoutSparkline]);
export let ultraCompactTemplateSerializationInfo = templateSerializationInfo.concat([mainValue, subValue, bottomValue, bottomSubValue1, bottomSubValue2, deltaIndicator, cardLayoutSparkline]);
