// devexpress-dashboard/model/items/card/metadata/_card-row-element.js
/**
* DevExpress Dashboard (_card-row-element.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { itemType } from '../../../metadata/_base-metadata';
export let hAlignment = { propertyName: 'hAlignment', modelName: '@HAlignment' };
export let cardRowElementSerializationsInfo = [itemType, hAlignment];
export let cardRowElementColor = { propertyName: 'color', modelName: '@ForeColor' };
export let cardRowFontFamily = { propertyName: 'fontFamily', modelName: '@FontFamily' };
export let cardRowFontSize = { propertyName: 'fontSize', modelName: '@FontSize' };
export let predefinedForeColor = { propertyName: 'predefinedForeColor', modelName: '@PredefinedForeColor' };
export let cardRowTextElementBaseSerializationInfo = cardRowElementSerializationsInfo.concat([cardRowElementColor, cardRowFontFamily, cardRowFontSize, predefinedForeColor]);
export let cardRowDataElementValueType = { propertyName: 'valueType', modelName: '@ValueType', defaultVal: 'DimensionValue' };
export let cardRowDataElementDimensionIndex = { propertyName: 'dimensionIndex', modelName: '@DimensionIndex', defaultVal: 0 };
export let cardRowDataElementSerializationInfo = cardRowTextElementBaseSerializationInfo.concat([cardRowDataElementValueType, cardRowDataElementDimensionIndex]);
export let text = { propertyName: 'text', modelName: '@Text' };
export let cardRowTextElementSerializationInfo = cardRowTextElementBaseSerializationInfo.concat([text]);
export let size = { propertyName: 'size', modelName: '@Size', defaultVal: 16 };
export let cardRowIndicatorElementSerializationInfo = cardRowElementSerializationsInfo.concat([size]);
