// devexpress-dashboard/designer/conditional-formatting/custom-style-settings/_popover-color-picker.js
/**
* DevExpress Dashboard (_popover-color-picker.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getLocalizationById } from '../../../data/localization/_default';
export class PopoverColorPicker {
    constructor(setColor) {
        this.setColor = setColor;
        this.color = ko.observable('');
        this.visible = ko.observable(false);
        this.buttonItems = ko.observableArray([
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: getLocalizationById('DashboardWebStringId.ButtonConfirm'), onClick: () => {
                        this.setColor(this.color());
                        this.visible(false);
                    } } },
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: getLocalizationById('DashboardStringId.ButtonCancel'), onClick: () => this.visible(false) } }
        ]);
    }
    show(color) {
        this.color(color);
        this.visible(true);
    }
}
