// devexpress-dashboard/model/items/map/metadata/_chorolpeth-map-item.js
/**
* DevExpress Dashboard (_chorolpeth-map-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { dataItemLinkSerializationsInfo } from '../../../data-item/metadata/_data-item';
import { parseBool, PropertyCategory } from '../../../metadata/_base-metadata';
import { colorLegend, mapDashboardItemSerializationsInfo } from './_map-item';
export let maps = {
    propertyName: 'maps',
    modelName: 'Maps',
    displayName: 'DashboardWebStringId.Binding.Maps',
    array: true
};
export let includeSummaryValueToShapeTitle = {
    propertyName: 'includeSummaryValueToShapeTitle',
    modelName: '@IncludeSummaryValueToShapeTitle',
    displayName: 'DashboardWebStringId.Map.IncludeSummaryValueToShapeTitle',
    defaultVal: false,
    simpleFormAdapterItem: 'yesNoButtonGroupEditor',
    from: parseBool
};
export let attributeDimension = {
    propertyName: '__attributeDimension',
    modelName: 'AttributeDimension',
    displayName: 'DashboardWebStringId.Map.AttributeDimension',
    info: dataItemLinkSerializationsInfo
};
export let attributeName = {
    propertyName: 'attributeName',
    modelName: '@AttributeName',
    displayName: 'DashboardWebStringId.Map.AttributeName',
    defaultVal: '',
    category: PropertyCategory.Map
};
export let tooltipAttributeName = {
    propertyName: 'tooltipAttributeName',
    modelName: '@ToolTopAttributeName',
    displayName: 'DashboardWebStringId.Map.TooltipAttribute',
    defaultVal: '',
    category: PropertyCategory.Map
};
export let choroplethMapDashboardItemSerializationsInfo = mapDashboardItemSerializationsInfo.concat([maps, attributeDimension, attributeName, includeSummaryValueToShapeTitle, tooltipAttributeName, colorLegend]);
