// devexpress-dashboard/model/parameters/dynamic-list-lookup-settings.js
/**
* DevExpress Dashboard (dynamic-list-lookup-settings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { SerializableModel } from '../serializable-model';
import { dynamicListLookUpSettingsSerializationsInfo, sortOrder } from './metadata/_dynamic-list-lookup-settings';
export class DynamicListLookUpSettings extends SerializableModel {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
        let clearMembers = () => {
            this.valueMemberName(null);
            this.displayMemberName(null);
            this.sortByMember(null);
        };
        this.addDisposable(this._dataSource = ko.computed({
            read: () => {
                return this.dataSource();
            },
            write: (val) => {
                this.dataMember(null);
                clearMembers();
                this.dataSource(val);
            }
        }));
        this.addDisposable(this._dataMember = ko.computed({
            read: () => {
                return this.dataMember();
            },
            write: (val) => {
                clearMembers();
                this.dataMember(val);
            }
        }));
    }
    getInfo() {
        return dynamicListLookUpSettingsSerializationsInfo;
    }
    equals(target) {
        return this.dataMember() === target.dataMember() &&
            this.dataSource() === target.dataSource() &&
            this.valueMemberName() === target.valueMemberName() &&
            this.displayMemberName() === target.displayMemberName() &&
            this.sortByMember() === target.sortByMember() &&
            this.sortOrder() === target.sortOrder();
    }
    isPropertyDisabled(propertyName) {
        if (propertyName === sortOrder.propertyName) {
            return !this.sortByMember();
        }
        return false;
    }
}
DynamicListLookUpSettings.modelName = 'DynamicListLookUpSettings';
export let _dynamicListLookUpSettingsSerializationInfo = { propertyName: 'dynamicListLookUpSettings', displayName: 'DashboardWebStringId.Parameters.DynamicListLookUpSettings', type: DynamicListLookUpSettings, defaultVal: null, alwaysSerialize: true };
