// devexpress-dashboard/model/parameters/parameter.js
/**
* DevExpress Dashboard (parameter.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { deepExtend } from '../../data/_jquery-helpers';
import { toStringArray } from '../internal/_date-utils';
import { collectionItemType } from '../internal/_utils';
import { TypedSerializableModel } from '../serializable-model';
import { ParameterHelper } from './_parameters-helper';
import { _dynamicListLookUpSettingsSerializationInfo, DynamicListLookUpSettings } from './dynamic-list-lookup-settings';
import { LookUpValue } from './look-up-value';
import { dashboardParameterSerializationsInfo, defaultValue } from './metadata/_parameter';
import { _staticListLookUpSettingsSerializationInfo, StaticListLookUpSettings } from './static-list-lookup-settings';
export function _getParametersQuery(parameters) {
    return parameters.map(p => {
        return {
            name: p.name(),
            value: toStringArray(p._actualValue()),
            type: p.type(),
            allowMultiselect: p.allowMultiselect(),
            selectAll: p.allowMultiselect() && p.selectAllValues() && p._actualValue() === Parameter.SelectAllValue
        };
    });
}
export class Parameter extends TypedSerializableModel {
    constructor(modelJson = {}, serializer = new ModelSerializer(), _allParameters) {
        super(modelJson, serializer);
        this._allParameters = _allParameters;
        this.lookUpSourceType = ko.observable('None');
        this.staticListLookUpSettings = ko.observable(null);
        this.dynamicListLookUpSettings = ko.observable(null);
        this._paramDialogValue = ko.observable();
        this.addDisposable(this._value = ko.computed({
            read: () => this._paramDialogValue(),
            write: val => this._paramDialogValue(val)
        }));
        this.defaultValues = deserializeArray(modelJson.Values, (item) => new LookUpValue(item, serializer));
        this.addDisposable(this._valuesOfDefaultValues = ko.computed(() => {
            return this.defaultValues() && this.defaultValues().map(val => val.value()) || null;
        }));
        if (this._type()) {
            var typeParts = this._type().split(',');
            if (typeParts.length > 1) {
                this._type(typeParts[0]);
            }
        }
        this.addDisposable(this.type = ko.pureComputed({
            read: () => {
                return this._type();
            },
            write: (val) => {
                var oldVal = this._type();
                if (val !== oldVal) {
                    this._value(undefined);
                    if (this.staticListLookUpSettings()) {
                        this.staticListLookUpSettings()._updateValuesType(val);
                    }
                    this._type(val);
                    if (val === 'System.DateTime') {
                        this.defaultValue(ParameterHelper.getDefaultTypeValue(val));
                    }
                    else {
                        this.defaultValue(ParameterHelper.convertSingleValue(this.defaultValue(), val, this.allowNull()));
                    }
                }
            }
        }));
        this.defaultValue(ParameterHelper.convertSingleValue(this.defaultValue(), this.type(), this.allowNull()));
        if (modelJson.hasOwnProperty(StaticListLookUpSettings.modelName)) {
            this.staticListLookUpSettings(new StaticListLookUpSettings(modelJson[StaticListLookUpSettings.modelName], serializer));
            delete this['_model'][StaticListLookUpSettings.modelName];
            this.staticListLookUpSettings()._updateValuesType(this.type());
            this.lookUpSourceType('StaticListLookUpSettings');
        }
        else if (modelJson.hasOwnProperty(DynamicListLookUpSettings.modelName)) {
            this.dynamicListLookUpSettings(new DynamicListLookUpSettings(modelJson[DynamicListLookUpSettings.modelName], serializer));
            delete this['_model'][DynamicListLookUpSettings.modelName];
            this.lookUpSourceType('DynamicListLookUpSettings');
        }
        else {
            this.lookUpSourceType('None');
        }
        this.addDisposable(this.lookUpSourceType = ko.computed({
            read: () => {
                if (!!this.staticListLookUpSettings()) {
                    return 'StaticListLookUpSettings';
                }
                else if (this.dynamicListLookUpSettings()) {
                    return 'DynamicListLookUpSettings';
                }
                else {
                    return 'None';
                }
            },
            write: (val) => {
                this.dynamicListLookUpSettings() && this.dynamicListLookUpSettings().dispose();
                switch (val) {
                    case 'StaticListLookUpSettings':
                        this.staticListLookUpSettings(new StaticListLookUpSettings());
                        this.dynamicListLookUpSettings(null);
                        this._resetDefaultValues();
                        break;
                    case 'DynamicListLookUpSettings':
                        this.staticListLookUpSettings(null);
                        this.dynamicListLookUpSettings(new DynamicListLookUpSettings());
                        this._resetDefaultValues();
                        break;
                    default:
                        this.staticListLookUpSettings(null);
                        this.dynamicListLookUpSettings(null);
                        this._resetDefaultValues();
                        this.allowMultiselect(false);
                        this.selectAllValues(false);
                        break;
                }
            }
        }));
        this.addDisposable(this._actualValue = ko.computed(() => {
            if ((!this.allowNull() && (this._value() === null || this._value() === undefined)) ||
                (this.allowNull() && this._value() === undefined)) {
                if (this.allowMultiselect()) {
                    if (this.selectAllValues()) {
                        if (!!this.staticListLookUpSettings()) {
                            return this.staticListLookUpSettings().values().map(val => val.value());
                        }
                        else if (!!this.dynamicListLookUpSettings()) {
                            return Parameter.SelectAllValue;
                        }
                    }
                    return this._valuesOfDefaultValues();
                }
                return this.defaultValue();
            }
            return this._value();
        }));
        this.addDisposable(this.containsDisplayMember = ko.computed(() => {
            return !!this.dynamicListLookUpSettings();
        }));
        this.addDisposable(this.defaultValue.subscribe(newDefaultValue => {
            if (!this.allowMultiselect()) {
                this._value(this.defaultValue());
            }
        }));
        this.addDisposable(ko.computed(() => {
            if (this.allowMultiselect()) {
                if (this.selectAllValues()) {
                    this._value(undefined);
                }
                else {
                    this._value(this._valuesOfDefaultValues());
                }
            }
            else {
                this._value(this.defaultValue());
            }
        }));
        this.addDisposable(this.allowMultiselect.subscribe(newAllowMultiselect => {
            if (!newAllowMultiselect) {
                this.selectAllValues(false);
            }
            this._resetDefaultValues();
        }));
        this.addDisposable(this.selectAllValues.subscribe(_ => {
            this._resetDefaultValues();
        }));
    }
    _patchSerializationsInfo(infos, propertyName, action) {
        var property = (infos.filter((prop) => { return prop.propertyName === propertyName; })[0]);
        if (!!property) {
            action(property);
        }
    }
    dispose() {
        super.dispose();
        this.dynamicListLookUpSettings() && this.dynamicListLookUpSettings().dispose();
    }
    getInfo() {
        var info = deepExtend([], dashboardParameterSerializationsInfo);
        if (this.type) {
            this._patchSerializationsInfo(info, defaultValue.propertyName, (prop) => { prop.defaultVal = ParameterHelper.getDefaultTypeValue(this.type()); });
        }
        if (this.staticListLookUpSettings && !!this.staticListLookUpSettings()) {
            this._patchSerializationsInfo(info, _staticListLookUpSettingsSerializationInfo.propertyName, (prop) => { prop.modelName = StaticListLookUpSettings.modelName; });
        }
        if (this.staticListLookUpSettings && !!this.dynamicListLookUpSettings()) {
            this._patchSerializationsInfo(info, _dynamicListLookUpSettingsSerializationInfo.propertyName, (prop) => { prop.modelName = DynamicListLookUpSettings.modelName; });
        }
        return info;
    }
    _resetDefaultValues() {
        this.defaultValue(ParameterHelper.convertSingleValue(null, this.type(), this.allowNull()));
        this.defaultValues(this.allowNull() ? null : []);
    }
    grabFrom(another) {
        this.name(another.name.peek());
        this._type(another._type.peek());
        this.allowNull(another.allowNull.peek());
        this.parameterVisible(another.parameterVisible.peek());
        this.description(another.description.peek());
        this.defaultValue(another.defaultValue.peek());
        this.allowMultiselect(another.allowMultiselect.peek());
        this.defaultValues(another.defaultValues.peek());
        this.selectAllValues(another.selectAllValues.peek());
        this.staticListLookUpSettings(another.staticListLookUpSettings.peek());
        this.dynamicListLookUpSettings(another.dynamicListLookUpSettings.peek());
    }
    _getDefaultItemType() {
        return 'Parameter';
    }
}
Parameter.SelectAllValue = '7BD68C11-DC21-4571-8EF6-AAB6E15355EF';
__decorate([
    collectionItemType('Value')
], Parameter.prototype, "defaultValues", void 0);
