// devexpress-dashboard/model/items/pivot/metadata/_pivot-item.js
/**
* DevExpress Dashboard (_pivot-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { columnsPropertyName, parseBool, rowsPropertyName, valuesPropertyName } from '../../../metadata/_base-metadata';
import { dataDashboardItemSerializationsInfo } from '../../metadata/_data-dashboard-item';
import { _baseInteractivityOptionsMeta } from '../../options/interactivity-options';
export let pivotColumns = { propertyName: columnsPropertyName, modelName: 'Columns', displayName: 'DashboardStringId.DescriptionColumns', array: true };
export let pivotRows = { propertyName: rowsPropertyName, modelName: 'Rows', displayName: 'DashboardStringId.DescriptionRows', array: true };
export let pivotValues = { propertyName: valuesPropertyName, modelName: 'Values', displayName: 'DashboardStringId.DescriptionValues', array: true };
export let autoExpandColumnGroups = { propertyName: 'autoExpandColumnGroups', modelName: 'AutoExpandColumnGroups', displayName: 'DashboardWebStringId.PivotGrid.AutoExpandColumnGroups', defaultVal: false, simpleFormAdapterItem: 'yesNoButtonGroupEditor', from: parseBool };
export let autoExpandRowGroups = { propertyName: 'autoExpandRowGroups', modelName: 'AutoExpandRowGroups', displayName: 'DashboardWebStringId.PivotGrid.AutoExpandRowGroups', defaultVal: false, simpleFormAdapterItem: 'yesNoButtonGroupEditor', from: parseBool };
export let showColumnGrandTotals = { propertyName: 'showColumnGrandTotals', modelName: 'ShowColumnGrandTotals', displayName: 'DashboardWebStringId.PivotGrid.ColumnGrandTotals', defaultVal: true, simpleFormAdapterItem: 'onOffButtonGroupEditor', from: parseBool };
export let showRowGrandTotals = { propertyName: 'showRowGrandTotals', modelName: 'ShowRowGrandTotals', displayName: 'DashboardWebStringId.PivotGrid.RowGrandTotals', defaultVal: true, simpleFormAdapterItem: 'onOffButtonGroupEditor', from: parseBool };
export let showColumnTotals = { propertyName: 'showColumnTotals', modelName: 'ShowColumnTotals', displayName: 'DashboardWebStringId.PivotGrid.ColumnTotals', defaultVal: true, simpleFormAdapterItem: 'onOffButtonGroupEditor', from: parseBool };
export let showRowTotals = { propertyName: 'showRowTotals', modelName: 'ShowRowTotals', displayName: 'DashboardWebStringId.PivotGrid.RowTotals', defaultVal: true, simpleFormAdapterItem: 'onOffButtonGroupEditor', from: parseBool };
export let valuesPosition = {
    propertyName: 'valuesPosition', modelName: 'ValuesPosition', displayName: 'DashboardWebStringId.PivotGrid.ValuesPosition', defaultVal: 'Columns', simpleFormAdapterItem: 'buttonGroupEditor',
    values: {
        'Columns': 'DashboardWebStringId.PivotGrid.ValuesPositionColumns',
        'Rows': 'DashboardWebStringId.PivotGrid.ValuesPositionRows'
    }
};
export let rowTotalsPosition = {
    propertyName: 'rowTotalsPosition', modelName: 'RowTotalsPosition', displayName: 'DashboardWebStringId.PivotGrid.RowTotalsPosition', defaultVal: 'Bottom', simpleFormAdapterItem: 'buttonGroupEditor',
    values: {
        'Bottom': 'DashboardWebStringId.PivotGrid.RowTotalsPositionBottom',
        'Top': 'DashboardWebStringId.PivotGrid.RowTotalsPositionTop'
    }
};
export let columnTotalsPosition = {
    propertyName: 'columnTotalsPosition', modelName: 'ColumnTotalsPosition', displayName: 'DashboardWebStringId.PivotGrid.ColumnTotalsPosition', defaultVal: 'Far', simpleFormAdapterItem: 'buttonGroupEditor',
    values: {
        'Far': 'DashboardWebStringId.PivotGrid.ColumnTotalsPositionFar',
        'Near': 'DashboardWebStringId.PivotGrid.ColumnTotalsPositionNear'
    }
};
export let layoutType = {
    propertyName: 'layoutType', modelName: 'LayoutType', displayName: 'DashboardWebStringId.PivotGrid.LayoutType', defaultVal: 'Compact', simpleFormAdapterItem: 'buttonGroupEditor',
    values: {
        'Compact': 'DashboardWebStringId.PivotGrid.LayoutTypeCompact',
        'Tabular': 'DashboardWebStringId.PivotGrid.LayoutTypeTabular'
    }
};
export let pivotDashboardItemSerializationsInfo = dataDashboardItemSerializationsInfo.concat([pivotColumns, pivotRows, pivotValues, autoExpandColumnGroups, autoExpandRowGroups, showColumnGrandTotals, showRowGrandTotals, showColumnTotals, showRowTotals, valuesPosition, rowTotalsPosition, columnTotalsPosition, layoutType, _baseInteractivityOptionsMeta]);
