// devexpress-dashboard/designer/currency-selector/_currency-selector.js
/**
* DevExpress Dashboard (_currency-selector.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { getLocalizationById } from '../../data/localization/_default';
import { DashboardFormatHelper } from '../../data/_format-helper';
import { subscribeWithPrev } from '../../model/internal/_knockout-utils';
import { CURRENCIES } from './_currencies';
export class CurrencyInfo {
    constructor(name, displayText, previewText = '') {
        this.name = name;
        this.displayText = displayText;
        this.previewText = previewText;
    }
}
export class CultureInfo {
}
export class CurrencySelector {
    constructor(currencyCultureName, disabled) {
        this.disabled = disabled;
        this._defaultCurrency = null;
        this._getDefaultCurrencyInfo = function () {
            if (!this._defaultCurrency) {
                this._defaultCurrency = new CurrencyInfo(null, getLocalizationById('DashboardStringId.DashboardCurrencyUseCurrentCurrency'), '');
                this._defaultCurrency.cultures = [{
                        name: null,
                        displayText: getLocalizationById('DashboardStringId.DashboardCurrencyUseCurrentCurrency')
                    }];
            }
            return this._defaultCurrency;
        };
        this.getPreviewText = function (value, currency) {
            if (currency) {
                var dashboardFormat = {
                    format: 'currency',
                    currency: currency
                };
                return DashboardFormatHelper.format(value, dashboardFormat);
            }
            else {
                return null;
            }
        };
        this.selectedCurrency = ko.observable(this._getDefaultCurrencyInfo());
        this.selectedCulture = ko.observable(this._getDefaultCurrencyInfo().cultures[0]);
        this.previewPositive = ko.pureComputed(() => this.getPreviewText(123, this.selectedCurrency().name));
        this.previewNegative = ko.pureComputed(() => this.getPreviewText(-123, this.selectedCurrency().name));
        this.previewWarning = ko.pureComputed(() => this.selectedCurrency().name ? null : getLocalizationById('DashboardWebStringId.Currency.DefaultCurrencyWarning'));
        this.currencies = ko.observableArray([this._getDefaultCurrencyInfo()].concat(CURRENCIES.map((currency) => {
            var currencyInfo = new CurrencyInfo(currency.name, currency.displayName, this.getPreviewText(123, currency.name));
            currencyInfo.cultures = currency.cultures.map(culture => {
                var cultureInfo = new CultureInfo();
                cultureInfo.name = culture.name;
                cultureInfo.displayText = culture.displayName;
                if (cultureInfo.name === currencyCultureName()) {
                    this.selectedCurrency(currencyInfo);
                    this.selectedCulture(cultureInfo);
                }
                return cultureInfo;
            });
            return currencyInfo;
        }).sort((c1, c2) => c1.displayText.localeCompare(c2.displayText))));
        subscribeWithPrev(this.selectedCurrency, (prevCurrencyInfo, currencyInfo) => {
            if (currencyInfo && prevCurrencyInfo !== currencyInfo) {
                this.selectedCulture(currencyInfo.cultures[0]);
            }
        });
        this.selectedCulture.subscribe(currencyCultureInfo => {
            if (currencyCultureInfo) {
                currencyCultureName(currencyCultureInfo.name);
            }
        });
    }
}
ko.components.register('dx-dashboard-currency-selector', {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            return new CurrencySelector(params.currencyCultureName, ko.observable(false));
        }
    },
    template: { element: 'dx-dash-currency-selector' }
});
