// devexpress-dashboard/common/_helpers.js
/**
* DevExpress Dashboard (_helpers.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { createJQueryDeferred } from '../data/_jquery-helpers';
import { getDimensionExpression } from '../model/data-item/_helpers';
import { getDataFields } from './_data-source-browser';
export function getDimensionsExpressions(dashboardItem, dataFieldProvider) {
    let deferred = createJQueryDeferred();
    let fieldsNames = dashboardItem._dimensions.map(d => d.dataMember());
    getDataFields(fieldsNames, dashboardItem.dataSource(), dashboardItem.dataMember(), dataFieldProvider)
        .done((fields) => {
        let dimensionsExpressions = {};
        dashboardItem._dimensions.map(dim => {
            let field = fields.filter(f => f.dataMember() === dim.dataMember())[0];
            dimensionsExpressions[dim.uniqueName()] = getDimensionExpression(dim.dataMember(), dim.dateTimeGroupInterval(), dim.textGroupInterval(), field.fieldType());
        });
        deferred.resolve(dimensionsExpressions);
    });
    return deferred.promise();
}
export class CancellationToken {
    constructor(source) {
        this._source = source;
    }
    static get None() {
        return new CancellationToken(createJQueryDeferred());
    }
    get canceled() {
        return this._source.state() === 'rejected';
    }
}
export function toCancelableSequenceAction(action) {
    let cancelableSequenceAction = action;
    if (cancelableSequenceAction.cancelableAction)
        return cancelableSequenceAction;
    else
        return Object.assign(Object.assign({}, action), { cancelableAction: (options, cancellationToken) => cancelableSequenceAction.action(options) });
}
export function combineOptions(previousOptions, options) {
    let result = Object.assign({}, previousOptions);
    options.forEach(op => {
        if (op.surfaceLeft !== previousOptions.surfaceLeft)
            result.surfaceLeft = op.surfaceLeft;
        if (op.surfaceTop !== previousOptions.surfaceTop)
            result.surfaceTop = op.surfaceTop;
    });
    return result;
}
export const requestParamsValidator = {
    isValid(requestParams) {
        return typeof requestParams.dashboardId !== 'symbol'
            && typeof requestParams.dashboardId !== 'object'
            || requestParams.dashboardId === null;
    }
};
