// devexpress-dashboard/common/resource-manager.js
/**
* DevExpress Dashboard (resource-manager.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { setLocalization } from '../data/localization/_localization-initializer';
import { defineObsoleteMethod } from '../model/internal/_obsolete-helper';
import '../resources/_bindings';
import { resources } from '../resources/_resources';
var resId = 'dx-dashboard-control-resources';
export class ResourceManager {
    static embedBundledResources() {
        this._embedBundledResourcesCore('callFromApi');
    }
    static removeEmbeddedResources() {
        this._removeEmbeddedResourcesCore('callFromApi');
    }
    static _embedBundledResourcesCore(keyElement) {
        if (!document.body)
            throw 'HTML <body> element is not created yet.';
        if (!document.getElementById(resId)) {
            let res = '';
            for (let key in resources) {
                res += resources[key];
            }
            var div = document.createElement('div');
            div.style.display = 'none';
            div.innerHTML = res;
            div.id = resId;
            document.body.insertBefore(div, document.body.childNodes[0]);
        }
        if (ResourceManager._resourceManagerKeys.indexOf(keyElement) === -1) {
            ResourceManager._resourceManagerKeys.push(keyElement);
        }
    }
    static _removeEmbeddedResourcesCore(keyElement) {
        let resourceElement = document.getElementById(resId);
        if (resourceElement) {
            var index = ResourceManager._resourceManagerKeys.indexOf(keyElement);
            if (index !== -1) {
                if (keyElement === 'callFromApi' || ResourceManager._resourceManagerKeys.length === 1) {
                    resourceElement.parentNode.removeChild(resourceElement);
                    ResourceManager._resourceManagerKeys = [];
                }
                else if (typeof keyElement !== 'string') {
                    ResourceManager._resourceManagerKeys.splice(index, 1);
                }
            }
        }
    }
    static setLocalizationMessages(localizationMessages) {
        setLocalization(localizationMessages);
    }
    static registerIcon(icon) {
        var div = document.createElement('div');
        div.style.position = 'absolute';
        div.style.visibility = 'hidden';
        div.style.left = '-10px';
        div.style.top = '-10px';
        div.style.width = '1px';
        div.style.height = '1px';
        if (typeof icon === 'string') {
            div.innerHTML = icon;
        }
        else {
            div.appendChild(icon);
        }
        document.body.insertBefore(div, document.body.childNodes[0]);
    }
}
ResourceManager._resourceManagerKeys = [];
defineObsoleteMethod({
    target: ResourceManager,
    memberName: 'removeEmbededResources',
    oldMemberDisplayName: 'ResourceManager.removeEmbededResources',
    newMemberDisplayName: 'ResourceManager.removeEmbeddedResources',
    action: () => ResourceManager.removeEmbeddedResources()
});
