// devexpress-dashboard/styles/_constants.js
/**
* DevExpress Dashboard (_constants.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
const materialGridBaseHeaderHeight = 40;
const materialGridBaseHeaderLineHeight = 16;
const materialGridBaseCellHeight = 36;
const materialGridBaseCellLineHeight = 16;
const materialPopupContentPadding = 16;
export const devExtremeConstants = {
    materialPopupToolbarHeight: 42,
    materialPopupContentPadding: materialPopupContentPadding,
    materialGridBaseHeaderHeight: materialGridBaseHeaderHeight,
    materialGridBaseHeaderLineHeight: materialGridBaseHeaderLineHeight,
    materialGridBaseHeaderCellVerticalPadding: Math.round((materialGridBaseHeaderHeight - materialGridBaseHeaderLineHeight) / 2),
    materialGridBaseCellHeight: materialGridBaseCellHeight,
    materialGridBaseCellLineHeight: materialGridBaseCellLineHeight,
    materialGridBaseCellVerticalPadding: Math.round((materialGridBaseCellHeight - materialGridBaseCellLineHeight) / 2),
    materialButtonHeight: 28,
    materialGridBaseCellFontSize: 13,
    materialGridBaseRowBorder: 1,
    dxPopupTitleBorderBottom: 1,
    dxDatagridHeadersBorder: 1,
};
export const dashboardConstants = {
    materialParametersDialogFormPaddingTop: 18,
    materialParametersDialogFormPaddingBottom: 30 - devExtremeConstants.materialPopupContentPadding,
};
