// devexpress-dashboard/model/items/filter-items/date-filter-item.js
/**
* DevExpress Dashboard (date-filter-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { deepExtend } from '../../../data/_jquery-helpers';
import { RangeFilterSelectionValidator } from '../../../viewer-parts/viewer-items/range-selector-item/_range-filter-selection-validator';
import { AcceptableShapingType } from '../../data-item/data-item';
import { collectionItemType } from '../../internal/_utils';
import { itemTypesMap } from '../../serializable-model';
import { DataDashboardItem } from '../data-dashboard-item';
import { _subscribeToDimension } from '../range-filter/_range-filter-item-helper';
import { RangeStateController } from '../range-filter/_range-state-controller';
import { DateTimePeriod } from '../range-filter/date-time-period';
import { dateFilterDashboardItemSerializationsInfo, dateFilterDimension } from './metadata/_date-filter-item';
export class DateFilterItem extends DataDashboardItem {
    constructor(dashboardItemJSON = {}, serializer = new ModelSerializer()) {
        super(dashboardItemJSON, serializer);
        this._defaultDateTimePeriodIndexSubscription = null;
        this.currentSelectedDateTimePeriodName = ko.observable();
        this._attachDataItem(this, dateFilterDimension.propertyName);
        this.__dimension._specifics.acceptableShapingType = AcceptableShapingType.RangeDate;
        this.dateTimePeriods = deserializeArray(dashboardItemJSON.DateTimePeriods, (item) => new DateTimePeriod(item, serializer));
        this.addDisposable(this._stateController = new RangeStateController(this));
        this._stateController.initialize();
        _subscribeToDimension(this.dimension, this.dateTimePeriods, this._dataManager);
    }
    _getSelectionByPeriod(period) {
        let range = period.getDateTimeValue();
        return [[range.startValue, range.endValue]];
    }
    _clearBindings() {
        super._clearBindings();
        this.disposeObservableArray(this.dateTimePeriods);
        this.dateTimePeriods.removeAll();
    }
    _getInfoCore() {
        return dateFilterDashboardItemSerializationsInfo;
    }
    _getDefaultItemType() {
        return 'DateFilter';
    }
    _getClearMasterFilterSupported() { return true; }
    _getIsMasterFilter() { return true; }
    _getIgnoreMasterFilter() { return this.interactivityOptions.ignoreMasterFilters(); }
    _getIsVisualInteractivitySupported() { return false; }
    _getInteractivityDimensionLinks() {
        return [this.__dimension];
    }
    _getDisplayFilterValues(limitCount) {
        return this._stateController._getDisplayFilterValues(limitCount);
    }
    _updateContentViewModel(content) {
        super._updateContentViewModel(content);
        (content.ViewModel.DateTimePeriods || []).forEach(p => {
            p.Start && p.Start.Date && (p.Start.Date = new Date(p.Start.Date));
            p.End && p.End.Date && (p.End.Date = new Date(p.End.Date));
        });
    }
    _setState(itemState) {
        this._stateController.setState(itemState);
    }
    _removeSelectionFromState(state) {
        let itemState = super._removeSelectionFromState(deepExtend({}, state));
        return this._stateController.removeSelectionFromState(itemState);
    }
    _setPredefinedPeriodToState(state, periodName) {
        let itemState = super._setPredefinedPeriodToState(deepExtend({}, state), periodName);
        return this._stateController.setPredefinedPeriodToState(itemState, periodName);
    }
    _setSelectionToState(state, selection) {
        return this._stateController.setSelectionToState(state, selection);
    }
    _getValidatedSelection(selectionValues) {
        let validatedValues = super._getValidatedSelection(selectionValues);
        if (validatedValues && this.dimension()) {
            let validatedRange = RangeFilterSelectionValidator.validate({
                startValue: validatedValues[0][0],
                endValue: validatedValues[0][1]
            }, this.dimension()._actualDateTimeGroupInterval() === 'Year');
            validatedValues[0] = [validatedRange.startValue, validatedRange.endValue];
        }
        return validatedValues;
    }
    _extendContentState(content) {
        super._extendContentState(content);
        this._stateController.extendContentState(content);
    }
    _hasSelection() {
        if (!!this.currentSelectedDateTimePeriodName())
            return true;
        else if (super._hasSelection()) {
            if (!!this._selectionValues()[0][0] || !!this._selectionValues()[0][1])
                return true;
        }
        return false;
    }
    _isSortingEnabled() {
        return false;
    }
    _isTopNEnabled(dataItem) {
        return false;
    }
    _getEntireRange() {
        if (this._multiData) {
            var measures = this._multiData.getMeasures();
            var values = [];
            for (var i = 0; i < measures.length; i++)
                values.push(this._multiData.getMeasureValue(measures[i].id).getValue());
            values.sort((a, b) => a > b ? 1 : -1);
            return values;
        }
        return [];
    }
}
__decorate([
    collectionItemType('DateTimePeriod')
], DateFilterItem.prototype, "dateTimePeriods", void 0);
itemTypesMap['DateFilter'] = { type: DateFilterItem, groupName: 'filter', title: 'DashboardStringId.DefaultNameDateFilterItem', index: 350 };
