// devexpress-richedit/bin/nspell.webpack.config.js
/**
* DevExpress WebRichEdit (nspell.webpack.config.js)
* Version: 23.2.7
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs
*/
const path = require('path');
const TerserPlugin = require('terser-webpack-plugin');
const source = path.resolve(__dirname, 'nspell-index.js');
const scriptFileName = 'nspell';
const libraryName = ['NSpell']; // this creates window.NSpell namespace for bundle

module.exports = {
    entry: {
        [scriptFileName]: source,
        [scriptFileName + '.min']: source,
    },
    output: {
        path: path.resolve(__dirname, '..', 'dist', 'custom'),
        filename: '[name].js',
        library: libraryName,
        libraryTarget: "self",
    },
    module: {
        rules: [
            {
                test: /\.(aff|dic)$/i,
                type: 'asset/source'
            }
        ]
    },
    optimization: {
        minimize: true,
        minimizer: [new TerserPlugin({test: /\.min\.js$/})],
    },
};
