// devexpress-dashboard/viewer-parts/viewer-items/_widget-viewer-item.js
/**
* DevExpress Dashboard (_widget-viewer-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { deepExtend } from '../../data/_jquery-helpers';
import { allowSelectValue } from '../../data/_utils';
import { dxWidgetsViewer as WidgetsViewer } from '../widgets/widgets-viewer/_widgets-viewer';
import { baseItem } from './_base-item';
export class widgetViewerItem extends baseItem {
    constructor(container, options) {
        super(container, options);
    }
    _clearSelectionUnsafe() {
        this.widgetsViewer.clearSelections();
    }
    getInfoUnsafe() {
        return deepExtend(super.getInfoUnsafe(), this.widgetsViewer.getSizeParams());
    }
    renderContentUnsafe(element, changeExisting, afterRenderCallback) {
        var that = this, options = that._getWidgetViewerOptions(), isInAsyncRendering = true;
        options.viewer.onAllItemsRenderComplete = function () {
            if (that.widgetsViewer) {
                afterRenderCallback();
            }
            else {
                isInAsyncRendering = false;
            }
        };
        if (changeExisting && that.widgetsViewer) {
            that.widgetsViewer.option(options);
        }
        else {
            that.widgetsViewer = that.createWidgetViewer(element, options);
        }
        return isInAsyncRendering;
    }
    createWidgetViewer(element, options) {
        return new WidgetsViewer(element, options);
    }
    _getContainerPositionUnsafe() {
        var that = this, position = super._getContainerPositionUnsafe(), itemInfo = that.getInfo(), scrollSize = itemInfo && itemInfo.scroll && itemInfo.scroll.vertical ? itemInfo.scroll.size : 0;
        position.offsetX -= scrollSize;
        return position;
    }
    _getSpecificWidgetViewerOptions() {
        return {
            itemOptions: {
                encodeHtml: this._isEncodeHtml(),
                itemWidgetOptionsPrepared: (options) => this._raiseItemWidgetOptionsPrepared(options)
            }
        };
    }
    _getWidgetType() {
        return null;
    }
    _isHoverEnabled() {
        return this._selectionMode() !== 'none';
    }
    _configureHover(selectionValues) {
        var hoverEnabled = selectionValues !== null && this._isHoverEnabled() && allowSelectValue(selectionValues);
        return {
            hoverEnabled,
            cursor: hoverEnabled ? 'pointer' : 'default'
        };
    }
    _getWidgetViewerOptions() {
        let viewModel = this.options.ViewModel;
        let contentDescription = viewModel ? viewModel.ContentDescription : undefined;
        if (this.dataController)
            this.dataController.setSourceItemProperties = (sourceItem, element, properties) => this._setSourceItemProperties(sourceItem, element, properties);
        let commonOptions = {
            dataSource: this._getDataSource(),
            viewer: {
                redrawOnResize: false,
                onclick: this._getOnClickHandler(),
                onhover: this._getOnHoverHandler(),
                widgetType: this._getWidgetType(),
                method: contentDescription ? this._convertContentArrangementMode(contentDescription.ArrangementMode) : 'auto',
                count: contentDescription ? contentDescription.LineCount : 1,
                supportAnimation: this._supportAnimation(),
            }
        };
        var resultOptions = deepExtend(commonOptions, this._getSpecificWidgetViewerOptions());
        this._ensureOptions(resultOptions);
        return resultOptions;
    }
    _supportAnimation() {
        return false;
    }
    _getDataSource() {
        if (this.dataController)
            return this.dataController.getDataSource();
    }
    _getElementInteractionValue(element, viewModel) {
        return element.tag;
    }
    _getOnClickHandler() {
        var that = this;
        return function (e) {
            that._raiseItemClick(e.item);
        };
    }
    _getOnHoverHandler() {
        var that = this;
        return function (e) {
            that._raiseItemHover(e.item, e.state);
        };
    }
    _convertContentArrangementMode(contentArrangementMode) {
        switch (contentArrangementMode) {
            case 'FixedColumnCount':
                return 'column';
            case 'FixedRowCount':
                return 'row';
            default:
                return 'auto';
        }
    }
    _resizeUnsafe() {
        super._resizeUnsafe();
        this.widgetsViewer.redraw();
    }
    updateContentStateUnsafe() {
        this.widgetsViewer.itemsList.forEach(viewer => {
            viewer.setHoverEnabledState(this._getCustomHoverEnabled());
        });
    }
    _setSourceItemProperties(sourceItem, elementModel, props) {
    }
    _isMultiDataSupported() {
        return true;
    }
    _ensureOptions(options) {
    }
    dispose() {
        super.dispose();
        if (this.widgetsViewer)
            this.widgetsViewer.dispose();
    }
}
