// devexpress-dashboard/model/format-rules/style-settings/appearance-settings.js
/**
* DevExpress Dashboard (appearance-settings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import { Color } from '../../color';
import { appearanceSettingsSerializationsInfo } from './metadata/_appearance-settings';
import { StyleSettingsBase } from './style-settings-base';
export class AppearanceSettings extends StyleSettingsBase {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return appearanceSettingsSerializationsInfo;
    }
    equals(style) {
        return (style instanceof AppearanceSettings) &&
            this.appearanceType() === style.appearanceType() &&
            this.fontFamily() == style.fontFamily() &&
            this.fontStyle() == style.fontStyle() &&
            Color.equals(this.backColor(), style.backColor()) &&
            Color.equals(this.foreColor(), style.foreColor());
    }
    clone() {
        return new AppearanceSettings({
            appearanceType: this.appearanceType(),
            fontFamily: this.fontFamily(),
            fontStyle: this.fontStyle(),
            backColor: this.backColor(),
            foreColor: this.foreColor()
        });
    }
    setSpecificType(type) {
        this.appearanceType(type);
    }
    init() {
        this.appearanceType('PaleRed');
    }
}
