// devexpress-dashboard/model/format-rules/style-settings/color-style-settings.js
/**
* DevExpress Dashboard (color-style-settings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import { Color } from '../../color';
import { colorStyleSettingsSerializationsInfo } from './metadata/_color-style-settings';
import { StyleSettingsBase } from './style-settings-base';
export class ColorStyleSettings extends StyleSettingsBase {
    constructor(modelJson = {}, serializer = new ModelSerializer()) {
        super(modelJson, serializer);
    }
    getInfo() {
        return colorStyleSettingsSerializationsInfo;
    }
    equals(style) {
        return (style instanceof ColorStyleSettings) &&
            this.predefinedColor() === style.predefinedColor() &&
            Color.equals(this.color(), style.color());
    }
    clone() {
        var style = new ColorStyleSettings({});
        style.predefinedColor(this.predefinedColor());
        style.color(this.color());
        return style;
    }
    setSpecificType(type) {
        this.predefinedColor(type);
    }
    init() {
        this.predefinedColor('Red');
    }
}
