// devexpress-dashboard/common/docking-layout/tab-container-item/_tab-container-item-bindings.js
/**
* DevExpress Dashboard (_tab-container-item-bindings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { debounce } from '../../../data/_utils';
import { DashboardUIItemStateController } from '../../viewer/_dashboard-item-bindings';
import { createItemSizeUpdater, setElementSize } from '../../viewer/_element-size-utils';
import { GroupViewModel } from '../group-item/_group-item-bindings';
import { DashboardTabsViewModel } from './_dashboard-tabs-view-model';
ko.bindingHandlers['dx-dashboard-layout-tab-binding'] = {
    init: function (element, valueAccessor, _, __, bindingContext) {
        var _disposables = [];
        var params = ko.unwrap(valueAccessor());
        element.classList.add('dx-dashboard-item-container');
        if (params.ignoreBorder())
            element.classList.add('dx-dashboard-ignore-border');
        let sizeController = params.sizeController;
        setElementSize(element, sizeController);
        _disposables.push(params.sizeController.width.subscribe(newValue => { setElementSize(element, sizeController); }));
        _disposables.push(params.sizeController.height.subscribe(newValue => { setElementSize(element, sizeController); }));
        _disposables.push(params.ignoreBorder.subscribe(newValue => {
            if (newValue) {
                element.classList.add('dx-dashboard-ignore-border');
            }
            else {
                element.classList.remove('dx-dashboard-ignore-border');
            }
        }));
        var uiStateController = new DashboardUIItemStateController(element, params.dashboardItem, params.sizeController, bindingContext, () => {
            var itemViewerAdapter = params.dashboardContext.viewerItemsManager.create(params.dashboardItem, element, params.dashboardContext, params.localContext, viewerItem => {
                setElementSize(element, sizeController);
                _disposables.push(createItemSizeUpdater(viewerItem, params.sizeController));
            });
            itemViewerAdapter.initialize();
            return itemViewerAdapter;
        }, ['dx-layout-item-empty-inside-tab']);
        uiStateController.render();
        ko.utils.domNodeDisposal.addDisposeCallback(element, function () {
            _disposables.forEach(d => d.dispose());
            uiStateController.dispose();
        });
        return { controlsDescendantBindings: true };
    }
};
export class TabContainerViewModel extends GroupViewModel {
    constructor(params) {
        super(params);
        this.activeItems = ko.observableArray();
        this._activeItemsUpdateDebounced = debounce(() => {
            this._updateActiveItems();
        }, 1);
        this._updateActiveItems();
        let isItemShownInsteadTabPage = ko.computed(() => {
            let activeLayoutItem = this.layoutItem().items().length > 0 ? this.layoutItem().items()[0] : undefined;
            if (!activeLayoutItem)
                return false;
            return !!activeLayoutItem._parent().viewModel.activeTabPage()
                && activeLayoutItem._parent().viewModel.activeTabPage() !== activeLayoutItem.viewModel.item();
        });
        this.addDisposable(isItemShownInsteadTabPage);
        this.addDisposable(ko.computed(() => {
            let newPadding = this.defaultPadding;
            if (isItemShownInsteadTabPage()) {
                newPadding = this.layoutItem().isDesignMode() ? 2 : 0;
            }
            this.padding(newPadding);
        }));
        this.addDisposable((this.layoutItem().items.subscribe(newValue => {
            this.activeItems.removeAll();
            this._activeItemsUpdateDebounced();
        })));
    }
    _updateActiveItems() {
        this.layoutItem().items().forEach(item => {
            this.activeItems.push(item);
        });
    }
}
ko.components.register('dx-dashboard-layout-tab-container', {
    viewModel: TabContainerViewModel,
    template: { element: 'dx-dashboard-layout-tab-container' }
});
ko.components.register('dashboard-layout-tabs', {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            return new DashboardTabsViewModel(params.layoutItem, params.headerHeight, componentInfo.element);
        }
    },
    template: { element: 'dx-dashboard-layout-tabs' }
});
