// devexpress-dashboard/model/parameters/metadata/_dynamic-list-lookup-settings.js
/**
* DevExpress Dashboard (_dynamic-list-lookup-settings.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export let dataSourceNameSerialization = { propertyName: 'dataSource', modelName: '@DataSourceName', defaultVal: null };
export let dataMemberSerialization = { propertyName: 'dataMember', modelName: '@DataMember', defaultVal: null };
export let valueMemberSerialization = { propertyName: 'valueMemberName', modelName: '@ValueMember', displayName: 'DevExpress.DashboardCommon.DynamicListLookUpSettings.ValueMember', defaultVal: null };
export let displayMemberSerialization = { propertyName: 'displayMemberName', modelName: '@DisplayMember', displayName: 'DevExpress.DashboardCommon.DynamicListLookUpSettings.DisplayMember', defaultVal: null };
export let sortByMember = { propertyName: 'sortByMember', modelName: '@SortByMember', displayName: 'DashboardWebStringId.Parameters.SortByMember', defaultVal: null };
export let sortOrder = {
    propertyName: 'sortOrder', modelName: '@SortOrder', displayName: 'DevExpress.DashboardCommon.DynamicListLookUpSettings.SortOrder', defaultVal: 'Ascending', simpleFormAdapterItem: 'listEditor', values: {
        'Ascending': 'DevExpress.DashboardCommon.DimensionSortOrder.Ascending',
        'Descending': 'DevExpress.DashboardCommon.DimensionSortOrder.Descending',
        'None': 'DevExpress.DashboardCommon.DimensionSortOrder.None'
    }
};
export let _dataMemberSerialization = { propertyName: '_dataMember', displayName: 'DevExpress.DashboardCommon.DynamicListLookUpSettings.DataMember' };
export let _dataSourceNameSerialization = { propertyName: '_dataSource', displayName: 'DevExpress.DashboardCommon.DynamicListLookUpSettings.DataSource' };
export let dynamicListLookUpSettingsSerializationsInfo = [_dataSourceNameSerialization, dataSourceNameSerialization, _dataMemberSerialization, dataMemberSerialization, valueMemberSerialization, displayMemberSerialization, sortByMember, sortOrder];
