// devexpress-dashboard/model/items/grid/grid-options.js
/**
* DevExpress Dashboard (grid-options.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import { SerializableModel } from '../../serializable-model';
import { gridOptionsSerializationsInfo } from './metadata/_grid-options';
export class GridOptions extends SerializableModel {
    constructor(JSON = {}, serializer = new ModelSerializer()) {
        super(JSON, serializer);
    }
    getInfo() {
        return gridOptionsSerializationsInfo;
    }
    _getViewModel() {
        return {
            AllowCellMerge: this.allowCellMerge(),
            ColumnWidthMode: this.columnWidthMode(),
            EnableBandedRows: this.enableBandedRows(),
            ShowHorizontalLines: this.showHorizontalLines(),
            ShowVerticalLines: this.showVerticalLines(),
            ShowColumnHeaders: this.showColumnHeaders(),
            WordWrap: this.wordWrap()
        };
    }
}
