// devexpress-dashboard/model/items/custom-item/metadata/_custom-item.js
/**
* DevExpress Dashboard (_custom-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { PropertyCategory } from '../../../metadata/_base-metadata';
import { dataDashboardItemSerializationsInfo } from '../../metadata/_data-dashboard-item';
export let customItemType = { propertyName: 'customItemType', modelName: '@CustomItemType' };
export let sliceTables = { propertyName: 'sliceTables', modelName: 'SliceTables', array: true, category: PropertyCategory.Data };
export let interactivityTargets = { propertyName: 'interactivityTargets', modelName: 'InteractivityTargets', array: true };
export let coloringDimensions = { propertyName: 'coloringDimensions', modelName: 'ColoringDimensions', array: true };
export let coloringMeasures = { propertyName: 'coloringMeasures', modelName: 'ColoringMeasures', array: true };
export let customDashboardItemSerializationsInfo = dataDashboardItemSerializationsInfo.concat([customItemType, sliceTables, interactivityTargets]);
