// devexpress-dashboard/model/items/pivot/pivot-item.js
/**
* DevExpress Dashboard (pivot-item.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { deserializeArray, ModelSerializer } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { AcceptableShapingType, DataItemLink } from '../../data-item/data-item';
import { Measure } from '../../data-item/measure';
import { showGrandTotals, showTotals, showValues } from '../../data-item/metadata/_data-item';
import { PivotWindowDefinition } from '../../data-item/window-definition/pivot-calc-window-definition';
import { collectionItemType } from '../../internal/_utils';
import { itemTypesMap } from '../../serializable-model';
import { DataDashboardItem } from '../data-dashboard-item';
import { pivotColumns, pivotDashboardItemSerializationsInfo, pivotRows, pivotValues } from './metadata/_pivot-item';
export class PivotItem extends DataDashboardItem {
    constructor(dashboardItemJSON = {}, serializer = new ModelSerializer()) {
        super(dashboardItemJSON, serializer);
        this.columns = ko.observableArray([]);
        this.rows = ko.observableArray([]);
        this.values = ko.observableArray([]);
        this.__columns = deserializeArray(dashboardItemJSON.Columns, (item) => new DataItemLink(this, item, serializer));
        this.__rows = deserializeArray(dashboardItemJSON.Rows, (item) => new DataItemLink(this, item, serializer));
        this.__values = deserializeArray(dashboardItemJSON.Values, (item) => new DataItemLink(this, item, serializer));
        this._subscribeDataItemLinkArrays(pivotColumns, pivotRows, pivotValues);
        this.addDisposable(this.__values.subscribe((newValue) => {
            this._setCustomOptionsPropertyOnMeasures(newValue);
            this._setLinkCollectionAcceptableShapingType(newValue, AcceptableShapingType.String);
        }));
        this.addDisposable(this.__rows.subscribe((newValue) => {
            this._setCustomOptionsPropertyOnDimensions(newValue);
        }));
        this.addDisposable(this.__columns.subscribe((newValue) => {
            this._setCustomOptionsPropertyOnDimensions(newValue);
        }));
        this._setCustomOptionsPropertyOnMeasures(this.__values());
        this._setLinkCollectionAcceptableShapingType(this.__values(), AcceptableShapingType.String);
        this._setCustomOptionsPropertyOnDimensions(this.__rows());
        this._setCustomOptionsPropertyOnDimensions(this.__columns());
        this.addDisposable(this.autoExpandColumnGroups.subscribe((newValue) => {
            this._expandingManager.resetExpandingParams();
            this._expandingManager.resetColumnViewState(newValue);
        }));
        this.addDisposable(this.autoExpandRowGroups.subscribe((newValue) => {
            this._expandingManager.resetExpandingParams();
            this._expandingManager.resetRowViewState(newValue);
        }));
    }
    _clearBindings() {
        super._clearBindings();
        this.disposeObservableArray(this.__columns);
        this.disposeObservableArray(this.__rows);
        this.disposeObservableArray(this.__values);
        this.__columns.removeAll();
        this.__rows.removeAll();
        this.__values.removeAll();
    }
    _getInfoCore() {
        return pivotDashboardItemSerializationsInfo;
    }
    _canChangeDataItemVisibilityOptions(dataItem) {
        return this.__values().some(value => value.dataItem() === dataItem) || this.__rows().some(value => value.dataItem() === dataItem) || this.__columns().some(value => value.dataItem() === dataItem);
    }
    _isTotalsVisibilityOptionEnabled(dataItem) {
        if (dataItem instanceof Measure)
            return true;
        var localRows = this.__rows();
        if (localRows.length > 0 && localRows.some(row => row.dataItem() == dataItem)) {
            return localRows.length > 0 && this.layoutType() != 'Compact' && dataItem != localRows[localRows.length - 1].dataItem();
        }
        var localColumns = this.__columns();
        return localColumns.length > 0 && localColumns.some(column => column.dataItem() == dataItem) && dataItem != localColumns[localColumns.length - 1].dataItem();
    }
    _setCustomOptionsPropertyOnMeasures(links) {
        links.forEach((value) => {
            value._specifics.customOptionsProperties = [{
                    serializationInfo: showValues
                }, {
                    serializationInfo: showTotals
                }, {
                    serializationInfo: showGrandTotals
                }];
        });
    }
    _setCustomOptionsPropertyOnDimensions(links) {
        links.forEach((row) => {
            row._specifics.customOptionsProperties = [{
                    serializationInfo: showTotals,
                    filter: (dataItem) => { return this._canChangeDataItemVisibilityOptions(dataItem); },
                    disabledRule: (dataItem) => { return !this._isTotalsVisibilityOptionEnabled(dataItem); }
                }];
        });
    }
    _getDefaultItemType() {
        return 'Pivot';
    }
    _extendContentState(content) {
        content.PivotExpandViewState = this._expandingManager.getPivotExpandViewState();
    }
    _getIgnoreMasterFilter() { return this.interactivityOptions.ignoreMasterFilters(); }
    _getDefaultCalculationWindowDefinition() {
        return new PivotWindowDefinition();
    }
    _getIsVisualInteractivitySupported() { return false; }
}
__decorate([
    collectionItemType('Column')
], PivotItem.prototype, "__columns", void 0);
__decorate([
    collectionItemType('Row')
], PivotItem.prototype, "__rows", void 0);
__decorate([
    collectionItemType('Value')
], PivotItem.prototype, "__values", void 0);
itemTypesMap['Pivot'] = { type: PivotItem, groupName: 'common', title: 'DashboardStringId.DefaultNamePivotItem', index: 20 };
