// devexpress-dashboard/viewer-parts/viewer-items/elements/_item-loading.js
/**
* DevExpress Dashboard (_item-loading.js)
* Version:  23.2.7
* Build date: Jun 25, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import dxOverlay from 'devextreme/ui/overlay/ui.overlay';
import 'devextreme/ui/toast';
import * as $ from 'jquery';
import { getLocalizationById } from '../../../data/localization/_default';
import { $unwrap } from '../../../data/_jquery-helpers';
import { zIndex } from '../../../data/_z-index';
export class ItemLoadingElement {
    show(container) {
        if (!this._overlay) {
            var overlayDiv = document.createElement('div');
            var contentTemplate = document.createElement('div');
            contentTemplate.classList.add('dx-dashboard-item-loading-panel');
            var h1 = document.createElement('h1');
            h1.innerText = getLocalizationById('DashboardStringId.MessageLoading');
            contentTemplate.appendChild(h1);
            this._overlay = new dxOverlay(overlayDiv, {
                shading: true,
                animation: {
                    show: { type: 'fade', duration: 150, from: 0, to: 1, delay: 150 },
                    hide: { type: 'fade', duration: 150, to: 0, delay: 150 }
                },
                width: 'auto',
                height: 'auto',
                target: container,
                container: container,
                contentTemplate: contentTemplate,
                hideOnOutsideClick: false,
                propagateOutsideClick: true,
                position: {
                    boundary: container,
                    my: 'bottom right',
                    at: 'bottom right',
                    offset: '-10 -10',
                    of: container
                },
                onShowing: (e) => {
                    let wrapper = $unwrap(e.component._$wrapper);
                    wrapper.style.zIndex = (zIndex.dashboardItemShield - 2).toString();
                    let overlayContent = wrapper.querySelector('.dx-overlay-content');
                    if (overlayContent) {
                        overlayContent.style.zIndex = (zIndex.dashboardItemShield - 1).toString();
                    }
                },
                wrapperAttr: {
                    class: 'dx-dashboard-loading-indicator'
                }
            });
            overlayDiv.style.zIndex = zIndex.dashboardItemShield.toString();
            $.fn.constructor(container).prepend(overlayDiv);
        }
        this._overlay.show();
    }
    hide() {
        if (this._overlay) {
            this._overlay.hide();
        }
    }
    resize() {
        if (this._overlay) {
            this._overlay.repaint();
        }
    }
}
